<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Microbial Genomics Workshop: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/dc/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/dc/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/dc/favicon-16x16.png">
<link rel="manifest" href="favicons/dc/site.webmanifest">
<link rel="mask-icon" href="favicons/dc/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav data"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="dc-logo" alt="Data Carpentry" src="assets/images/data-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="assets/images/data-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Microbial Genomics Workshop
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics Workshop
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="resources.html">Resources</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics Workshop
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-why-cloud-computing.html">1. Computing Concepts &amp; Servers</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-shell.html">2. Introducing the Shell</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-the-filesystem.html">3. Navigating the Filesystem</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-working-with-files.html">4. Working with Files and Scripts</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-tidiness.html">5. Project Tidiness</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-project-planning.html">6. Planning a Genomics Project</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-background.html">7. Sequencing Data Background</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-quality-control.html">8. Assessing Read Quality (FastQC)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-trimming.html">9. Trimming Reads (fastp)</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-assembly-spades.html">10. De Novo Assembly with SPAdes</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-annotation-prokka.html">11. Genome Annotation with Prokka</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="resources.html">Resources</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-why-cloud-computing"><p>Content from <a href="01-why-cloud-computing.html">Computing Concepts &amp; Servers</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/01-why-cloud-computing.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we use a remote server for microbial genomics?</li>
<li>How is a server different from my laptop?</li>
<li>How will I connect to and work on the workshop server?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the difference between local and remote computing.</li>
<li>Identify the main hardware resources on a server (CPU, RAM,
storage).</li>
<li>Interpret the basic information shown when you log in to a Linux
server.</li>
<li>Connect to the workshop server using <code>ssh</code> or a terminal
program.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Introduction to
Cloud Computing for Genomics</em> and <em>Introduction to the Command
Line for Genomics</em> lessons.</p>
</blockquote>
<section><h2 class="section-heading" id="local-vs-remote-computing">Local vs remote computing<a class="anchor" aria-label="anchor" href="#local-vs-remote-computing"></a>
</h2>
<hr class="half-width">
<p>Most people are used to working on a <strong>local computer</strong>:
a laptop or desktop they can see and touch. For genomics, local machines
are often not enough:</p>
<ul>
<li>Sequencing data files can be <em>many gigabytes</em> in size.</li>
<li>Assembly and mapping tools can use <em>tens of gigabytes of
RAM</em>.</li>
<li>Analyses may run for <em>hours</em> or <em>days</em>.</li>
</ul>
<p>A <strong>server</strong> (or <em>remote machine</em>) is just a more
powerful computer that you log into over the network. In this workshop,
everyone will use the <strong>same type of remote Linux server</strong>,
pre-configured with the tools and data we need.</p>
<div id="vocabulary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="vocabulary" class="callout-inner">
<h3 class="callout-title">Vocabulary</h3>
<div class="callout-content">
<ul>
<li>
<strong>Local machine</strong> – the computer in front of you.</li>
<li>
<strong>Remote server</strong> – a computer in a different physical
location that you connect to over the network.</li>
<li>
<strong>Client</strong> – the program on your local machine that
talks to the server (Terminal, PuTTY, Windows Terminal, etc.).</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="what-resources-does-a-server-have">What resources does a server have?<a class="anchor" aria-label="anchor" href="#what-resources-does-a-server-have"></a>
</h2>
<hr class="half-width">
<p>Just like your laptop, a server has:</p>
<ul>
<li>
<strong>CPUs</strong> (cores): do the actual calculations.</li>
<li>
<strong>RAM</strong> (memory): holds data and code <em>while</em>
programs are running.</li>
<li>
<strong>Disk storage</strong>: where files live long-term (reads,
assemblies, scripts).</li>
<li>
<strong>Network connection</strong>: lets you log in and transfer
data.</li>
</ul>
<p>Genomics workloads tend to be limited by <strong>RAM</strong> and
<strong>disk</strong>. It is useful to know that:</p>
<ul>
<li>If you run out of RAM, programs may crash or become very slow.</li>
<li>If you run out of disk space, tools may refuse to run or fail with
messages like <em>No space left on device</em>.</li>
</ul>
<p>Your instructor or local sysadmin can tell you typical limits on the
system you are using.</p>
</section><section><h2 class="section-heading" id="connecting-to-the-workshop-server">Connecting to the workshop server<a class="anchor" aria-label="anchor" href="#connecting-to-the-workshop-server"></a>
</h2>
<hr class="half-width">
<p>Your instructor will provide:</p>
<ul>
<li>a <strong>host name</strong> (something like
<code>genomics-server.university.edu</code>), and</li>
<li>a <strong>username</strong> and <strong>password</strong> for your
workshop account.</li>
</ul>
<div class="section level3">
<h3 id="macos-and-linux">macOS and Linux<a class="anchor" aria-label="anchor" href="#macos-and-linux"></a>
</h3>
<p>On macOS and Linux, we use the built-in <strong>Terminal</strong>
application and the <code>ssh</code> command.</p>
<p>Open a terminal and type (replacing the pieces in angle brackets with
the values your instructor gives you):</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>server-address<span class="op">&gt;</span></span></code></pre>
</div>
<p>For example:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ssh</span> learner01@genomics-server.university.edu</span></code></pre>
</div>
<p>The first time you connect, you may see a security question about the
host key; answer <code>yes</code>. Then enter the password when
prompted. Note that the cursor will not move while you type the password
– this is normal.</p>
</div>
<div class="section level3">
<h3 id="windows">Windows<a class="anchor" aria-label="anchor" href="#windows"></a>
</h3>
<p>On Windows, you can use a terminal program such as:</p>
<ul>
<li>
<strong>Windows Terminal</strong> or <strong>PowerShell</strong> (on
recent Windows 10/11),</li>
<li>
<strong>PuTTY</strong>, or</li>
<li>
<strong>MobaXterm</strong>.</li>
</ul>
<p>Your instructor will demonstrate the recommended option for this
workshop. In all cases you will enter the <strong>host name</strong>,
<strong>username</strong>, and <strong>password</strong> that were
provided.</p>
</div>
</section><section><h2 class="section-heading" id="first-look-at-the-shell-prompt">First look at the shell prompt<a class="anchor" aria-label="anchor" href="#first-look-at-the-shell-prompt"></a>
</h2>
<hr class="half-width">
<p>After you log in, you will see a message of the day followed by a
line that ends with a <strong>prompt</strong>, often something like:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
<p>This usually shows:</p>
<ul>
<li>your <strong>username</strong>,</li>
<li>the <strong>server name</strong>, and</li>
<li>your current location in the filesystem (<code>~</code> means your
home directory).</li>
</ul>
<p>Everything you type from now on will be interpreted as
<strong>commands</strong> to the server, not your local machine.</p>
<p>Try the following commands to verify where you are:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">whoami</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">hostname</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<ul>
<li>
<code>whoami</code> prints your username on the remote server.</li>
<li>
<code>hostname</code> prints the name of the computer you are logged
into.</li>
<li>
<code>pwd</code> prints the present working directory (your current
folder).</li>
</ul>
<div id="exercise-local-or-remote" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-local-or-remote" class="callout-inner">
<h3 class="callout-title">Exercise: Local or remote?</h3>
<div class="callout-content">
<p>With a partner, decide whether each of the following is running on
your <strong>local machine</strong> or the <strong>remote
server</strong>:</p>
<ol style="list-style-type: decimal">
<li>The browser window showing these lesson pages.</li>
<li>The terminal or SSH client you are using.</li>
<li>The <code>whoami</code> and <code>pwd</code> commands you just
ran.</li>
<li>The FastQC and SPAdes runs we will do later.</li>
</ol>
<p>Discuss why it matters where each thing is happening.</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>We use a <strong>remote server</strong> for genomics because it has
more CPU, RAM, and storage than a typical laptop.</li>
<li>You connect to the server using an <strong>SSH client</strong>
(Terminal, PuTTY, MobaXterm, etc.).</li>
<li>After logging in, you type commands at a <strong>shell
prompt</strong> that is running on the <strong>remote</strong>
machine.</li>
<li>Simple commands like <code>whoami</code>, <code>hostname</code>, and
<code>pwd</code> help you confirm where you are and who you are logged
in as.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-introduction-to-shell"><p>Content from <a href="02-introduction-to-shell.html">Introducing the Shell</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/02-introduction-to-shell.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the Unix shell?</li>
<li>How do I run basic commands on the remote server?</li>
<li>How do I find help for a command?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what the shell is and why genomics workflows rely on
it.</li>
<li>Run basic commands such as <code>whoami</code>, <code>pwd</code>,
and <code>ls</code>.</li>
<li>Describe the general structure of a shell command.</li>
<li>Use tab completion and command history to work more
efficiently.</li>
<li>Use <code>--help</code> and <code>man</code> to find basic usage
information.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Introduction to
the Command Line for Genomics</em> lesson.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-the-shell">What is the shell?<a class="anchor" aria-label="anchor" href="#what-is-the-shell"></a>
</h2>
<hr class="half-width">
<p>The <strong>shell</strong> is a program that takes the commands you
type and asks the operating system to run them. Instead of pointing and
clicking in a graphical interface, you work by typing <strong>text
commands</strong>.</p>
<p>For genomics work:</p>
<ul>
<li>Many bioinformatics tools are only available on the command
line.</li>
<li>Many tasks involve many files or samples, which are easier to handle
with scripts.</li>
<li>Remote servers often only provide a text interface.</li>
</ul></section><section><h2 class="section-heading" id="the-shell-prompt">The shell prompt<a class="anchor" aria-label="anchor" href="#the-shell-prompt"></a>
</h2>
<hr class="half-width">
<p>On our server, your prompt will look something like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
<p>We will write commands in the lesson like this:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>You should <strong>type only what appears after the
<code>$</code></strong> and then press <kbd>Enter</kbd>.</p>
</section><section><h2 class="section-heading" id="your-first-commands">Your first commands<a class="anchor" aria-label="anchor" href="#your-first-commands"></a>
</h2>
<hr class="half-width">
<p>Try these commands on the server:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">whoami</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<ul>
<li>
<code>whoami</code> shows your username.</li>
<li>
<code>pwd</code> (<em>print working directory</em>) shows where you
are in the filesystem.</li>
<li>
<code>ls</code> lists files and directories in your current
location.</li>
</ul>
<p>Run <code>ls</code> a few times in a row – nothing changes, but this
is a good way to get used to the cycle of <strong>type → Enter → see
output</strong>.</p>
</section><section><h2 class="section-heading" id="the-structure-of-a-command">The structure of a command<a class="anchor" aria-label="anchor" href="#the-structure-of-a-command"></a>
</h2>
<hr class="half-width">
<p>Most shell commands follow this pattern:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">command</span>  <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span>  <span class="pp">[</span><span class="ss">arguments</span><span class="pp">]</span></span></code></pre>
</div>
<ul>
<li>
<strong>command</strong> – the program you are running
(e.g. <code>ls</code>)</li>
<li>
<strong>options</strong> – extra flags that change behavior (usually
start with <code>-</code>)</li>
<li>
<strong>arguments</strong> – what the command should act on (files,
directories, etc.)</li>
</ul>
<p>For example:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> shell_data</span></code></pre>
</div>
<ul>
<li>
<code>ls</code> – list files</li>
<li>
<code>-lh</code> – options: <code>-l</code> (long format) and
<code>-h</code> (human-readable sizes)</li>
<li>
<code>shell_data</code> – argument: which directory to list</li>
</ul></section><section><h2 class="section-heading" id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<hr class="half-width">
<p>Most commands provide some form of help. Two common patterns:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">--help</span></span></code></pre>
</div>
<p>Many programs print a short usage message with <code>--help</code> or
<code>-h</code>.</p>
<p>You can also use manual pages:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">man</span> ls</span></code></pre>
</div>
<ul>
<li>Use the arrow keys or <kbd>Space</kbd> to scroll.</li>
<li>Press <kbd>q</kbd> to quit the manual page.</li>
</ul>
<p>Not every command has a man page, but many core tools do.</p>
</section><section><h2 class="section-heading" id="working-more-efficiently-history-and-tab-completion">Working more efficiently: history and tab completion<a class="anchor" aria-label="anchor" href="#working-more-efficiently-history-and-tab-completion"></a>
</h2>
<hr class="half-width">
<p>The shell keeps a <strong>history</strong> of commands you have
typed.</p>
<ul>
<li>Press the <strong>up arrow</strong> to cycle backwards through
previous commands.</li>
<li>Press the <strong>down arrow</strong> to move forward again.</li>
</ul>
<p>You can also use <strong>tab completion</strong>:</p>
<ol style="list-style-type: decimal">
<li>Type <code>ls she</code> and then press <kbd>Tab</kbd>.</li>
<li>If there is only one match (e.g. <code>shell_data</code>), the shell
completes it.</li>
<li>If there are multiple matches, press <kbd>Tab</kbd> twice to see
them.</li>
</ol>
<p>Tab completion works for:</p>
<ul>
<li>command names,</li>
<li>directory names, and</li>
<li>file names.</li>
</ul>
<div id="exercise-inspecting-the-environment" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspecting-the-environment" class="callout-inner">
<h3 class="callout-title">Exercise: Inspecting the environment</h3>
<div class="callout-content">
<p>On the remote server, run the following commands:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">whoami</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">hostname</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span></span></code></pre>
</div>
<ol style="list-style-type: decimal">
<li>What does each command tell you?</li>
<li>What do you think the <code>-h</code> option to <code>ls</code>
does?</li>
<li>Try typing <code>ls -l</code> and then press the up arrow and edit
the command to add <code>h</code> (so it becomes
<code>ls -lh</code>).</li>
</ol>
<p>Discuss your answers with a neighbor.</p>
</div>
</div>
</div>
<div id="exercise-getting-help" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-getting-help" class="callout-inner">
<h3 class="callout-title">Exercise: Getting help</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Use <code>--help</code> and/or <code>man</code> to answer these
questions:</p>
<ul>
<li>Which option to <code>ls</code> shows hidden files (those whose
names start with <code>.</code>)?</li>
<li>Which option sorts files by size?</li>
</ul>
</li>
<li><p>Try out your answers on the command line.</p></li>
</ol>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The <strong>shell</strong> lets you control a computer by typing
commands instead of using a graphical interface.</li>
<li>A typical command has the form
<code>command [options] [arguments]</code>.</li>
<li>
<code>whoami</code>, <code>pwd</code>, and <code>ls</code> are
useful first commands to confirm where you are and what is in a
directory.</li>
<li>Use <code>--help</code> or <code>man</code> to get basic information
about most commands.</li>
<li>Command history and tab completion save time and reduce typing
errors.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-the-filesystem"><p>Content from <a href="03-the-filesystem.html">Navigating the Filesystem</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/03-the-filesystem.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How is data organized on the server?</li>
<li>How do I move between directories?</li>
<li>What is the difference between absolute and relative paths?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the filesystem as a tree of directories and files.</li>
<li>Use <code>pwd</code>, <code>ls</code>, and <code>cd</code> to move
around the filesystem.</li>
<li>Distinguish between absolute and relative paths.</li>
<li>Use shortcuts like <code>.</code>, <code>..</code>, and
<code>~</code> when specifying paths.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Navigating Files
and Directories</em> episode in <em>Introduction to the Command Line for
Genomics</em>.</p>
</blockquote>
<section><h2 class="section-heading" id="the-filesystem-as-a-tree">The filesystem as a tree<a class="anchor" aria-label="anchor" href="#the-filesystem-as-a-tree"></a>
</h2>
<hr class="half-width">
<p>The server stores data in a <strong>filesystem</strong>, which you
can picture as a tree:</p>
<ul>
<li>the <strong>root</strong> (<code>/</code>) at the top,</li>
<li>directories (folders) as branches,</li>
<li>files as leaves.</li>
</ul>
<p>Your personal working area is your <strong>home directory</strong>.
On this server, it will look something like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">/home/learner01</span></span></code></pre>
</div>
<p>The shell uses <strong>paths</strong> (like addresses) to locate
files and directories.</p>
</section><section><h2 class="section-heading" id="where-are-we-pwd">Where are we? <code>pwd</code>
<a class="anchor" aria-label="anchor" href="#where-are-we-pwd"></a>
</h2>
<hr class="half-width">
<p>Use <code>pwd</code> to <strong>print the working
directory</strong>:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>You should see something like <code>/home/learner01</code>. This is
your starting point.</p>
</section><section><h2 class="section-heading" id="what-is-here-ls">What is here? <code>ls</code>
<a class="anchor" aria-label="anchor" href="#what-is-here-ls"></a>
</h2>
<hr class="half-width">
<p>Use <code>ls</code> to list directory contents:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>You should see a directory prepared for the workshop data, for
example <code>microbial_workshop</code> or <code>shell_data</code>. Your
instructor will tell you the exact name.</p>
<p>You can give <code>ls</code> a <strong>path</strong> as an
argument:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ls</span> shell_data</span></code></pre>
</div>
<p>This lists the contents of <code>shell_data</code> without changing
where you are.</p>
</section><section><h2 class="section-heading" id="moving-around-cd">Moving around: <code>cd</code>
<a class="anchor" aria-label="anchor" href="#moving-around-cd"></a>
</h2>
<hr class="half-width">
<p>Use <code>cd</code> (<em>change directory</em>) to move between
directories.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> shell_data</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>Now <code>pwd</code> should show that you are inside
<code>shell_data</code>.</p>
<p>To go <strong>back up one level</strong>, use:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre>
</div>
<p>To go directly to your home directory from anywhere:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span></span></code></pre>
</div>
<p>or</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> ~</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="absolute-vs-relative-paths">Absolute vs relative paths<a class="anchor" aria-label="anchor" href="#absolute-vs-relative-paths"></a>
</h2>
<hr class="half-width">
<p>There are two ways to describe a location:</p>
<ul>
<li>
<strong>Absolute path</strong> – starts from the root <code>/</code>
and works downward.
<ul>
<li>Example: <code>/home/learner01/shell_data/fastq</code>
</li>
</ul>
</li>
<li>
<strong>Relative path</strong> – starts from your current directory.
<ul>
<li>Example: if you are already in <code>/home/learner01</code>, then
<code>shell_data/fastq</code> is a relative path to the same place.</li>
</ul>
</li>
</ul>
<p>Try these commands and compare:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="bu">cd</span>               <span class="co"># go to your home directory</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="bu">cd</span> shell_data    <span class="co"># relative path</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="bu">cd</span>               <span class="co"># back home</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="bu">cd</span> /home/learner01/shell_data   <span class="co"># absolute path (adjust for your username)</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>Both commands put you in the same directory.</p>
</section><section><h2 class="section-heading" id="helpful-shortcuts">Helpful shortcuts<a class="anchor" aria-label="anchor" href="#helpful-shortcuts"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<code>.</code> – the current directory</li>
<li>
<code>..</code> – the parent directory</li>
<li>
<code>~</code> – your home directory</li>
</ul>
<p>These shortcuts work in paths. For example:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> ~/shell_data</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">ls</span> ../</span></code></pre>
</div>
<p>The second command lists the contents of the directory <em>above</em>
<code>shell_data</code>.</p>
<div id="exercise-navigating-the-workshop-data" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-navigating-the-workshop-data" class="callout-inner">
<h3 class="callout-title">Exercise: Navigating the workshop data</h3>
<div class="callout-content">
<p>Your instructor will tell you the name of the main data directory for
this workshop (for example, <code>microbial_workshop</code> or
<code>shell_data</code>). Starting from your home directory:</p>
<ol style="list-style-type: decimal">
<li>Use <code>cd</code> and <code>ls</code> to find the directory that
contains the raw FASTQ files.</li>
<li>Use <code>pwd</code> to record the <strong>absolute path</strong> to
that directory.</li>
<li>From the FASTQ directory, use a <strong>relative path</strong> with
<code>cd</code> to return to your home directory.</li>
</ol>
<p>Compare answers with your neighbor. Did you find the same paths?</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The filesystem is organized as a <strong>tree</strong> of
directories and files.</li>
<li>
<code>pwd</code> tells you where you are; <code>ls</code> shows what
is in a directory; <code>cd</code> changes your location.</li>
<li>
<strong>Absolute paths</strong> start from <code>/</code>;
<strong>relative paths</strong> start from your current directory.</li>
<li>Shortcuts like <code>.</code>, <code>..</code>, and <code>~</code>
make it easier to specify paths.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-working-with-files"><p>Content from <a href="04-working-with-files.html">Working with Files and Scripts</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/04-working-with-files.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I inspect the contents of text files from the command
line?</li>
<li>How do I create, copy, move, and remove files and directories?</li>
<li>How can I automate a few commands in a simple shell script?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use <code>less</code>, <code>head</code>, and <code>tail</code> to
view text files.</li>
<li>Create, copy, move, and remove files and directories with
<code>mkdir</code>, <code>cp</code>, <code>mv</code>, and
<code>rm</code>.</li>
<li>Write and run a small shell script using <code>bash</code>.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Working with
Files and Directories</em> and <em>Writing Scripts and Working with
Data</em> episodes in <em>Introduction to the Command Line for
Genomics</em>.</p>
</blockquote>
<section><h2 class="section-heading" id="looking-at-files-less-head-and-tail">Looking at files: <code>less</code>, <code>head</code>, and
<code>tail</code>
<a class="anchor" aria-label="anchor" href="#looking-at-files-less-head-and-tail"></a>
</h2>
<hr class="half-width">
<p>Genomics data and results are usually stored in <strong>plain text
files</strong>. We can inspect them from the command line without
opening a separate editor.</p>
<p>Navigate to the directory that contains a small text file (for
example, a README or an example FASTQ). Your instructor will give a
specific path.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> ~/shell_data</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>Use <code>head</code> to see the first few lines of a file:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">head</span> README.txt</span></code></pre>
</div>
<p>Use <code>tail</code> to see the last few lines:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">tail</span> README.txt</span></code></pre>
</div>
<p>For longer files, <code>less</code> lets you scroll:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">less</span> README.txt</span></code></pre>
</div>
<ul>
<li>Use the arrow keys or <kbd>Space</kbd> to move.</li>
<li>Press <kbd>q</kbd> to quit <code>less</code>.</li>
</ul></section><section><h2 class="section-heading" id="creating-and-organizing-directories">Creating and organizing directories<a class="anchor" aria-label="anchor" href="#creating-and-organizing-directories"></a>
</h2>
<hr class="half-width">
<p>We often want to separate <strong>raw data</strong>,
<strong>intermediate files</strong>, and <strong>results</strong>.</p>
<p>From your home directory, create a project folder structure:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">mkdir</span> microbial_project</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="bu">cd</span> microbial_project</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">mkdir</span> data raw qc scripts results</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>Now you have a tidy place to keep files you will create in later
episodes.</p>
</section><section><h2 class="section-heading" id="copying-moving-and-removing-files">Copying, moving, and removing files<a class="anchor" aria-label="anchor" href="#copying-moving-and-removing-files"></a>
</h2>
<hr class="half-width">
<p>The main commands are:</p>
<ul>
<li>
<code>cp</code> – copy files</li>
<li>
<code>mv</code> – move or rename files</li>
<li>
<code>rm</code> – remove (delete) files</li>
<li>
<code>mkdir</code> – create directories</li>
<li>
<code>rmdir</code> – remove empty directories</li>
</ul>
<p>Examples (run from inside <code>microbial_project</code>):</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">cp</span> ~/shell_data/README.txt data/</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">ls</span> data</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">mv</span> data/README.txt data/project_readme.txt</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">ls</span> data</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="fu">rm</span> data/project_readme.txt</span></code></pre>
</div>
<p>Be careful with <code>rm</code> – deleted files are
<strong>not</strong> sent to a trash can. They are gone unless you have
a backup.</p>
</section><section><h2 class="section-heading" id="a-first-shell-script">A first shell script<a class="anchor" aria-label="anchor" href="#a-first-shell-script"></a>
</h2>
<hr class="half-width">
<p>Typing the same sequence of commands over and over is error-prone. A
<strong>shell script</strong> is just a text file containing commands
that you want to run together.</p>
<p>In the <code>scripts</code> directory, create a script called
<code>hello.sh</code>:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project/scripts</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">nano</span> hello.sh</span></code></pre>
</div>
<p>In the editor, type:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Hello from the microbial genomics workshop!"</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<ul>
<li>The first line is called a <strong>shebang</strong> and tells the
system to run the script with <code>bash</code>.</li>
<li>The remaining lines are the same commands you would type at the
prompt.</li>
</ul>
<p>Save the file and exit the editor (your instructor will demonstrate
the key strokes if you are using <code>nano</code>).</p>
<p>Run the script:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">bash</span> hello.sh</span></code></pre>
</div>
<p>You should see the message, then the working directory, then a
listing of the files in <code>scripts</code>.</p>
<p>Optionally, you can mark the script as
<strong>executable</strong>:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">chmod</span> +x hello.sh</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">./hello.sh</span></span></code></pre>
</div>
<p>Now you can run it directly with <code>./hello.sh</code>.</p>
<div id="exercise-making-a-project-script" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-making-a-project-script" class="callout-inner">
<h3 class="callout-title">Exercise: Making a project script</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>From inside <code>~/microbial_project/scripts</code>, create a
script called <code>setup_project.sh</code>.</li>
<li>The script should:
<ul>
<li>print the current working directory,</li>
<li>create directories called <code>tmp</code> and <code>logs</code> (if
they do not already exist),</li>
<li>list the contents of the project directory.</li>
</ul>
</li>
<li>Run your script from the project root
(<code>~/microbial_project</code>).</li>
</ol>
<blockquote>
<p>Hint: use <code>mkdir -p</code> to avoid errors if the directory
already exists.</p>
</blockquote>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Use <code>head</code>, <code>tail</code>, and <code>less</code> to
inspect text files on the command line.</li>
<li>Use <code>mkdir</code>, <code>cp</code>, <code>mv</code>, and
<code>rm</code> to manage files and directories.</li>
<li>A <strong>shell script</strong> is a text file containing commands
that can be run together with <code>bash scriptname.sh</code>.</li>
<li>Scripts help automate repetitive tasks and make your work more
reproducible.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-tidiness"><p>Content from <a href="05-tidiness.html">Project Tidiness</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/05-tidiness.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<section><h2 class="section-heading" id="suggested-layout">Suggested layout<a class="anchor" aria-label="anchor" href="#suggested-layout"></a>
</h2>
<hr class="half-width">
<pre><code>project/
├── data/raw/
├── data/trimmed/
├── assembly/
└── annotation/</code></pre>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Keep raw data immutable.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-06-project-planning"><p>Content from <a href="06-project-planning.html">Planning a Genomics Project</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/06-project-planning.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>Know: - organism - expected genome size - sequencing approach -
analysis goals</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Genome size expectations help interpret the assembly.</li>
</ul>
</div>
</div>
</div></section><section id="aio-07-background"><p>Content from <a href="07-background.html">Sequencing Data Background</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/07-background.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the difference between FASTA and FASTQ files?</li>
<li>What information is stored in a sequencing read?</li>
<li>What are paired-end reads, and why do we use them?</li>
<li>Where will our sequencing files live on the rampers server?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Distinguish between FASTA and FASTQ formats.</li>
<li>Describe the four lines that make up a single FASTQ record.</li>
<li>Explain what paired-end reads are and how they are named.</li>
<li>Locate the raw FASTQ files for this workshop in your project
directory.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <strong>Wrangling
Genomics</strong> background lesson and a short sequence file formats
overview, condensed for microbial assembly.</p>
</blockquote>
<section><h2 class="section-heading" id="sequences-reads-and-genomes">Sequences, reads, and genomes<a class="anchor" aria-label="anchor" href="#sequences-reads-and-genomes"></a>
</h2>
<hr class="half-width">
<ul>
<li>A <strong>genome</strong> is the complete DNA sequence of an
organism.</li>
<li>A <strong>read</strong> is a short piece of DNA sequence produced by
a sequencer.</li>
<li>To reconstruct a genome, we need many overlapping reads.</li>
</ul>
<p>In this workshop, we work with <strong>Illumina paired-end
reads</strong> from a single microbial genome. The goal is to assemble
these reads into <strong>contigs</strong> and then annotate genes.</p>
</section><section><h2 class="section-heading" id="fasta-vs-fastq">FASTA vs FASTQ<a class="anchor" aria-label="anchor" href="#fasta-vs-fastq"></a>
</h2>
<hr class="half-width">
<p>Both FASTA and FASTQ are text formats used to store sequences.</p>
<ul>
<li>
<strong>FASTA</strong> stores only the
<strong>sequence</strong>.</li>
<li>
<strong>FASTQ</strong> stores the <strong>sequence plus a quality
score</strong> for each base.</li>
</ul>
<p>A FASTA record looks like:</p>
<pre class="text"><code>&gt;sequence_id
ACGTACGTACGT...</code></pre>
<p>A FASTQ record always has <strong>four lines per read</strong>:</p>
<ol style="list-style-type: decimal">
<li>
<code>@</code> + sequence identifier<br>
</li>
<li>DNA sequence<br>
</li>
<li>
<code>+</code> (optionally followed by the identifier again)<br>
</li>
<li>Quality string (one character per base)</li>
</ol>
<p>For example:</p>
<pre class="text"><code>@read0001
ACGTACGTACGT
+
IIIIHHHFFFEE</code></pre>
<p>The quality string encodes how confident the sequencer is about each
base (higher characters = better quality). We will not decode the exact
scores in this workshop, but we will look at <strong>quality
profiles</strong> with FastQC.</p>
</section><section><h2 class="section-heading" id="compressed-fastq-files">Compressed FASTQ files<a class="anchor" aria-label="anchor" href="#compressed-fastq-files"></a>
</h2>
<hr class="half-width">
<p>On the rampers server, sequencing reads are usually stored in
compressed files with the <code>.fastq.gz</code> extension. These are
FASTQ files that have been compressed with <code>gzip</code> to save
disk space.</p>
<p>Example filenames:</p>
<ul>
<li>
<code>sample_R1.fastq.gz</code> – forward reads<br>
</li>
<li>
<code>sample_R2.fastq.gz</code> – reverse reads</li>
</ul>
<p>The <code>R1</code> and <code>R2</code> indicate the two
<strong>ends</strong> of the same DNA fragment.</p>
</section><section><h2 class="section-heading" id="paired-end-reads">Paired-end reads<a class="anchor" aria-label="anchor" href="#paired-end-reads"></a>
</h2>
<hr class="half-width">
<p>In <strong>paired-end sequencing</strong>, the machine reads from
both ends of each DNA fragment, producing two reads per fragment:</p>
<ul>
<li>
<code>R1</code> – read from one end</li>
<li>
<code>R2</code> – read from the other end</li>
</ul>
<p>These pairs:</p>
<ul>
<li>improve mapping and assembly,</li>
<li>help resolve small repeats,</li>
<li>give more confidence about insert size.</li>
</ul>
<p>Assemblers like SPAdes expect these reads to be provided as matching
<code>R1</code> and <code>R2</code> FASTQ files.</p>
</section><section><h2 class="section-heading" id="where-are-our-reads-on-the-rampers-server">Where are our reads on the rampers server?<a class="anchor" aria-label="anchor" href="#where-are-our-reads-on-the-rampers-server"></a>
</h2>
<hr class="half-width">
<p>For this workshop we will keep raw data in:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">~/microbial_project/data/raw</span></span></code></pre>
</div>
<p>Check that directory now:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">ls</span> data/raw</span></code></pre>
</div>
<p>You should see one or more pairs of FASTQ files
(e.g. <code>sample_R1.fastq.gz</code> and
<code>sample_R2.fastq.gz</code>). Your instructor will describe the
specific dataset.</p>
<div id="exercise-counting-reads-conceptual" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-counting-reads-conceptual" class="callout-inner">
<h3 class="callout-title">Exercise: Counting reads (conceptual)</h3>
<div class="callout-content">
<p>The total number of reads in a FASTQ file is the number of lines
divided by 4.</p>
<p>Without actually running the commands, answer:</p>
<ol style="list-style-type: decimal">
<li>Which command would show you the number of lines in a compressed
FASTQ?</li>
<li>How would you compute the approximate number of reads from
that?</li>
</ol>
<p>Your instructor may demonstrate this on the rampers server.</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>FASTA stores <strong>sequence only</strong>; FASTQ stores
<strong>sequence plus quality</strong>.</li>
<li>Each FASTQ read uses four lines: identifier, sequence, separator,
qualities.</li>
<li>Illumina paired-end data is stored as <code>R1</code> and
<code>R2</code> FASTQ files.</li>
<li>We keep raw reads in <code>~/microbial_project/data/raw</code> on
the rampers server.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-08-quality-control"><p>Content from <a href="08-quality-control.html">Assessing Read Quality (FastQC)</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/08-quality-control.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can we quickly assess the quality of our sequencing reads?</li>
<li>What kinds of problems can we detect with FastQC?</li>
<li>Where should we store QC reports in our project structure?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Run FastQC on raw FASTQ files.</li>
<li>Locate and interpret key plots in a FastQC report.</li>
<li>Identify common quality issues (low-quality tails, adapter
contamination).</li>
<li>Store QC reports in a dedicated <code>qc/</code> directory.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <strong>Quality
Control</strong> episode in <em>Wrangling Genomics</em>, using FastQC on
microbial reads.</p>
</blockquote>
<section><h2 class="section-heading" id="why-do-quality-control">Why do quality control?<a class="anchor" aria-label="anchor" href="#why-do-quality-control"></a>
</h2>
<hr class="half-width">
<p>Sequencing machines make mistakes. Problems such as:</p>
<ul>
<li>declining quality at the ends of reads,</li>
<li>leftover adapter sequences,</li>
<li>unexpected sequence content,</li>
</ul>
<p>can cause trouble for assembly and downstream analyses.</p>
<p>Before trimming or assembly, we run <strong>FastQC</strong> to get an
overview of read quality and potential issues.</p>
</section><section><h2 class="section-heading" id="running-fastqc-on-the-rampers-server">Running FastQC on the rampers server<a class="anchor" aria-label="anchor" href="#running-fastqc-on-the-rampers-server"></a>
</h2>
<hr class="half-width">
<p>We will keep all QC outputs in a <code>qc/</code> directory inside
our project.</p>
<p>From your project root:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> qc</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">fastqc</span> data/raw/<span class="pp">*</span>.fastq.gz <span class="at">-o</span> qc</span></code></pre>
</div>
<p>This command:</p>
<ul>
<li>runs FastQC on all <code>.fastq.gz</code> files in
<code>data/raw/</code>,</li>
<li>writes HTML and <code>.zip</code> reports into
<code>qc/</code>.</li>
</ul>
<p>Check the output:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span> qc</span></code></pre>
</div>
<p>You should see files like <code>sample_R1_fastqc.html</code> and
<code>sample_R2_fastqc.html</code>.</p>
</section><section><h2 class="section-heading" id="viewing-fastqc-reports">Viewing FastQC reports<a class="anchor" aria-label="anchor" href="#viewing-fastqc-reports"></a>
</h2>
<hr class="half-width">
<p>FastQC produces HTML files meant to be viewed in a web browser. There
are a few options for seeing them:</p>
<ul>
<li>Instructor demo: the instructor downloads one report and shows it on
a projector.</li>
<li>Download yourself: use <code>scp</code> or an SFTP client to copy
<code>qc/*.html</code> to your local machine and open them in a
browser.</li>
</ul>
<p>The details of downloading will depend on your system; your
instructor will show an example.</p>
</section><section><h2 class="section-heading" id="key-fastqc-modules">Key FastQC modules<a class="anchor" aria-label="anchor" href="#key-fastqc-modules"></a>
</h2>
<hr class="half-width">
<p>When viewing a FastQC report, pay particular attention to:</p>
<ul>
<li>
<strong>Per base sequence quality</strong>
<ul>
<li>Boxplots of quality scores at each base position.<br>
</li>
<li>We like to see high-quality scores across most of the read.</li>
</ul>
</li>
<li>
<strong>Per sequence quality scores</strong>
<ul>
<li>Distribution of mean quality per read.<br>
</li>
<li>A unimodal distribution with high scores is good.</li>
</ul>
</li>
<li>
<strong>Per base sequence content</strong>
<ul>
<li>Proportions of A/C/G/T at each position.<br>
</li>
<li>Weird patterns at the start of reads can reflect library prep
bias.</li>
</ul>
</li>
<li>
<strong>Adapter content / overrepresented sequences</strong>
<ul>
<li>Presence of adapter sequences or unexpected motifs.<br>
</li>
<li>If adapters are present, trimming is essential.</li>
</ul>
</li>
</ul>
<p>FastQC uses a red/orange/green flag system (FAIL/WARN/PASS). These
are guidelines, not absolute rules; context matters.</p>
</section><section><h2 class="section-heading" id="deciding-what-to-do-next">Deciding what to do next<a class="anchor" aria-label="anchor" href="#deciding-what-to-do-next"></a>
</h2>
<hr class="half-width">
<p>FastQC does <strong>not</strong> modify your data. It only reports.
Based on the report, we decide:</p>
<ul>
<li>whether to trim low-quality bases from read ends,</li>
<li>whether to remove adapter sequences,</li>
<li>whether to discard very short or very poor-quality reads.</li>
</ul>
<p>In the next episode, we will use <strong>fastp</strong> to perform
trimming.</p>
<div id="exercise-interpreting-a-fastqc-report" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-interpreting-a-fastqc-report" class="callout-inner">
<h3 class="callout-title">Exercise: Interpreting a FastQC report</h3>
<div class="callout-content">
<p>Working from the FastQC report shown by your instructor (or one you
download):</p>
<ol style="list-style-type: decimal">
<li>Is the overall read quality good, fair, or poor?</li>
<li>Does quality drop toward the end of the reads?</li>
<li>Is there evidence of adapter contamination?</li>
<li>Based on the report, what trimming steps seem necessary?</li>
</ol>
<p>Discuss your answers with a neighbor.</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>FastQC provides a quick overview of sequencing read quality.</li>
<li>We store FastQC reports in a <code>qc/</code> directory inside the
project.</li>
<li>Key plots show base quality, sequence quality distribution, and
adapter content.</li>
<li>FastQC helps us decide how aggressively to trim reads before
assembly.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-09-trimming"><p>Content from <a href="09-trimming.html">Trimming Reads (fastp)</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/09-trimming.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we trim sequencing reads before assembly?</li>
<li>How can we remove low-quality bases and adapter sequences?</li>
<li>Where do trimmed reads fit in our project directory structure?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why trimming improves assembly quality.</li>
<li>Run <code>fastp</code> on paired-end reads to produce cleaned FASTQ
files.</li>
<li>Save trimmed reads under <code>data/trimmed/</code> and new QC
reports under <code>qc/</code>.</li>
<li>Compare the number of reads before and after trimming.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <strong>Trimming and
Filtering</strong> episode in <em>Wrangling Genomics</em>, using
<code>fastp</code> instead of Trimmomatic.</p>
</blockquote>
<section><h2 class="section-heading" id="why-trim-reads">Why trim reads?<a class="anchor" aria-label="anchor" href="#why-trim-reads"></a>
</h2>
<hr class="half-width">
<p>FastQC often shows:</p>
<ul>
<li>lower quality toward the ends of reads,</li>
<li>leftover adapter sequence,</li>
<li>some reads that are too short or poor quality to be useful.</li>
</ul>
<p><strong>Trimming</strong> aims to:</p>
<ul>
<li>remove low-quality bases from read ends,</li>
<li>remove adapters,</li>
<li>discard very short reads after trimming.</li>
</ul>
<p>The result is a smaller set of <strong>higher-quality reads</strong>
that usually assemble better.</p>
</section><section><h2 class="section-heading" id="project-locations-for-trimmed-reads">Project locations for trimmed reads<a class="anchor" aria-label="anchor" href="#project-locations-for-trimmed-reads"></a>
</h2>
<hr class="half-width">
<p>We will keep trimmed reads in:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">~/microbial_project/data/trimmed</span></span></code></pre>
</div>
<p>and any additional QC reports from fastp in:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">~/microbial_project/qc</span></span></code></pre>
</div>
<p>Create the trimmed directory if needed:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> data/trimmed</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="running-fastp-on-paired-end-reads">Running fastp on paired-end reads<a class="anchor" aria-label="anchor" href="#running-fastp-on-paired-end-reads"></a>
</h2>
<hr class="half-width">
<p>Here is a typical <code>fastp</code> command for one sample:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">fastp</span>   <span class="at">-i</span> data/raw/sample_R1.fastq.gz   <span class="at">-I</span> data/raw/sample_R2.fastq.gz   <span class="at">-o</span> data/trimmed/sample_R1.trimmed.fastq.gz   <span class="at">-O</span> data/trimmed/sample_R2.trimmed.fastq.gz   <span class="at">-h</span> qc/fastp_sample.html   <span class="at">-j</span> qc/fastp_sample.json</span></code></pre>
</div>
<p>This will:</p>
<ul>
<li>read <code>R1</code> and <code>R2</code> files from
<code>data/raw/</code>,</li>
<li>write trimmed reads to <code>data/trimmed/</code>,</li>
<li>produce an HTML and JSON report summarizing the trimming.</li>
</ul>
<p>Your instructor will provide the exact sample filenames used on the
rampers server, or a script that runs <code>fastp</code> for multiple
samples.</p>
</section><section><h2 class="section-heading" id="inspecting-fastp-reports">Inspecting fastp reports<a class="anchor" aria-label="anchor" href="#inspecting-fastp-reports"></a>
</h2>
<hr class="half-width">
<p>Like FastQC, <code>fastp</code> produces an HTML report with:</p>
<ul>
<li>basic read statistics,</li>
<li>quality distributions before and after trimming,</li>
<li>adapter removal statistics,</li>
<li>length distributions.</li>
</ul>
<p>The instructor may demonstrate:</p>
<ul>
<li>how many reads were removed,</li>
<li>how much shorter reads became on average,</li>
<li>whether adapters were detected and trimmed.</li>
</ul></section><section><h2 class="section-heading" id="checking-trimmed-files-on-the-server">Checking trimmed files on the server<a class="anchor" aria-label="anchor" href="#checking-trimmed-files-on-the-server"></a>
</h2>
<hr class="half-width">
<p>We can use simple commands to compare raw and trimmed files. For
example:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># list raw and trimmed files</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">ls</span> data/raw</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">ls</span> data/trimmed</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co"># count lines in one raw vs trimmed file (conceptual)</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="fu">zcat</span> data/raw/sample_R1.fastq.gz <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="fu">zcat</span> data/trimmed/sample_R1.trimmed.fastq.gz <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre>
</div>
<p>Remember to divide the line counts by 4 to get approximate read
counts.</p>
<div id="exercise-reasonable-trimming-or-too-aggressive" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-reasonable-trimming-or-too-aggressive" class="callout-inner">
<h3 class="callout-title">Exercise: Reasonable trimming or too
aggressive?</h3>
<div class="callout-content">
<p>Using the <code>fastp</code> report demonstrated by your instructor
(or provided as an example), answer:</p>
<ol style="list-style-type: decimal">
<li>What percentage of reads were kept after trimming?</li>
<li>Are the per-base quality scores improved after trimming?</li>
<li>Did any reads become very short? Would you consider changing the
minimum length setting?</li>
</ol>
<p>Discuss with your neighbor how you would adjust trimming parameters
if this were your own project.</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Trimming removes low-quality bases and adapters, improving
downstream assembly.</li>
<li>We keep trimmed reads in <code>data/trimmed/</code> and trimming
reports in <code>qc/</code>.</li>
<li>
<code>fastp</code> can handle paired-end trimming and produce
summary reports.</li>
<li>Comparing raw and trimmed statistics helps ensure trimming is
reasonable.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-10-assembly-spades"><p>Content from <a href="10-assembly-spades.html">De Novo Assembly with SPAdes</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/10-assembly-spades.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do short reads become longer contigs in an assembly?</li>
<li>What are contigs, scaffolds, and N50?</li>
<li>How do we run SPAdes on the rampers server using trimmed reads?</li>
<li>How can we quickly inspect the assembly output?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe at a high level how de novo assembly reconstructs genomes
from reads.</li>
<li>Define contigs, scaffolds, and N50 in the context of
assemblies.</li>
<li>Run SPAdes on trimmed paired-end reads to generate an assembly.</li>
<li>Inspect the resulting <code>contigs.fasta</code> file and basic
assembly statistics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from short microbial assembly teaching
materials and the SPAdes documentation, customized for the rampers
server project layout.</p>
</blockquote>
<section><h2 class="section-heading" id="from-reads-to-contigs">From reads to contigs<a class="anchor" aria-label="anchor" href="#from-reads-to-contigs"></a>
</h2>
<hr class="half-width">
<p>Illumina reads are typically 100–300 bases long, while microbial
genomes have millions of bases. An <strong>assembler</strong>:</p>
<ul>
<li>finds overlaps between reads,</li>
<li>builds a graph of how reads connect,</li>
<li>collapses paths through that graph into longer sequences called
<strong>contigs</strong>.</li>
</ul>
<p>Key terms:</p>
<ul>
<li>
<strong>contig</strong> – a contiguous stretch of assembled
sequence.</li>
<li>
<strong>scaffold</strong> – a set of contigs ordered and oriented
with gaps (Ns) between them.</li>
<li>
<strong>N50</strong> – the contig length such that 50% of the
assembly lies in contigs of that length or longer.</li>
</ul>
<p>Larger N50 and fewer contigs usually mean a more contiguous assembly,
but context (genome size, repeats, coverage) matters.</p>
</section><section><h2 class="section-heading" id="in-person-assembly-activity-offline">In-person assembly activity (offline)<a class="anchor" aria-label="anchor" href="#in-person-assembly-activity-offline"></a>
</h2>
<hr class="half-width">
<p>Before running SPAdes, we will do a <strong>paper-based
assembly</strong>:</p>
<ul>
<li>You will receive strips of paper with short “reads” from a simple
story.</li>
<li>Your task is to find overlaps and reconstruct the original
story.</li>
<li>You will notice challenges like repeats and ambiguous joins.</li>
</ul>
<p>This activity builds intuition for what assemblies are doing behind
the scenes.</p>
</section><section><h2 class="section-heading" id="running-spades-on-the-rampers-server">Running SPAdes on the rampers server<a class="anchor" aria-label="anchor" href="#running-spades-on-the-rampers-server"></a>
</h2>
<hr class="half-width">
<p>We will run SPAdes using the trimmed reads in
<code>data/trimmed/</code> and write all output to an
<code>assembly/</code> directory.</p>
<p>From your project root:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> assembly</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">spades.py</span>   <span class="at">-1</span> data/trimmed/sample_R1.trimmed.fastq.gz   <span class="at">-2</span> data/trimmed/sample_R2.trimmed.fastq.gz   <span class="at">-o</span> assembly/spades_sample</span></code></pre>
</div>
<p>Notes:</p>
<ul>
<li>
<code>-1</code> and <code>-2</code> specify the paired-end read
files.</li>
<li>
<code>-o</code> sets the output directory.</li>
<li>SPAdes may take several minutes depending on data size and server
load.</li>
</ul>
<p>Your instructor may provide a script that runs SPAdes with the
correct sample names for the workshop dataset.</p>
</section><section><h2 class="section-heading" id="inspecting-contigs-fasta">Inspecting <code>contigs.fasta</code>
<a class="anchor" aria-label="anchor" href="#inspecting-contigs-fasta"></a>
</h2>
<hr class="half-width">
<p>When SPAdes finishes, list the output directory:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span> assembly/spades_sample</span></code></pre>
</div>
<p>You should see files including <code>contigs.fasta</code> and
<code>scaffolds.fasta</code> (we will focus on
<code>contigs.fasta</code>).</p>
<p>To count the number of contigs:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"^&gt;"</span> assembly/spades_sample/contigs.fasta</span></code></pre>
</div>
<p>To look at the first contig:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 20 assembly/spades_sample/contigs.fasta</span></code></pre>
</div>
<p>If <code>seqkit</code> is available on the rampers server, we can get
quick stats:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">seqkit</span> stats assembly/spades_sample/contigs.fasta</span></code></pre>
</div>
<p>This will report total length, N50, and other metrics. Compare the
total length to the expected genome size for the organism.</p>
<p>If <code>seqkit</code> is not available, your instructor may provide
precomputed assembly statistics for discussion.</p>
<div id="exercise-interpreting-assembly-metrics" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-interpreting-assembly-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Interpreting assembly metrics</h3>
<div class="callout-content">
<p>Using the stats shown by your instructor (or from
<code>seqkit stats</code>), answer:</p>
<ol style="list-style-type: decimal">
<li>Approximately how large is the assembly (total length)?</li>
<li>How many contigs are there?</li>
<li>What is the N50 value?</li>
<li>Does the total length seem reasonable for a microbial genome of this
type?</li>
</ol>
<p>Discuss whether you would consider this assembly “good enough” for
basic annotation.</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Assemblers like SPAdes reconstruct genomes from overlapping short
reads, producing <strong>contigs</strong>.</li>
<li>N50 summarizes assembly contiguity: larger N50 usually means a more
contiguous assembly.</li>
<li>We run SPAdes on trimmed reads and store outputs in
<code>assembly/</code>.</li>
<li>
<code>contigs.fasta</code> is the main file we use for downstream
annotation.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-11-annotation-prokka"><p>Content from <a href="11-annotation-prokka.html">Genome Annotation with Prokka</a></p>
<hr>
<p>Last updated on 2025-12-03 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/11-annotation-prokka.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What does it mean to “annotate” a genome?</li>
<li>What features does Prokka predict from an assembled microbial
genome?</li>
<li>How do we run Prokka on the contigs from SPAdes?</li>
<li>What useful information is stored in the GFF and FAA files?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the goal of genome annotation in microbial genomics.</li>
<li>Run Prokka on an assembled genome to predict genes and other
features.</li>
<li>Identify key output files from Prokka (GFF, FAA, FFN, etc.).</li>
<li>Perform simple inspections of annotation results on the rampers
server.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from short microbial annotation teaching
materials and the Prokka documentation.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-genome-annotation">What is genome annotation?<a class="anchor" aria-label="anchor" href="#what-is-genome-annotation"></a>
</h2>
<hr class="half-width">
<p>An assembly (like <code>contigs.fasta</code>) is just raw sequence.
<strong>Annotation</strong> adds biological meaning by predicting:</p>
<ul>
<li>protein-coding genes (CDS),</li>
<li>rRNA and tRNA genes,</li>
<li>product names and functional descriptions (where possible).</li>
</ul>
<p>Prokka is a rapid annotation pipeline tuned for <strong>bacterial,
archaeal, and viral</strong> genomes.</p>
</section><section><h2 class="section-heading" id="inputs-and-outputs-of-prokka">Inputs and outputs of Prokka<a class="anchor" aria-label="anchor" href="#inputs-and-outputs-of-prokka"></a>
</h2>
<hr class="half-width">
<p><strong>Input:</strong></p>
<ul>
<li>a FASTA file with assembled contigs
(e.g. <code>contigs.fasta</code>).</li>
</ul>
<p><strong>Outputs (in an output directory):</strong></p>
<ul>
<li>
<code>*.gff</code> – combined sequence and annotation in one
file.</li>
<li>
<code>*.faa</code> – amino acid sequences of predicted
proteins.</li>
<li>
<code>*.ffn</code> – nucleotide sequences of predicted genes.</li>
<li>
<code>*.tbl</code> – feature table.</li>
<li>
<code>*.txt</code> – summary of the run.</li>
</ul>
<p>These files can be used for downstream analysis (e.g. searching for
particular genes, comparing annotations, building phylogenies).</p>
</section><section><h2 class="section-heading" id="running-prokka-on-the-rampers-server">Running Prokka on the rampers server<a class="anchor" aria-label="anchor" href="#running-prokka-on-the-rampers-server"></a>
</h2>
<hr class="half-width">
<p>We will annotate the SPAdes contigs produced in the previous
episode.</p>
<p>From your project root:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> ~/microbial_project</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> annotation</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">prokka</span>   <span class="at">--outdir</span> annotation/sample   <span class="at">--prefix</span> sample   assembly/spades_sample/contigs.fasta</span></code></pre>
</div>
<p>Options used:</p>
<ul>
<li>
<code>--outdir</code> – where to put the output files.</li>
<li>
<code>--prefix</code> – prefix for output filenames.</li>
<li>the final argument is the input FASTA file.</li>
</ul>
<p>Prokka will print progress messages and may take a few minutes
depending on the size of the assembly and the server load.</p>
<p>After it finishes, list the outputs:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span> annotation/sample</span></code></pre>
</div>
<p>Look for <code>sample.gff</code>, <code>sample.faa</code>, and
<code>sample.txt</code>.</p>
</section><section><h2 class="section-heading" id="inspecting-annotation-results">Inspecting annotation results<a class="anchor" aria-label="anchor" href="#inspecting-annotation-results"></a>
</h2>
<hr class="half-width">
<p>To get a quick sense of what Prokka found:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"	CDS	"</span> annotation/sample/sample.gff</span></code></pre>
</div>
<p>This counts how many CDS features were annotated.</p>
<p>You can also inspect the summary file:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">less</span> annotation/sample/sample.txt</span></code></pre>
</div>
<p>Look for lines describing:</p>
<ul>
<li>the number of contigs and bases,</li>
<li>the number of CDS, rRNA, tRNA predictions,</li>
<li>the annotation databases used.</li>
</ul>
<p>To preview protein sequences:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span> annotation/sample/sample.faa</span></code></pre>
</div>
<p>Each entry has a header line with the locus tag and product name (if
any), followed by the amino acid sequence.</p>
<div id="exercise-exploring-prokka-output" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-exploring-prokka-output" class="callout-inner">
<h3 class="callout-title">Exercise: Exploring Prokka output</h3>
<div class="callout-content">
<p>Using <code>sample.gff</code> and <code>sample.faa</code>, try to
answer:</p>
<ol style="list-style-type: decimal">
<li>Approximately how many protein-coding genes were predicted?</li>
<li>Can you find a gene annotated as something like “DNA gyrase” or
another familiar function?</li>
<li>How might you extract all protein sequences for a particular
functional category (conceptually)?</li>
</ol>
<p>Discuss how you might use these files in a downstream analysis
(e.g. searching for specific genes, building a phylogenetic tree).</p>
</div>
</div>
</div>
<div id="key-points" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div id="key-points" class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Annotation adds biological meaning to an assembly by predicting
genes and other features.</li>
<li>Prokka is a rapid pipeline for annotating microbial genomes.</li>
<li>Key outputs include GFF (features + coordinates) and FAA (protein
sequences).</li>
<li>We run Prokka on <code>contigs.fasta</code> from SPAdes and store
results in <code>annotation/</code> for downstream analysis.</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:ekwealor@sfsu.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.3" class="external-link">sandpaper (0.17.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.8" class="external-link">varnish (1.0.8)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://jenna-tb-ekwealor.github.io/microbial-genomics/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "genomics, microbiome, assembly, bioinformatics, Unix shell, sequencing, annotation, contigs",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://jenna-tb-ekwealor.github.io/microbial-genomics/aio.html",
  "identifier": "https://jenna-tb-ekwealor.github.io/microbial-genomics/aio.html",
  "dateCreated": "2025-12-25",
  "dateModified": "2025-12-03",
  "datePublished": "2025-12-03"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

