<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Microbial Genomics Workshop: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="../favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Microbial Genomics Workshop
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics Workshop
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="Extra_lesson.html">File manipulation</a></li>
<li><a class="dropdown-item" href="etherpad_template.html">Etherpad template</a></li>
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="resources.html">Resources</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics Workshop
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-why-cloud-computing.html">1. Why Cloud Computing for Microbial Genomics</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-shell.html">2. Introducing the Shell</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-the-filesystem.html">3. The Filesystem</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-working-with-files.html">4. Working With Files</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-tidiness.html">5. Tidiness &amp; Project Organization</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-project-planning.html">6. Project Planning</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-background.html">7. Sequence File Formats</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-quality-control.html">8. Quality Control of Reads</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-trimming.html">9. Trimming Reads</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-assembly-spades.html">10. Genome Assembly with SPAdes</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-annotation-prokka.html">11. Genome Annotation with Prokka</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="Extra_lesson.html">File manipulation</a></li>
<li><a href="etherpad_template.html">Etherpad template</a></li>
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="resources.html">Resources</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-why-cloud-computing"><p>Content from <a href="01-why-cloud-computing.html">Why Cloud Computing for Microbial Genomics</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/01-why-cloud-computing.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why is cloud/HPC computing necessary for microbial genomics?</li>
<li>What limitations do laptops have for assembly and annotation?</li>
<li>How does using a remote Unix server improve reproducibility and
workflow stability?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why genomics workflows require more RAM/CPU than personal
computers provide.</li>
<li>Describe the benefits of using a remote server for QC, trimming,
assembly, and annotation.</li>
<li>Recognize how compute resources (CPUs, RAM, storage) impact SPAdes
and Prokka.</li>
<li>Interpret server resource information and understand why it matters
for microbial genomics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode integrates core concepts from the Unix tutorial, genome
assembly workflow, and annotation workflow to explain <em>why</em> we
rely on powerful remote systems.</p>
</blockquote>
<section><h2 class="section-heading" id="why-local-machines-arent-enough">Why local machines aren’t enough<a class="anchor" aria-label="anchor" href="#why-local-machines-arent-enough"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics data are large:</p>
<ul>
<li>Illumina paired-end datasets can be large, even exceeding <strong>20
GB</strong>.</li>
<li>Each FASTQ file may contain <strong>millions of reads</strong>.</li>
<li>Assembly tools like <strong>SPAdes</strong> can require more than
<strong>16 GB of RAM</strong>.</li>
<li>Annotation tools like <strong>Prokka</strong> load large databases
into memory (RAM).</li>
</ul>
<p>Most laptops cannot handle these resource requirements. Typical
problems include:</p>
<ul>
<li>Programs freezing due to insufficient memory.</li>
<li>Disk space running out during assembly.</li>
<li>Analyses taking many hours or failing entirely.</li>
</ul>
<p>A remote server solves these issues!</p>
</section><section><h2 class="section-heading" id="practical-limitations-of-remote-servers-what-to-expect">Practical limitations of remote servers (what to expect)<a class="anchor" aria-label="anchor" href="#practical-limitations-of-remote-servers-what-to-expect"></a>
</h2>
<hr class="half-width">
<p>While cloud and HPC servers provide the resources needed for
microbial genomics, they also come with a few practical constraints:</p>
<ul>
<li>
<strong>Shared resources</strong> – CPU and RAM may be used by many
users at once.</li>
<li>
<strong>Storage quotas</strong> – You may run out of disk space and
need to clean up or request more.</li>
<li>
<strong>Internet dependence</strong> – If your connection drops,
your terminal session may disconnect.</li>
<li>
<strong>Software environments</strong> – Module versions or paths
may differ from tutorials online.</li>
<li>
<strong>Login nodes vs compute nodes</strong> – Heavy jobs should
not be run on login nodes on some HPC systems.</li>
</ul>
<p>These constraints are normal in scientific computing.<br>
One goal of this workshop is learning how to work effectively within
them.</p>
<div id="vocabulary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="vocabulary" class="callout-inner">
<h3 class="callout-title">Vocabulary</h3>
<div class="callout-content">
<ul>
<li>
<strong>CPU / core</strong> — A processing unit that executes
instructions.</li>
<li>
<strong>RAM</strong> — Memory used by running programs; a key
limiter for assembly.</li>
<li>
<strong>Disk storage</strong> — Where FASTQ, FASTA, output files,
and logs live.</li>
<li>
<strong>Network filesystem</strong> — Shared storage accessible to
all users.</li>
<li>
<strong>Remote server</strong> — A powerful, multi-user machine
accessed via SSH.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="checking-your-available-resources">Checking your available resources<a class="anchor" aria-label="anchor" href="#checking-your-available-resources"></a>
</h2>
<hr class="half-width">
<p>Before running heavy tools later (FastQC, trimming, SPAdes, Prokka),
it is useful to know how to check:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">df</span> <span class="at">-h</span> .          <span class="co"># check available disk space</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">free</span> <span class="at">-h</span>          <span class="co"># check memory (RAM)</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">nproc</span>            <span class="co"># check number of CPU cores</span></span></code></pre>
</div>
<p>These commands help you understand what the server can handle and
avoid failed runs due to insufficient resources.</p>
</section><section><h2 class="section-heading" id="why-cloud-or-hpc-systems-excel-at-microbial-genomics">Why cloud or HPC systems excel at microbial genomics<a class="anchor" aria-label="anchor" href="#why-cloud-or-hpc-systems-excel-at-microbial-genomics"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="more-cpus-faster-assembly">1. More CPUs → Faster assembly<a class="anchor" aria-label="anchor" href="#more-cpus-faster-assembly"></a>
</h3>
<p>Later, we’ll use the assembly softwsre SPAdes. SPAdes supports
threading, which means it can do multiple tasks (or pieces of the task)
simultaneously:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">spades.py</span> <span class="at">--threads</span> 10</span></code></pre>
</div>
<p>More threads = faster graph construction and contig generation (i.e.,
faster assembly).</p>
</div>
<div class="section level3">
<h3 id="more-ram-successful-assembly">2. More RAM → Successful assembly<a class="anchor" aria-label="anchor" href="#more-ram-successful-assembly"></a>
</h3>
<p>If SPAdes runs out of memory, it crashes.<br>
Servers commonly provide <strong>32–256 GB RAM</strong>, or more.</p>
</div>
<div class="section level3">
<h3 id="high-throughput-storage">3. High-throughput storage<a class="anchor" aria-label="anchor" href="#high-throughput-storage"></a>
</h3>
<p>Assemblies generate many large temporary files. A server’s storage
avoids the “No space left on device” errors common on laptops.</p>
</div>
<div class="section level3">
<h3 id="reproducibility">4. Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h3>
<p>Shared environments load consistent versions:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">fastqc</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">trimmomatic</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">spades.py</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="ex">prokka</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="stability">5. Stability<a class="anchor" aria-label="anchor" href="#stability"></a>
</h3>
<p>Jobs can run for <strong>hours</strong> without interruption using
<code>screen</code> or <code>tmux</code>.</p>
</div>
</section><section><h2 class="section-heading" id="remote-server-etiquette-best-practices">Remote server etiquette &amp; best practices<a class="anchor" aria-label="anchor" href="#remote-server-etiquette-best-practices"></a>
</h2>
<hr class="half-width">
<p>When working on shared servers:</p>
<ul>
<li>Keep file organization tidy (later episodes show how).</li>
<li>Do not run heavy jobs in your home directory if a
<code>scratch/</code> or <code>work/</code> space is provided.</li>
<li>Avoid running long jobs in a normal SSH session — use
<code>screen</code> or <code>tmux</code>.</li>
<li>Delete large temporary files when no longer needed.</li>
<li>Be mindful that other learners are using the same machine.</li>
</ul>
<p>Following these practices ensures that everyone can complete the
workshop successfully.</p>
</section><section><h2 class="section-heading" id="logging-into-the-remote-system">Logging into the remote system<a class="anchor" aria-label="anchor" href="#logging-into-the-remote-system"></a>
</h2>
<hr class="half-width">
<p>Instructors will provide:</p>
<ul>
<li>server hostname<br>
</li>
<li>username<br>
</li>
<li>temporary password</li>
</ul>
<p>Connect using SSH:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>server-address<span class="op">&gt;</span></span></code></pre>
</div>
<p>On macOS/Linux, use:</p>
<ul>
<li>Terminal</li>
</ul>
<p>On Windows, use:</p>
<ul>
<li>Windows Terminal<br>
</li>
<li>PowerShell<br>
</li>
<li>PuTTY<br>
</li>
<li>MobaXterm</li>
</ul>
<p>Once connected, you will see a shell prompt:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="useful-commands-for-orientation">Useful commands for orientation<a class="anchor" aria-label="anchor" href="#useful-commands-for-orientation"></a>
</h2>
<hr class="half-width">
<p>Here are a few commands you will use repeatedly throughout the
workshop:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">whoami</span>      <span class="co"># confirm your user name on the server</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">hostname</span>    <span class="co"># confirm which machine you are on</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="bu">pwd</span>         <span class="co"># print your current directory</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span>      <span class="co"># list files with readable sizes</span></span></code></pre>
</div>
<p>We will lean heavily on these in every subsequent episode.</p>
<div id="exercise-orient-yourself-on-the-server" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-orient-yourself-on-the-server" class="callout-inner">
<h3 class="callout-title">Exercise: Orient Yourself on the Server</h3>
<div class="callout-content">
<p>Run the above commands to learn about your current session:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>These confirm you are on the <strong>remote</strong> machine.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="how-this-episode-fits-into-the-whole-workflow">How this episode fits into the whole workflow<a class="anchor" aria-label="anchor" href="#how-this-episode-fits-into-the-whole-workflow"></a>
</h2>
<hr class="half-width">
<p>In later episodes, you will:</p>
<ol style="list-style-type: decimal">
<li>Use the shell to move, inspect, and organize FASTQ files.<br>
</li>
<li>Run FastQC to diagnose read quality.</li>
<li>Trim low-quality bases and adapters.</li>
<li>Assemble the genome using SPAdes (requires high RAM).</li>
<li>Annotate genes using Prokka (requires many CPUs).</li>
<li>Interpret assembly and annotation results.</li>
</ol>
<p>Understanding <strong>why</strong> we use a remote server now will
make the computational steps in the rest of the workshop much
smoother.</p>
<div id="exercise-why-cant-we-run-spades-on-a-laptop" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-why-cant-we-run-spades-on-a-laptop" class="callout-inner">
<h3 class="callout-title">Exercise: Why can’t we run SPAdes on a
laptop?</h3>
<div class="callout-content">
<p>Discuss:</p>
<ol style="list-style-type: decimal">
<li>Why does SPAdes need so much RAM?</li>
<li>How large are typical FASTQ input files?</li>
<li>What happens if your laptop suspends or sleeps during a long
run?</li>
<li>How would cloud/HPC resources solve these problems?</li>
</ol>
</div>
</div>
</div>
<div id="exercise-estimate-your-dataset-size" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-estimate-your-dataset-size" class="callout-inner">
<h3 class="callout-title">Exercise: Estimate your dataset size</h3>
<div class="callout-content">
<p>Later in the workshop you will work with FASTQ files. Before
assembly, check their sizes:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">du</span> <span class="at">-h</span> reads/</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How large are they?</li>
<li>Would your laptop have enough free disk space?</li>
<li>Would you feel confident running SPAdes locally?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Sequencing data are large and computationally expensive to
assemble.</li>
<li>Cloud/HPC computing provides the RAM, CPUs, and disk space needed
for assembly and annotation.</li>
<li>You access the remote system through SSH and run all commands <em>on
the server</em>, not locally.</li>
<li>Tools like FastQC, Trimmomatic, SPAdes, and Prokka all benefit from
shared, stable compute environments.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-introduction-to-shell"><p>Content from <a href="02-introduction-to-shell.html">Introducing the Shell</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/02-introduction-to-shell.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the Unix shell?</li>
<li>How do I run commands on the workshop server?</li>
<li>How do I navigate files and folders using basic Unix commands?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what the shell is and why it is essential for microbial
genomics workflows.</li>
<li>Use <code>pwd</code>, <code>ls</code>, and <code>cd</code> to
navigate the directory structure.</li>
<li>Distinguish between absolute and relative paths.</li>
<li>Understand how commands, arguments, and options fit together.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Introduction to
the Command Line for Genomics</em> lesson and enhanced with concepts
needed for microbial genome QC, trimming, assembly, and annotation
workflows.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-the-shell">What is the shell?<a class="anchor" aria-label="anchor" href="#what-is-the-shell"></a>
</h2>
<hr class="half-width">
<p>The <strong>shell</strong> is a program that interprets your commands
and runs other programs.<br>
Rather than clicking through folders, you <strong>type commands</strong>
that give you precise control over:</p>
<ul>
<li>navigating files,</li>
<li>running tools (FastQC, SPAdes, Prokka),</li>
<li>organizing sequencing data,</li>
<li>automating workflows.</li>
</ul>
<p>Because genome-scale tools run on <strong>remote Linux
servers</strong>, learning the shell is essential.</p>
<p>When you log in via SSH, you land at a prompt that looks something
like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
<p>Everything typed after this prompt is interpreted <em>by the remote
machine</em>.</p>
</section><section><h2 class="section-heading" id="confirming-your-location">Confirming your location<a class="anchor" aria-label="anchor" href="#confirming-your-location"></a>
</h2>
<hr class="half-width">
<p>You can check where you are using:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>This prints the <strong>present working directory</strong>, which
will usually be your home directory when you first log in.</p>
</section><section><h2 class="section-heading" id="listing-files-with-ls">Listing files with <code>ls</code>
<a class="anchor" aria-label="anchor" href="#listing-files-with-ls"></a>
</h2>
<hr class="half-width">
<p>Use <code>ls</code> to see what’s in your current directory:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>Common useful variations:</p>
<ul>
<li>
<code>ls -l</code> – long listing with details<br>
</li>
<li>
<code>ls -a</code> – show hidden files (those beginning with
<code>.</code>)<br>
</li>
<li>
<code>ls -lh</code> – human-readable sizes</li>
</ul></section><section><h2 class="section-heading" id="moving-around-with-cd">Moving around with <code>cd</code>
<a class="anchor" aria-label="anchor" href="#moving-around-with-cd"></a>
</h2>
<hr class="half-width">
<p>You will move between directories throughout the workshop:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">cd</span> directoryname</span></code></pre>
</div>
<p>To go <strong>home</strong>:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> ~</span></code></pre>
</div>
<p>To go <strong>up one level</strong>:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre>
</div>
<p>To move multiple levels:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> ../../</span></code></pre>
</div>
<p>To go to a specific path:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> /home/learner01/myproject/</span></code></pre>
</div>
<div id="absolute-vs-relative-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="absolute-vs-relative-paths" class="callout-inner">
<h3 class="callout-title">Absolute vs relative paths</h3>
<div class="callout-content">
<ul>
<li><p><strong>Absolute path:</strong> starts from the root
(<code>/</code>), e.g.<br><code>/home/learner01/myproject/raw_reads/</code></p></li>
<li><p><strong>Relative path:</strong> starts from your <em>current</em>
directory, e.g.<br><code>raw_reads/ERR435025_R1.fastq.gz</code></p></li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="command-structure">Command structure<a class="anchor" aria-label="anchor" href="#command-structure"></a>
</h2>
<hr class="half-width">
<p>Commands often look like this:</p>
<pre><code>command -option argument</code></pre>
<p>Example:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> /home/learner01/</span></code></pre>
</div>
<div id="common-beginner-mistakes" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="common-beginner-mistakes" class="callout-inner">
<h3 class="callout-title">Common beginner mistakes</h3>
<div class="callout-content">
<table class="table">
<colgroup>
<col width="26%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th>Mistake</th>
<th>What to check / avoid</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Running commands in the wrong directory</td>
<td>Always run <code>pwd</code> to confirm location</td>
</tr>
<tr class="even">
<td>Using relative paths carelessly</td>
<td>Prefer absolute paths for important commands</td>
</tr>
<tr class="odd">
<td>Overwriting or deleting files by accident</td>
<td>Double-check filenames; avoid spaces in names</td>
</tr>
<tr class="even">
<td>Confusing local vs remote context</td>
<td>After SSH login, all commands run on the <strong>remote
server</strong>
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-where-are-you" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-where-are-you" class="callout-inner">
<h3 class="callout-title">Exercise: Where are you?</h3>
<div class="callout-content">
<p>Run:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-navigate-with-cd" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-navigate-with-cd" class="callout-inner">
<h3 class="callout-title">Exercise: Navigate with <code>cd</code>
</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">mkdir</span> practice</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="bu">cd</span> practice</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-inspect-files-with-ls" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-files-with-ls" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect files with
<code>ls</code>
</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-a</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-practice-navigation-and-paths" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-practice-navigation-and-paths" class="callout-inner">
<h3 class="callout-title">Exercise: Practice navigation and paths</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Use <code>pwd</code> to confirm your current directory.<br>
</li>
<li>Create a directory named <code>practice1</code>.<br>
</li>
<li>Change into it.<br>
</li>
<li>Create and move into a subdirectory:</li>
</ol>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">mkdir</span> subdir</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="bu">cd</span> subdir</span></code></pre>
</div>
<ol start="5" style="list-style-type: decimal">
<li>Navigate back to your home directory in one command.</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The <strong>shell</strong> is the primary interface for microbial
genomics workflows.</li>
<li>
<code>pwd</code>, <code>ls</code>, and <code>cd</code> are essential
tools for navigation.</li>
<li>Understanding paths is critical for running tools like FastQC,
SPAdes, and Prokka.</li>
<li>All commands run on the <strong>remote server</strong>, not your
local machine.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-03-the-filesystem"><p>Content from <a href="03-the-filesystem.html">The Filesystem</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/03-the-filesystem.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the directory (folder) structure on a Linux server, and why
should you care?</li>
<li>How do you create, navigate, and manage directories in the
shell?</li>
<li>What are hidden files, and how can you view them?</li>
<li>Why does a consistent project directory structure matter for
microbial genomics?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the root filesystem and the concept of a home
directory.</li>
<li>Navigate directories using absolute and relative paths.</li>
<li>Create, remove, move, and rename files and directories.</li>
<li>View hidden files and interpret file permissions.</li>
<li>Apply good directory organization for microbial genomics
workflows.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-the-linux-filesystem">What is the Linux filesystem?<a class="anchor" aria-label="anchor" href="#what-is-the-linux-filesystem"></a>
</h2>
<hr class="half-width">
<p>On the server, all files live under the root directory
<code>/</code>.<br>
As a learner, you will mainly work in your <strong>home
directory</strong>, for example:</p>
<pre><code>/home/learner01/</code></pre>
<p>This is where all your project files — raw reads, QC results,
assemblies, and annotations — should live.</p>
</section><section><h2 class="section-heading" id="why-directory-structure-matters-for-genomics">Why directory structure matters for genomics<a class="anchor" aria-label="anchor" href="#why-directory-structure-matters-for-genomics"></a>
</h2>
<hr class="half-width">
<p>A microbial genomics project generates many files (FASTQ, QC outputs,
trimmed reads, assembly files, annotation outputs). A clear structure
prevents confusion and supports reproducibility.</p>
<p>A recommended layout:</p>
<pre><code>myproject/
├── metadata/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
</section><section><h2 class="section-heading" id="navigating-directories-with-pwd-ls-and-cd">Navigating directories with <code>pwd</code>, <code>ls</code>, and
<code>cd</code>
<a class="anchor" aria-label="anchor" href="#navigating-directories-with-pwd-ls-and-cd"></a>
</h2>
<hr class="half-width">
<p>Check where you are:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>List files:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span>      <span class="co"># long format</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-a</span>      <span class="co"># include hidden files</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span>     <span class="co"># human-readable sizes</span></span></code></pre>
</div>
<p>Move between directories:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> directoryname</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="bu">cd</span> ~        <span class="co"># home</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="bu">cd</span> ..       <span class="co"># up one level</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="bu">cd</span> ../../   <span class="co"># up two levels</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="bu">cd</span> /home/learner01/myproject/</span></code></pre>
</div>
<div id="absolute-vs-relative-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="absolute-vs-relative-paths" class="callout-inner">
<h3 class="callout-title">Absolute vs relative paths</h3>
<div class="callout-content">
<p><strong>Absolute paths</strong> start from <code>/</code>.
Example:</p>
<pre><code>/home/learner01/myproject/raw_reads/</code></pre>
<p><strong>Relative paths</strong> depend on where you currently
are:</p>
<pre><code><span><span class="va">raw_reads</span><span class="op">/</span><span class="va">sample.fastq.gz</span></span></code></pre>
<p>Understanding paths is essential later when running FastQC, SPAdes,
and Prokka.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="creating-moving-copying-and-deleting-files">Creating, moving, copying, and deleting files<a class="anchor" aria-label="anchor" href="#creating-moving-copying-and-deleting-files"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">mkdir</span> newdir</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">cp</span> file1 file2</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> dir1 dir2     <span class="co"># copy directories</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">mv</span> old new</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="fu">rm</span> file</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> directory     <span class="co"># remove directory and contents (dangerous!)</span></span></code></pre>
</div>
<div id="permissions-basics" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="permissions-basics" class="callout-inner">
<h3 class="callout-title">Permissions basics</h3>
<div class="callout-content">
<p><code>ls -l</code> shows file permissions:</p>
<ul>
<li>
<code>r</code> = read<br>
</li>
<li>
<code>w</code> = write<br>
</li>
<li>
<code>x</code> = execute</li>
</ul>
<p>Directories need <code>x</code> permission to be entered.</p>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-explore-your-home-directory" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-explore-your-home-directory" class="callout-inner">
<h3 class="callout-title">Exercise: Explore your home directory</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Run <code>pwd</code> — where are you?<br>
</li>
<li>Use <code>ls -l</code> — what files or directories do you see?<br>
</li>
<li>Use <code>ls -a</code> — what hidden files appear?</li>
</ol>
</div>
</div>
</div>
<div id="exercise-create-your-project-skeleton" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-create-your-project-skeleton" class="callout-inner">
<h3 class="callout-title">Exercise: Create your project skeleton</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">mkdir</span> myproject</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="bu">cd</span> myproject</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">mkdir</span> metadata raw_reads qc trimmed assembly annotation scripts</span></code></pre>
</div>
<p>Run <code>ls -l</code>.<br>
What permissions and sizes do you see?</p>
</div>
</div>
</div>
<div id="exercise-copy-move-and-delete-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-copy-move-and-delete-files" class="callout-inner">
<h3 class="callout-title">Exercise: Copy, move, and delete files</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"example"</span> <span class="op">&gt;</span> testfile.txt</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">cp</span> testfile.txt copy_testfile.txt</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="fu">mv</span> copy_testfile.txt renamed_testfile.txt</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">rm</span> testfile.txt</span></code></pre>
</div>
<p>Use <code>ls -l</code> to confirm results.</p>
</div>
</div>
</div>
<div id="exercise-clean-up-reflect" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-clean-up-reflect" class="callout-inner">
<h3 class="callout-title">Exercise: Clean up &amp; reflect</h3>
<div class="callout-content">
<p>Return to your home directory:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">cd</span> ~</span></code></pre>
</div>
<p>Why is it important that all genomics work stays inside a dedicated
project directory?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Linux uses a hierarchical filesystem with <code>/</code> at the
root.</li>
<li>
<code>pwd</code>, <code>ls</code>, and <code>cd</code> are essential
tools for navigating the shell.</li>
<li>Good file organization is critical for genomic workflows.</li>
<li>Absolute paths are reliable; relative paths depend on your current
location.</li>
<li>Use caution with <code>rm -r</code>.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-04-working-with-files"><p>Content from <a href="04-working-with-files.html">Working With Files</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/04-working-with-files.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 55 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I create, view, copy, move, and delete files on the
server?</li>
<li>How do redirection and appending work in the shell?</li>
<li>Why are file‑handling skills essential for a microbial genomics
workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create and view files using <code>touch</code>, <code>cat</code>,
<code>less</code>, and redirection (<code>&gt;</code> and
<code>&gt;&gt;</code>).</li>
<li>Copy, move, and rename files using <code>cp</code> and
<code>mv</code>.</li>
<li>Remove files and directories safely using <code>rm</code> and
<code>rmdir</code>.</li>
<li>Understand how these operations support QC, trimming, assembly, and
annotation workflows.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode expands foundational command-line file manipulation
skills. These skills are essential once you begin handling FASTQ files,
QC outputs, trimmed reads, SPAdes assemblies, and Prokka annotation
results.</p>
</blockquote>
<section><h2 class="section-heading" id="creating-files">Creating files<a class="anchor" aria-label="anchor" href="#creating-files"></a>
</h2>
<hr class="half-width">
<p>Use <code>touch</code> to create an empty file:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">touch</span> notes.txt</span></code></pre>
</div>
<p>Use <code>echo</code> with redirection to write text into a file:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"This is a line of text"</span> <span class="op">&gt;</span> file1.txt    <span class="co"># overwrite</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Another line"</span> <span class="op">&gt;&gt;</span> file1.txt             <span class="co"># append</span></span></code></pre>
</div>
<div id="overwrite-vs-append" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="overwrite-vs-append" class="callout-inner">
<h3 class="callout-title">Overwrite vs append</h3>
<div class="callout-content">
<ul>
<li>
<code>&gt;</code> <strong>overwrites</strong> an existing file<br>
</li>
<li>
<code>&gt;&gt;</code> <strong>appends</strong> to the end</li>
</ul>
<p>Overwriting raw sequencing data by mistake can ruin an analysis — use
redirection carefully.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="viewing-files">Viewing files<a class="anchor" aria-label="anchor" href="#viewing-files"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">cat</span> file1.txt      <span class="co"># print entire file</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">less</span> file1.txt     <span class="co"># scroll interactively (press q to quit)</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">head</span> file1.txt     <span class="co"># first 10 lines</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">tail</span> file1.txt     <span class="co"># last 10 lines</span></span></code></pre>
</div>
<p>These commands are essential later when examining FASTQ structure,
assembly summaries, and annotation outputs.</p>
</section><section><h2 class="section-heading" id="copying-and-moving-files">Copying and moving files<a class="anchor" aria-label="anchor" href="#copying-and-moving-files"></a>
</h2>
<hr class="half-width">
<p>Copy:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">cp</span> file1.txt backup_file1.txt</span></code></pre>
</div>
<p>Copy directories:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> raw_reads/ raw_reads_backup/</span></code></pre>
</div>
<p>Move or rename:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">mv</span> file1.txt renamed.txt</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">mv</span> renamed.txt raw_reads/</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="deleting-files-and-directories">Deleting files and directories<a class="anchor" aria-label="anchor" href="#deleting-files-and-directories"></a>
</h2>
<hr class="half-width">
<p>Remove a file:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">rm</span> file.txt</span></code></pre>
</div>
<p>Remove an empty directory:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">rmdir</span> mydir</span></code></pre>
</div>
<p>Remove a directory <em>and everything inside it</em>:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> results/</span></code></pre>
</div>
<div id="warning-rm--r-is-permanent" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="warning-rm--r-is-permanent" class="callout-inner">
<h3 class="callout-title">⚠️ Warning: <code>rm -r</code> is
permanent</h3>
<div class="callout-content">
<p>There is <strong>no undo</strong> in the Unix shell.<br>
Always double‑check your path before removing a directory.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="why-this-matters-for-genomics">Why this matters for genomics<a class="anchor" aria-label="anchor" href="#why-this-matters-for-genomics"></a>
</h2>
<hr class="half-width">
<p>Throughout this workshop you will:</p>
<ul>
<li>download or receive FASTQ files,<br>
</li>
<li>inspect QC reports,<br>
</li>
<li>produce trimmed reads,<br>
</li>
<li>create SPAdes assembly directories,<br>
</li>
<li>annotate genomes with Prokka.</li>
</ul>
<p>Each stage generates multiple files. Clean file-handling habits
prevent:</p>
<ul>
<li>overwriting raw data,</li>
<li>losing track of results,</li>
<li>running tools on the wrong inputs.</li>
</ul>
<p>Good file management = reproducible science.</p>
</section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-create-and-modify-a-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-create-and-modify-a-file" class="callout-inner">
<h3 class="callout-title">Exercise: Create and modify a file</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Create a file:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">touch</span> test.txt</span></code></pre>
</div>
</li>
<li>
<p>Add a line:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"first line"</span> <span class="op">&gt;</span> test.txt</span></code></pre>
</div>
</li>
<li>
<p>Append another line:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"second line"</span> <span class="op">&gt;&gt;</span> test.txt</span></code></pre>
</div>
</li>
<li><p>View it with <code>cat</code>.</p></li>
</ol>
<p><strong>Questions:</strong></p>
<ul>
<li>What changed between <code>&gt;</code> and
<code>&gt;&gt;</code>?</li>
<li>Why does this matter for logging workflows?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-copy-move-and-rename" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-copy-move-and-rename" class="callout-inner">
<h3 class="callout-title">Exercise: Copy, move, and rename</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">cp</span> test.txt copy_test.txt</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">mv</span> copy_test.txt renamed_test.txt</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="fu">mv</span> renamed_test.txt ../   <span class="co"># move one directory up</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Did the file appear where you expected?</li>
<li>How can you verify using <code>ls</code> and <code>pwd</code>?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-safe-deletions" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-safe-deletions" class="callout-inner">
<h3 class="callout-title">Exercise: Safe deletions</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Create a temporary directory:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">mkdir</span> tempdir</span></code></pre>
</div>
</li>
<li>
<p>Add a file to it:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"data"</span> <span class="op">&gt;</span> tempdir/data.txt</span></code></pre>
</div>
</li>
<li><p>Try to remove with <code>rmdir tempdir</code>. What
happens?</p></li>
<li>
<p>Remove correctly with:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> tempdir</span></code></pre>
</div>
</li>
</ol>
<p><strong>Questions:</strong><br>
- Why does <code>rmdir</code> only work on empty directories?<br>
- Why is <code>rm -r</code> dangerous?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>
<code>touch</code> creates files; <code>cat</code> and
<code>less</code> display them.</li>
<li>
<code>cp</code> copies; <code>mv</code> renames or moves
files/directories.</li>
<li>
<code>rm</code> and <code>rm -r</code> delete files and directories
— permanently.</li>
<li>Redirection (<code>&gt;</code> and <code>&gt;&gt;</code>) allows you
to create and modify files.</li>
<li>Good file-handling practices prevent mistakes during genomics
workflows.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-05-tidiness"><p>Content from <a href="05-tidiness.html">Tidiness &amp; Project Organization</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/05-tidiness.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why does tidy organization matter for microbial genomics?</li>
<li>How should we structure directories for a full analysis
workflow?</li>
<li>How do naming conventions support clarity and reproducibility?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Organize a microbial genomics project with a clear directory
structure.</li>
<li>Apply consistent file‑naming conventions.</li>
<li>Understand how tidiness supports transparent, reproducible
analysis.</li>
<li>Prepare a workspace that will be used for QC, trimming, assembly,
and annotation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="why-organization-matters-in-genomics">Why organization matters in genomics<a class="anchor" aria-label="anchor" href="#why-organization-matters-in-genomics"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics projects generate many files:<br>
raw FASTQ, QC reports, trimmed reads, assemblies, annotations, logs, and
more.<br>
Without organization, it becomes easy to:</p>
<ul>
<li>run tools on the wrong files,</li>
<li>overwrite or lose important results,</li>
<li>confuse raw vs processed data,</li>
<li>make mistakes that compromise downstream steps.</li>
</ul>
<p>But beyond convenience, tidy organization is essential for
<strong>reproducibility</strong>.</p>
<p>Reproducibility means that:</p>
<ul>
<li>
<em>You</em> can return to your project months later and understand
every step.<br>
</li>
<li>
<em>Another researcher</em> can recreate your workflow from raw
reads to final annotations.<br>
</li>
<li>Files and directories document the workflow clearly and
transparently.</li>
</ul>
<p>Good structure = good science.</p>
<div id="naming-conventions-and-reproducibility" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="naming-conventions-and-reproducibility" class="callout-inner">
<h3 class="callout-title">Naming conventions and reproducibility</h3>
<div class="callout-content">
<p>Consistent, descriptive naming makes your work understandable.</p>
<p>Good:</p>
<pre><code>sample01_R1.fastq.gz
sample01_R2.fastq.gz
trimmed/sample01_R1.trimmed.fastq.gz
assembly/sample01_assembly/
annotation/sample01_prokka/</code></pre>
<p>Avoid spaces or ambiguous names:</p>
<pre><code>my file.fastq
weirdStuff/
analysis2_final_final/</code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="recommended-project-directory-structure">Recommended project directory structure<a class="anchor" aria-label="anchor" href="#recommended-project-directory-structure"></a>
</h2>
<hr class="half-width">
<p>A clean layout helps separate raw data from processed data:</p>
<pre><code>myproject/
├── metadata/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
<p>Purpose of each:</p>
<ul>
<li>
<strong>metadata/</strong>: sample sheets, README files<br>
</li>
<li>
<strong>raw_reads/</strong>: original FASTQ (never modified)<br>
</li>
<li>
<strong>qc/</strong>: FastQC reports and summaries<br>
</li>
<li>
<strong>trimmed/</strong>: files after read trimming<br>
</li>
<li>
<strong>assembly/</strong>: SPAdes results and filtered
scaffolds<br>
</li>
<li>
<strong>annotation/</strong>: Prokka outputs<br>
</li>
<li>
<strong>scripts/</strong>: helper scripts, logs, and workflow
notes</li>
</ul>
<p>This structure mirrors the entire 8‑hour workflow and supports
transparent, reproducible analysis.</p>
</section><section><h2 class="section-heading" id="planning-for-downstream-tools">Planning for downstream tools<a class="anchor" aria-label="anchor" href="#planning-for-downstream-tools"></a>
</h2>
<hr class="half-width">
<p>Many tools assume or benefit from a predictable layout:</p>
<ul>
<li>FastQC outputs can go in <code>qc/</code>
</li>
<li>Trimming outputs go in <code>trimmed/</code>
</li>
<li>SPAdes creates an assembly directory inside
<code>assembly/</code>
</li>
<li>Prokka writes annotation files inside <code>annotation/</code>
</li>
</ul>
<p>Clear organization ensures:</p>
<ul>
<li>raw data stays untouched,</li>
<li>intermediate steps remain traceable,</li>
<li>results are easy to locate and interpret,</li>
<li>collaborators (or future you) can reproduce the workflow.</li>
</ul>
<div id="reproducibility-checklist" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="reproducibility-checklist" class="callout-inner">
<h3 class="callout-title">Reproducibility checklist</h3>
<div class="callout-content">
<ul class="task-list">
<li><label><input type="checkbox">Raw FASTQ lives
<strong>only</strong> in <code>raw_reads/</code><br></label></li>
<li><label><input type="checkbox">QC outputs stored in
<code>qc/</code><br></label></li>
<li><label><input type="checkbox">Trimmed reads stored in
<code>trimmed/</code><br></label></li>
<li><label><input type="checkbox">Assemblies saved in
<code>assembly/</code><br></label></li>
<li><label><input type="checkbox">Annotations saved in
<code>annotation/</code><br></label></li>
<li><label><input type="checkbox">Scripts, notes, and logs saved in
<code>scripts/</code><br></label></li>
<li><label><input type="checkbox">No spaces in filenames<br></label></li>
<li><label><input type="checkbox">Clear, descriptive naming for all
samples<br></label></li>
<li><label><input type="checkbox">README or metadata file explains the
workflow</label></li>
</ul>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-build-your-project-structure" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-build-your-project-structure" class="callout-inner">
<h3 class="callout-title">Exercise: Build your project structure</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> myproject/<span class="dt">{metadata</span><span class="op">,</span><span class="dt">raw_reads</span><span class="op">,</span><span class="dt">qc</span><span class="op">,</span><span class="dt">trimmed</span><span class="op">,</span><span class="dt">assembly</span><span class="op">,</span><span class="dt">annotation</span><span class="op">,</span><span class="dt">scripts}</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which directories will hold raw vs processed data?</li>
<li>Why should raw data be kept untouched?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-rename-for-tidiness" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-rename-for-tidiness" class="callout-inner">
<h3 class="callout-title">Exercise: Rename for tidiness</h3>
<div class="callout-content">
<p>Suppose you received files named:</p>
<pre><code><span><span class="va">reads1.fq.gz</span></span>
<span><span class="va">reads2.fq.gz</span></span></code></pre>
<p>Rename them:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">mv</span> reads1.fq.gz sample01_R1.fastq.gz</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">mv</span> reads2.fq.gz sample01_R2.fastq.gz</span></code></pre>
</div>
<p><strong>Why is this important?</strong></p>
</div>
</div>
</div>
<div id="exercise-add-documentation" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-add-documentation" class="callout-inner">
<h3 class="callout-title">Exercise: Add documentation</h3>
<div class="callout-content">
<p>Inside <code>metadata/</code>, create a basic README:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Project: Microbial Genome Workshop</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="st">Sample: sample01</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="st">Notes: Raw reads stored in raw_reads/, workflow steps recorded in scripts/"</span> <span class="op">&gt;</span> metadata/README.txt</span></code></pre>
</div>
<p>Why does this improve reproducibility?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Good organization prevents mistakes in multi‑step genomics
workflows.</li>
<li>Clean, consistent naming is essential for clarity and
reproducibility.</li>
<li>A standard directory layout helps trace raw → QC → trimmed →
assembly → annotation.</li>
<li>Documentation (metadata, README, logs) is part of reproducible
science.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-06-project-planning"><p>Content from <a href="06-project-planning.html">Project Planning</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/06-project-planning.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we plan a microbial genomics workflow before touching the
data?</li>
<li>What directories, metadata, and decisions must be made before QC,
trimming, assembly, and annotation?</li>
<li>How does thoughtful planning support reproducibility and reduce
errors?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Anticipate each stage of the microbial genomics workflow.</li>
<li>Identify what information (metadata) must be collected before
beginning analysis.</li>
<li>Create a project structure that supports reproducible, traceable
computation.</li>
<li>Understand how planning prevents mistakes and accelerates downstream
work.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode prepares learners for the complete workflow used in
Episodes 7–11: sequence formats → QC → trimming → assembly → annotation.
Good planning prevents downstream confusion and supports transparent,
reproducible bioinformatics.</p>
</blockquote>
<section><h2 class="section-heading" id="why-planning-matters">Why planning matters<a class="anchor" aria-label="anchor" href="#why-planning-matters"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics workflows involve many sequential steps:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Quality control</strong> (FastQC)<br>
</li>
<li>
<strong>Trimming</strong> (removing adapters and low-quality
bases)<br>
</li>
<li>
<strong>Assembly</strong> (SPAdes)<br>
</li>
<li>
<strong>Contig filtering &amp; evaluation</strong><br>
</li>
<li>
<strong>Annotation</strong> (Prokka)</li>
</ol>
<p>Each step generates multiple files, and each depends on the previous
step being correct.</p>
<p>Good planning ensures that:</p>
<ul>
<li>raw data stays untouched,</li>
<li>workflows are traceable and reproducible,</li>
<li>directories and filenames are predictable,</li>
<li>you avoid wasting compute time on mis‑labeled or misplaced
files.</li>
</ul>
<p>Reproducibility begins <strong>before</strong> you run your first
command.</p>
</section><section><h2 class="section-heading" id="planning-your-project-structure">Planning your project structure<a class="anchor" aria-label="anchor" href="#planning-your-project-structure"></a>
</h2>
<hr class="half-width">
<p>You created a tidy project layout in Episode 5.<br>
Now we expand the rationale behind it.</p>
<p>A clean directory layout enables:</p>
<ul>
<li>clear separation of raw vs processed data,</li>
<li>predictable paths for running tools,</li>
<li>easy troubleshooting when something fails,</li>
<li>transparency for collaborators or future you.</li>
</ul>
<p>A typical workflow structure:</p>
<pre><code>myproject/
├── metadata/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
</section><section><h2 class="section-heading" id="what-metadata-do-you-need">What metadata do you need?<a class="anchor" aria-label="anchor" href="#what-metadata-do-you-need"></a>
</h2>
<hr class="half-width">
<p>Before analysis begins, record:</p>
<ul>
<li>
<strong>sample IDs</strong><br>
</li>
<li>
<strong>R1 and R2 filenames</strong><br>
</li>
<li>
<strong>sequencing platform</strong><br>
</li>
<li>
<strong>expected genome size</strong><br>
</li>
<li>
<strong>notes about conditions or treatments</strong><br>
</li>
<li><strong>date of sequencing and provider</strong></li>
</ul>
<p>Place metadata in a file inside your project:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">cd</span> myproject/metadata</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="st">"sample_id	read1	read2</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="st">sample01	sample01_R1.fastq.gz	sample01_R2.fastq.gz"</span> <span class="op">&gt;</span> samples.tsv</span></code></pre>
</div>
<p>Metadata supports:</p>
<ul>
<li>reproducibility,</li>
<li>automation,</li>
<li>correct input selection,</li>
<li>downstream interpretation of assembly and annotation.</li>
</ul>
<div id="good-planning-reduces-errors" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="good-planning-reduces-errors" class="callout-inner">
<h3 class="callout-title">Good planning reduces errors</h3>
<div class="callout-content">
<p>Most failures in QC, trimming, SPAdes, and Prokka come from: - wrong
file paths, - inconsistent sample names, - missing metadata, - unclear
project structure.</p>
<p>Planning ahead prevents these issues and saves significant time.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="thinking-ahead-to-qc-trimming-and-assembly">Thinking ahead to QC, trimming, and assembly<a class="anchor" aria-label="anchor" href="#thinking-ahead-to-qc-trimming-and-assembly"></a>
</h2>
<hr class="half-width">
<p>Before beginning analysis, confirm:</p>
<ul>
<li>Are your raw FASTQ names consistent
(<code>sample01_R1.fastq.gz</code>,
<code>sample01_R2.fastq.gz</code>)?<br>
</li>
<li>Do you know where FastQC outputs will go (<code>qc/</code>)?<br>
</li>
<li>Will trimmed reads be written to (<code>trimmed/</code>)?<br>
</li>
<li>Do you have enough disk space for SPAdes output?<br>
</li>
<li>Where will Prokka results be saved (<code>annotation/</code>)?</li>
</ul>
<p>These decisions prevent workflow bottlenecks and unexpected errors
mid‑analysis.</p>
</section><section><h2 class="section-heading" id="planning-for-reproducibility">Planning for reproducibility<a class="anchor" aria-label="anchor" href="#planning-for-reproducibility"></a>
</h2>
<hr class="half-width">
<p>A reproducible project includes:</p>
<ul>
<li>a clear directory tree,</li>
<li>a metadata file,</li>
<li>saved scripts or command logs,</li>
<li>documented decisions (e.g., trimming parameters, SPAdes
settings),</li>
<li>no overwritten raw data,</li>
<li>version information for tools (FastQC, SPAdes, Prokka).</li>
</ul>
<p>This ensures others can verify, repeat, or extend your analysis.</p>
</section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-build-a-planning-checklist" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-build-a-planning-checklist" class="callout-inner">
<h3 class="callout-title">Exercise: Build a planning checklist</h3>
<div class="callout-content">
<p>Make a short checklist for your project:</p>
<ul>
<li>What directories will you need?</li>
<li>What metadata must be recorded?</li>
<li>What filenames will you use?</li>
<li>Where will QC, trimming, assembly, and annotation outputs go?</li>
</ul>
<p>Compare with a partner.</p>
</div>
</div>
</div>
<div id="exercise-create-or-update-your-metadata-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-create-or-update-your-metadata-file" class="callout-inner">
<h3 class="callout-title">Exercise: Create or update your metadata
file</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> myproject/metadata</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="st">"sample_id	read1	read2</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="st">sample01	sample01_R1.fastq.gz	sample01_R2.fastq.gz"</span> <span class="op">&gt;</span> samples.tsv</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What additional fields might help future users of this dataset?</li>
<li>How could you document SPAdes or Prokka parameters here?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-anticipate-downstream-needs" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-anticipate-downstream-needs" class="callout-inner">
<h3 class="callout-title">Exercise: Anticipate downstream needs</h3>
<div class="callout-content">
<p>Discuss:</p>
<ul>
<li>What problems might arise if trimmed reads overwrite raw reads?</li>
<li>Why do we avoid storing large SPAdes outputs inside random
directories?</li>
<li>How does planning help when teaching or collaborating?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Good planning prevents errors and supports reproducibility across
the workflow.</li>
<li>Metadata is essential for tracking sample names, files, and
decisions.</li>
<li>A consistent directory structure allows QC → trimming → assembly →
annotation to progress smoothly.</li>
<li>Reproducible science begins before data processing starts.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-07-background"><p>Content from <a href="07-background.html">Sequence File Formats</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/07-background.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are FASTA, FASTQ, and SAM/BAM file formats?</li>
<li>How are sequence, identifiers, and quality scores stored?</li>
<li>Why are many sequence files compressed, and what does
<code>.gz</code> mean?</li>
<li>How do these formats fit into the microbial genome assembly and
annotation workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Recognize and describe the structure of FASTA and FASTQ files.</li>
<li>Understand how quality scores are represented in FASTQ.</li>
<li>Explain the role of SAM/BAM files in mapping reads to
references.</li>
<li>Identify common filename extensions used in microbial genomics.</li>
<li>Connect file formats to later workflow steps: QC, trimming,
assembly, and annotation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode incorporates concepts from the <em>Sequence File
Formats</em> notes (FASTA, FASTQ, SAM/BAM) and prepares learners to
interpret the files used in later episodes for QC, assembly, and
annotation.</p>
</blockquote>
<section><h2 class="section-heading" id="overview-of-common-file-types">Overview of common file types<a class="anchor" aria-label="anchor" href="#overview-of-common-file-types"></a>
</h2>
<hr class="half-width">
<p>In microbial genomics, you will frequently encounter:</p>
<ul>
<li>
<strong>FASTA</strong> – sequence and identifier(s) only<br>
</li>
<li>
<strong>FASTQ</strong> – sequence <em>and</em> per-base quality
scores<br>
</li>
<li>
<strong>SAM/BAM</strong> – alignments of reads to a reference</li>
</ul>
<p>File extensions help you keep track:</p>
<ul>
<li>
<code>.fasta</code>, <code>.fa</code>, <code>.fna</code>,
<code>.faa</code>, <code>.ffn</code>, <code>.frn</code>
</li>
<li>
<code>.fastq</code>, <code>.fq</code>, <code>.fastq.gz</code>
</li>
<li>
<code>.sam</code>, <code>.bam</code>
</li>
</ul>
<div id="plain-text-vs-binary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="plain-text-vs-binary" class="callout-inner">
<h3 class="callout-title">Plain text vs binary</h3>
<div class="callout-content">
<ul>
<li>FASTA, FASTQ, and SAM are <strong>plain text</strong> (you can read
them with <code>less</code>).</li>
<li>BAM is a <strong>binary</strong> version of SAM (smaller and faster
to process, not human-readable).</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="fasta-format">FASTA format<a class="anchor" aria-label="anchor" href="#fasta-format"></a>
</h2>
<hr class="half-width">
<p>FASTA files store sequence(s) with headers. From your sequence
formats notes:</p>
<ul>
<li>Header lines start with <code>&gt;</code><br>
</li>
<li>The header is followed by one or more lines of sequence</li>
</ul>
<p>Example:</p>
<pre class="text"><code>&gt;gi|212499741|ref|YP_002308549.1| DNA polymerase [Bacteriophage APSE-2]
MQNLLFCDLETYSDIPINCGTHRYAENAEILLFAYAYNHAPVKVWDVTQDKTMPADLKAYFDDSEILTVW
HNGGMFDTVILKRVLNIDLPLSRVHDTLVQALAHGLPGALGLLCDIFNVNSDKAKDKEGKALISLLCKPR</code></pre>
<p>Key points:</p>
<ul>
<li>A FASTA file often contains <strong>many</strong> sequences, one
after another.</li>
<li>FASTA is used for genomes, contigs, proteins, etc.</li>
<li>Later, SPAdes assemblies and Prokka annotations will use
FASTA-formatted files.</li>
</ul></section><section><h2 class="section-heading" id="fastq-format">FASTQ format<a class="anchor" aria-label="anchor" href="#fastq-format"></a>
</h2>
<hr class="half-width">
<p>FASTQ is a <strong>high-density format for short reads</strong> (≤300
nt) that stores both sequence and quality.</p>
<p>Each record has <strong>four lines</strong>:</p>
<ol style="list-style-type: decimal">
<li>
<code>@</code> plus an identifier<br>
</li>
<li>the raw sequence<br>
</li>
<li>
<code>+</code> (optionally followed by the identifier)<br>
</li>
<li>ASCII-encoded quality string</li>
</ol>
<pre class="text"><code>@DBRHHJN1:387:H89AEADXX:2:1101:1811:1988 1:N:0:TCCTGAGCCTCTCTAT
GTATTGGATATGTCCTCTATTAAACCTTGTGATGAGGAAGCTGTCTGTCTCTTATACACATCTCCGAGCCCACGAGACTCCTGAGCA
+
BB@FFFFFHHHHHJJJJJJJJJJJJJJJJIJJJJJJJJJIJJJIJJJIIIJJJJIJIJJJJJJJJIJHIJJJHHFFDDEEEDDCDDD</code></pre>
<ul>
<li>Line 1: read name, including run, lane, tile, x/y position,
etc.</li>
<li>Line 2: sequence (A/C/G/T/N).</li>
<li>Line 4: quality scores encoded as ASCII characters.</li>
</ul>
<p>These FASTQ files are the <strong>starting point</strong> for QC,
trimming, and assembly.</p>
<div id="why-.fastq.gz" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-.fastq.gz" class="callout-inner">
<h3 class="callout-title">Why <code>.fastq.gz</code>?</h3>
<div class="callout-content">
<p>FASTQ files are large.<br>
They are often compressed with <strong>gzip</strong>, resulting in
<code>.fastq.gz</code> files.</p>
<p>Most tools (FastQC, trimmers, SPAdes) can read gzipped FASTQ
directly, so you rarely need to decompress them manually.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="sam-and-bam-formats">SAM and BAM formats<a class="anchor" aria-label="anchor" href="#sam-and-bam-formats"></a>
</h2>
<hr class="half-width">
<p>After assembling or mapping reads to a reference, alignments are
often stored in <strong>SAM</strong> (text) or <strong>BAM</strong>
(binary) format.</p>
<p>A (simplified) SAM snippet from your notes:</p>
<pre class="text"><code>@HD VN:1.3 SO:coordinate
@SQ SN:ref LN:45
r001 163 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *
r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *</code></pre>
<p>Each line after the header contains:</p>
<ul>
<li>read name</li>
<li>flags</li>
<li>reference name</li>
<li>position</li>
<li>CIGAR string (match/mismatch/insertions/deletions)</li>
<li>sequence and quality</li>
</ul>
<p>Although this workshop focuses on <strong>de novo assembly</strong>
and annotation, you may still encounter SAM/BAM when evaluating
assemblies or mapping reads back to contigs.</p>
</section><section><h2 class="section-heading" id="where-formats-fit-into-the-workflow">Where formats fit into the workflow<a class="anchor" aria-label="anchor" href="#where-formats-fit-into-the-workflow"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<strong>FASTQ</strong> – raw reads → input to QC and trimming<br>
</li>
<li>
<strong>FASTQ (trimmed)</strong> – input to SPAdes<br>
</li>
<li>
<strong>FASTA</strong> – contigs/scaffolds from assembly → input to
annotation (Prokka)<br>
</li>
<li>
<strong>FASTA (proteins, CDS)</strong> – output from Prokka<br>
</li>
<li>
<strong>SAM/BAM</strong> – optional, for read mapping to references
or assemblies</li>
</ul>
<p>Understanding these formats now will make later steps much less
mysterious.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-identify-format-by-eye" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-identify-format-by-eye" class="callout-inner">
<h3 class="callout-title">Exercise: Identify format by eye</h3>
<div class="callout-content">
<p>Your instructor will show you several short text snippets (or
files).<br>
For each one, decide whether it is:</p>
<ul>
<li>FASTA</li>
<li>FASTQ</li>
<li>SAM</li>
</ul>
<p><strong>Questions:</strong></p>
<ul>
<li>What clues tell you the format?</li>
<li>How many lines per record can you see?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-inspect-a-fastq-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-a-fastq-file" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect a FASTQ file</h3>
<div class="callout-content">
<p>On the server, navigate to where a sample FASTQ file is stored and
run:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">zcat</span> sample.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 8</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How many complete records are shown in 8 lines?</li>
<li>Which lines are the sequences?</li>
<li>Which lines are the quality scores?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-inspect-an-assembly-fasta-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-an-assembly-fasta-file" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect an assembly FASTA file</h3>
<div class="callout-content">
<p>Later, after running SPAdes, you will have a scaffolds FASTA
file.<br>
Practice on a small example (or instructor-provided file):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span> scaffolds.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What does each header look like?</li>
<li>Can you tell how many contigs are in the file
(<code>grep -c '^&gt;' scaffolds.fasta</code>)?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>FASTA stores sequence(s) with header lines starting with
<code>&gt;</code>.</li>
<li>FASTQ stores reads with <strong>four lines per record</strong>,
including quality scores.</li>
<li>Many FASTQ files are compressed as <code>.fastq.gz</code> to save
space.</li>
<li>SAM/BAM store read alignments to a reference genome or
assembly.</li>
<li>Understanding these formats is essential for QC, trimming, assembly,
and annotation.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-08-quality-control"><p>Content from <a href="08-quality-control.html">Quality Control of Reads</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/08-quality-control.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we perform quality control (QC) on raw sequencing reads?</li>
<li>What problems can QC reveal in microbial genomics datasets?</li>
<li>How do we run FastQC on the command line and interpret its key
outputs?</li>
<li>How does QC inform trimming and assembly decisions?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why QC is essential before trimming and assembly.</li>
<li>Run FastQC on compressed FASTQ files.</li>
<li>Interpret key FastQC metrics (per-base quality, GC content,
adapters, sequence duplication).</li>
<li>Connect QC results to trimming choices and assembly quality.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode prepares learners for trimming (Episode 9) and assembly
(Episode 10) by teaching how to evaluate FASTQ read quality using
FastQC.</p>
</blockquote>
<section><h2 class="section-heading" id="why-quality-control">Why quality control?<a class="anchor" aria-label="anchor" href="#why-quality-control"></a>
</h2>
<hr class="half-width">
<p>Illumina reads may contain:</p>
<ul>
<li>low-quality bases at the 3′ end,</li>
<li>leftover adapter sequences,</li>
<li>duplicated reads or biased sequence content,</li>
<li>poor-quality cycles or lanes.</li>
</ul>
<p>These issues can:</p>
<ul>
<li>reduce assembly contiguity,</li>
<li>inflate error rates,</li>
<li>produce misleading downstream analyses.</li>
</ul>
<p>QC helps you evaluate whether your sequencing run is suitable for
assembly and what trimming steps are needed.</p>
</section><section><h2 class="section-heading" id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h2>
<hr class="half-width">
<p>FastQC accepts <code>.fastq</code> and <code>.fastq.gz</code>
files.</p>
<p>We recommend placing outputs in a dedicated directory:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> qc</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc raw_reads/sample01_R1.fastq.gz raw_reads/sample01_R2.fastq.gz</span></code></pre>
</div>
<p>After running:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span> qc</span></code></pre>
</div>
<p>You should see:</p>
<ul>
<li><code>sample01_R1_fastqc.html</code></li>
<li><code>sample01_R1_fastqc.zip</code></li>
<li><code>sample01_R2_fastqc.html</code></li>
<li><code>sample01_R2_fastqc.zip</code></li>
</ul>
<div id="viewing-fastqc-reports" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="viewing-fastqc-reports" class="callout-inner">
<h3 class="callout-title">Viewing FastQC reports</h3>
<div class="callout-content">
<p>Most learners <strong>download the HTML files</strong> to their local
machine and open them in a browser.</p>
<p>Some servers also allow opening them via a remote browser
session.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="key-fastqc-modules">Key FastQC modules<a class="anchor" aria-label="anchor" href="#key-fastqc-modules"></a>
</h2>
<hr class="half-width">
<p>Below are the most important FastQC panels for microbial
genomics:</p>
<div class="section level4">
<h4 id="per-base-sequence-quality">
<strong>Per base sequence quality</strong><a class="anchor" aria-label="anchor" href="#per-base-sequence-quality"></a>
</h4>
<ul>
<li>Boxplots representing quality at each base.</li>
<li>Declining quality at the 3′ ends usually indicates the need for
trimming.</li>
</ul>
</div>
<div class="section level4">
<h4 id="per-sequence-quality-scores">
<strong>Per sequence quality scores</strong><a class="anchor" aria-label="anchor" href="#per-sequence-quality-scores"></a>
</h4>
<ul>
<li>Shows the distribution of average read quality.</li>
<li>Ideally, most reads should fall into high‑quality ranges.</li>
</ul>
</div>
<div class="section level4">
<h4 id="per-base-sequence-content">
<strong>Per base sequence content</strong><a class="anchor" aria-label="anchor" href="#per-base-sequence-content"></a>
</h4>
<ul>
<li>Early cycles often show imbalance; significant deviation across the
read is concerning.</li>
</ul>
</div>
<div class="section level4">
<h4 id="per-sequence-gc-content">
<strong>Per sequence GC content</strong><a class="anchor" aria-label="anchor" href="#per-sequence-gc-content"></a>
</h4>
<ul>
<li>Should approximate a normal distribution.</li>
<li>A multimodal distribution may indicate contamination or mixed
populations.</li>
</ul>
</div>
<div class="section level4">
<h4 id="overrepresented-sequences">
<strong>Overrepresented sequences</strong><a class="anchor" aria-label="anchor" href="#overrepresented-sequences"></a>
</h4>
<ul>
<li>Often indicates adapters, primers, or contamination.</li>
<li>Will guide which adapter sequences to remove in trimming.</li>
</ul>
</div>
<div class="section level4">
<h4 id="sequence-duplication-levels">
<strong>Sequence duplication levels</strong><a class="anchor" aria-label="anchor" href="#sequence-duplication-levels"></a>
</h4>
<ul>
<li>Can indicate library prep issues or PCR artifacts.</li>
</ul>
</div>
</section><section><h2 class="section-heading" id="connecting-qc-trimming-assembly">Connecting QC → trimming → assembly<a class="anchor" aria-label="anchor" href="#connecting-qc-trimming-assembly"></a>
</h2>
<hr class="half-width">
<p>QC findings determine trimming strategy (Episode 9):</p>
<ul>
<li>Low 3′ quality → trim ends<br>
</li>
<li>Adapter contamination → remove adapters<br>
</li>
<li>Short reads after trimming → adjust length filters</li>
</ul>
<p>This improves assembly quality in Episode 10.</p>
</section><section><h2 class="section-heading" id="optional-qc-on-trimmed-reads">Optional: QC on trimmed reads<a class="anchor" aria-label="anchor" href="#optional-qc-on-trimmed-reads"></a>
</h2>
<hr class="half-width">
<p>After trimming:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc trimmed/sample01_R1.trimmed.fastq.gz trimmed/sample01_R2.trimmed.fastq.gz</span></code></pre>
</div>
<p>Compare trimmed and untrimmed reports to assess improvements.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-run-fastqc" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-fastqc" class="callout-inner">
<h3 class="callout-title">Exercise: Run FastQC</h3>
<div class="callout-content">
<p>Run FastQC on one FASTQ file:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> qc</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc raw_reads/sample01_R1.fastq.gz</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What output files appeared in the <code>qc/</code> directory?</li>
<li>How will you view the HTML report?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-interpret-fastqc-metrics" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-interpret-fastqc-metrics" class="callout-inner">
<h3 class="callout-title">Exercise: Interpret FastQC metrics</h3>
<div class="callout-content">
<p>With an instructor-provided FastQC report:</p>
<ul>
<li>What does the <strong>Per base sequence quality</strong> panel
show?</li>
<li>Are there adapters in the <strong>Overrepresented sequences</strong>
panel?</li>
<li>Would you recommend trimming? Why?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-qc-after-trimming" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-qc-after-trimming" class="callout-inner">
<h3 class="callout-title">Exercise: QC after trimming</h3>
<div class="callout-content">
<p>Compare QC before and after trimming.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>Did quality scores improve?</li>
<li>Were adapters removed?</li>
<li>Are the reads ready for SPAdes assembly?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>QC is essential before trimming and assembly.</li>
<li>FastQC summarizes read quality, bias, GC content, and possible
contaminants.</li>
<li>QC results guide trimming decisions.</li>
<li>Improved trimmed reads lead to better assemblies.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-09-trimming"><p>Content from <a href="09-trimming.html">Trimming Reads</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/09-trimming.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we trim raw sequencing reads?</li>
<li>What kinds of issues can trimming fix?</li>
<li>How do we run a trimming tool on paired-end FASTQ files?</li>
<li>How does trimming improve downstream assembly with SPAdes?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why trimming is an important preprocessing step.</li>
<li>Describe common trimming tasks (adapter removal, quality trimming,
length filtering).</li>
<li>Run a representative trimming command on paired-end FASTQ
files.</li>
<li>Verify trimmed outputs and understand how they feed into SPAdes
assembly.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>Raw FASTQ files often contain adapters and low-quality base
calls.<br>
Trimming cleans the data so assembly (Episode 10) and annotation
(Episode 11) perform better.</p>
</blockquote>
<section><h2 class="section-heading" id="why-trim-reads">Why trim reads?<a class="anchor" aria-label="anchor" href="#why-trim-reads"></a>
</h2>
<hr class="half-width">
<p>FastQC (Episode 8) may show:</p>
<ul>
<li>low-quality tails at the 3′ end,</li>
<li>adapter contamination,</li>
<li>overrepresented sequences,</li>
<li>poor-quality cycles.</li>
</ul>
<p>If left untrimmed, these issues can lead to:</p>
<ul>
<li>fragmented assemblies,</li>
<li>misassemblies,</li>
<li>wasted compute time,</li>
<li>misleading gene predictions during annotation.</li>
</ul>
<p>Trimming improves read quality and reduces noise before assembly.</p>
</section><section><h2 class="section-heading" id="what-does-trimming-do">What does trimming do?<a class="anchor" aria-label="anchor" href="#what-does-trimming-do"></a>
</h2>
<hr class="half-width">
<p>Most trimming tools perform one or more of these steps:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Remove adapters</strong><br>
</li>
<li>
<strong>Trim low-quality bases</strong> from the ends<br>
</li>
<li>
<strong>Filter short reads</strong> below a given length<br>
</li>
<li>
<strong>Remove unpaired reads</strong> or keep them separately</li>
</ol>
<p>In this workshop, trimming is demonstrated with a representative
command pattern that applies to tools like <code>fastp</code>,
<code>cutadapt</code>, or <code>Trimmomatic</code>.</p>
</section><section><h2 class="section-heading" id="running-a-trimming-tool">Running a trimming tool<a class="anchor" aria-label="anchor" href="#running-a-trimming-tool"></a>
</h2>
<hr class="half-width">
<p>A generic trimming command looks like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">trimmer</span>   <span class="at">-i</span> raw_reads/sample01_R1.fastq.gz   <span class="at">-I</span> raw_reads/sample01_R2.fastq.gz   <span class="at">-o</span> trimmed/sample01_R1.trimmed.fastq.gz   <span class="at">-O</span> trimmed/sample01_R2.trimmed.fastq.gz   [additional options...]</span></code></pre>
</div>
<p>Replace <code>trimmer</code> with your tool of choice (e.g.,
<code>fastp</code>, <code>cutadapt</code>, etc.).</p>
<p>Key arguments:</p>
<ul>
<li>
<code>-i</code> / <code>-I</code> → input R1 and R2<br>
</li>
<li>
<code>-o</code> / <code>-O</code> → output trimmed reads<br>
</li>
<li>quality, adapter, or length options</li>
</ul>
<p>Always write trimmed reads into a <strong>separate
directory</strong>, not over raw reads.</p>
<div id="protect-your-raw-data" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="protect-your-raw-data" class="callout-inner">
<h3 class="callout-title">Protect your raw data</h3>
<div class="callout-content">
<p>Never overwrite your raw FASTQ files.<br>
Raw reads are considered archival.<br>
Always place trimmed reads in a new directory:</p>
<pre><code><span><span class="va">trimmed</span><span class="op">/</span><span class="va">sample01_R1.trimmed.fastq.gz</span></span></code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="checking-trimmed-outputs">Checking trimmed outputs<a class="anchor" aria-label="anchor" href="#checking-trimmed-outputs"></a>
</h2>
<hr class="half-width">
<p>After trimming:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span> trimmed/</span></code></pre>
</div>
<p>Confirm:</p>
<ul>
<li>both <code>R1</code> and <code>R2</code> trimmed FASTQ files
exist,</li>
<li>their sizes may be smaller than the originals,</li>
<li>record counts may change depending on trimming parameters.</li>
</ul>
<p>Optional: run FastQC again to confirm improvements (Episode 8).</p>
</section><section><h2 class="section-heading" id="how-trimming-improves-assembly">How trimming improves assembly<a class="anchor" aria-label="anchor" href="#how-trimming-improves-assembly"></a>
</h2>
<hr class="half-width">
<p>Cleaner reads → cleaner assembly graph.</p>
<p>Trimming helps:</p>
<ul>
<li>reduce ambiguous connections,</li>
<li>increase contig length,</li>
<li>reduce memory usage,</li>
<li>reduce adapter contamination in nodes,</li>
<li>improve downstream Prokka gene calls.</li>
</ul>
<p>This prepares your dataset for SPAdes (Episode 10).</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-plan-your-trimming-parameters" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-plan-your-trimming-parameters" class="callout-inner">
<h3 class="callout-title">Exercise: Plan your trimming parameters</h3>
<div class="callout-content">
<p>Based on FastQC:</p>
<ul>
<li>Do your reads have low-quality tails?</li>
<li>Are adapters present?</li>
<li>What minimum read length makes sense?</li>
</ul>
<p>Draft a trimming command with options you would use.</p>
</div>
</div>
</div>
<div id="exercise-run-a-trimming-tool" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-a-trimming-tool" class="callout-inner">
<h3 class="callout-title">Exercise: Run a trimming tool</h3>
<div class="callout-content">
<p>Your instructor will demonstrate a trimming command.<br>
Then:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ls</span> trimmed/</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Did trimmed files appear?</li>
<li>Are they smaller than raw files?</li>
<li>How many reads remain versus the original files?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-qc-after-trimming" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-qc-after-trimming" class="callout-inner">
<h3 class="callout-title">Exercise: QC after trimming</h3>
<div class="callout-content">
<p>Run FastQC again (optional):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc trimmed/sample01_R1.trimmed.fastq.gz               trimmed/sample01_R2.trimmed.fastq.gz</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Did 3′ quality improve?</li>
<li>Are adapters still detected?</li>
<li>Are the reads ready for SPAdes?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Trimming removes adapters and low-quality bases that interfere with
assembly.</li>
<li>Always keep raw reads untouched; trimmed reads belong in a separate
directory.</li>
<li>QC results guide trimming choices.</li>
<li>Good trimming improves SPAdes assembly quality and reduces noise in
annotation.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-10-assembly-spades"><p>Content from <a href="10-assembly-spades.html">Genome Assembly with SPAdes</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/10-assembly-spades.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 55 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What does <em>de novo</em> genome assembly do with our trimmed
reads?</li>
<li>How do we run SPAdes on a remote server using paired-end Illumina
data?</li>
<li>Where does SPAdes write its outputs, and which files matter for
downstream analysis?</li>
<li>How can we read basic assembly statistics (number of contigs, total
length, N50)?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain at a high level how <em>de novo</em> assembly reconstructs a
genome from reads.</li>
<li>Run SPAdes on trimmed paired-end reads in a reproducible project
structure.</li>
<li>Locate and interpret key SPAdes output files (logs,
<code>contigs.fasta</code>, <code>scaffolds.fasta</code>).</li>
<li>Extract and interpret simple assembly metrics that will inform
annotation with Prokka.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the SPAdes assembly workflow materials
and the genome assembly lecture, with a focus on running a small <em>de
novo</em> assembly on a shared server.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-de-novo-assembly">What is <em>de novo</em> assembly?<a class="anchor" aria-label="anchor" href="#what-is-de-novo-assembly"></a>
</h2>
<hr class="half-width">
<p>In this workshop we <strong>do not</strong> map reads to a reference
genome.<br>
Instead, we run a <em>de novo</em> assembly:</p>
<ul>
<li>Reads are broken into <strong>k-mers</strong> (short overlapping
words).</li>
<li>A <strong>graph</strong> is built that represents how k-mers
connect.</li>
<li>The assembler walks the graph to reconstruct
<strong>contigs</strong> (continuous sequences).</li>
<li>Contigs may then be linked into <strong>scaffolds</strong> when
paired-end information supports it.</li>
</ul>
<p>Key terms:</p>
<ul>
<li>
<strong>Contig</strong> – a continuous assembled sequence with no
gaps (Ns).</li>
<li>
<strong>Scaffold</strong> – a sequence that may contain gaps (Ns)
where contigs are linked.</li>
<li>
<strong>Coverage</strong> – average number of reads covering each
base.</li>
<li>
<strong>N50</strong> – a length such that 50% of the assembled bases
are in contigs of this length or longer.</li>
</ul>
<p>We will assemble our <strong>trimmed reads</strong> from Episode 9
into contigs using SPAdes.</p>
</section><section><h2 class="section-heading" id="inputs-and-outputs-for-spades">Inputs and outputs for SPAdes<a class="anchor" aria-label="anchor" href="#inputs-and-outputs-for-spades"></a>
</h2>
<hr class="half-width">
<p>We assume the following project structure from earlier episodes:</p>
<pre class="text"><code>myproject/
├── metadata/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
<p>For a single sample (<code>sample01</code>), you should have trimmed
reads in:</p>
<pre class="text"><code><span><span class="va">trimmed</span><span class="op">/</span><span class="va">sample01_R1.trimmed.fastq.gz</span></span>
<span><span class="va">trimmed</span><span class="op">/</span><span class="va">sample01_R2.trimmed.fastq.gz</span></span></code></pre>
<p>SPAdes will create a <strong>new directory</strong> inside
<code>assembly/</code>, for example:</p>
<pre class="text"><code>assembly/sample01_spades/</code></pre>
<p>This contains:</p>
<ul>
<li>
<code>contigs.fasta</code> – assembled contigs (main file we will
use for annotation)</li>
<li>
<code>scaffolds.fasta</code> – scaffolds; may contain Ns</li>
<li>
<code>assembly_graph.fastg</code> – graph representation of the
assembly</li>
<li>
<code>spades.log</code> – log file with parameters and progress</li>
<li>
<code>params.txt</code> – summary of command-line options used</li>
</ul>
<p>Keeping the output in <code>assembly/</code> helps maintain a tidy,
reproducible project.</p>
</section><section><h2 class="section-heading" id="running-spades">Running SPAdes<a class="anchor" aria-label="anchor" href="#running-spades"></a>
</h2>
<hr class="half-width">
<p>Change into your project directory:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">cd</span> ~/myproject</span></code></pre>
</div>
<p>Create an output directory name (SPAdes will create it):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> assembly</span></code></pre>
</div>
<p>Run SPAdes with paired-end trimmed reads:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">spades.py</span>   <span class="at">-1</span> trimmed/sample01_R1.trimmed.fastq.gz   <span class="at">-2</span> trimmed/sample01_R2.trimmed.fastq.gz   <span class="at">-o</span> assembly/sample01_spades   <span class="at">--threads</span> 4   <span class="at">--memory</span> 16</span></code></pre>
</div>
<p>Options:</p>
<ul>
<li>
<code>-1</code>, <code>-2</code> – forward and reverse reads</li>
<li>
<code>-o</code> – output directory</li>
<li>
<code>--threads</code> – number of CPU threads to use</li>
<li>
<code>--memory</code> – maximum RAM (in GB) to use</li>
</ul>
<p>Ask your instructor how many threads and how much memory are
appropriate for your server.</p>
<div id="running-spades-takes-time" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="running-spades-takes-time" class="callout-inner">
<h3 class="callout-title">Running SPAdes takes time</h3>
<div class="callout-content">
<p>Even on a small dataset, SPAdes can take <strong>several
minutes</strong>.<br>
On larger genomes, or with many samples, assemblies may take
<strong>hours</strong>.</p>
<ul>
<li>Do <strong>not</strong> run SPAdes on a login node if your system
uses a scheduler.</li>
<li>Make sure you understand where outputs are going before you
start.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="inspecting-the-spades-output">Inspecting the SPAdes output<a class="anchor" aria-label="anchor" href="#inspecting-the-spades-output"></a>
</h2>
<hr class="half-width">
<p>When SPAdes finishes, list the output directory:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">ls</span> assembly/sample01_spades</span></code></pre>
</div>
<p>You should see files including:</p>
<ul>
<li><code>contigs.fasta</code></li>
<li><code>scaffolds.fasta</code></li>
<li><code>spades.log</code></li>
<li><code>params.txt</code></li>
</ul>
<p>The main file we will use for annotation is:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">assembly/sample01_spades/contigs.fasta</span></span></code></pre>
</div>
<p>We can quickly inspect it:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">head</span> assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p>Each contig starts with a header line:</p>
<pre class="text"><code>&gt;NODE_1_length_48231_cov_38.7</code></pre>
<p>This contains:</p>
<ul>
<li>a contig ID (<code>NODE_1</code>)</li>
<li>its length in base pairs (<code>length_48231</code>)</li>
<li>estimated coverage (<code>cov_38.7</code>)</li>
</ul></section><section><h2 class="section-heading" id="simple-assembly-statistics">Simple assembly statistics<a class="anchor" aria-label="anchor" href="#simple-assembly-statistics"></a>
</h2>
<hr class="half-width">
<p>We can get basic stats from the contigs file.</p>
<p>Count contigs:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"^&gt;"</span> assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p>Estimate total assembly length (requires <code>seqkit</code>):</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">seqkit</span> stats assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p>Example output might include:</p>
<pre class="text"><code>file                              format  type  num_seqs  sum_len  min_len  avg_len  max_len
assembly/sample01_spades/contigs.fasta  FASTA  DNA      143     4.8e6    500      33706    182431</code></pre>
<p>This tells you:</p>
<ul>
<li>
<strong>num_seqs</strong> – number of contigs</li>
<li>
<strong>sum_len</strong> – total assembled length (approx genome
size)</li>
<li>
<strong>max_len</strong> – length of the longest contig</li>
</ul>
<p>N50 is also reported in some tools or logs, but even these three
numbers already give a sense of assembly quality.</p>
<div id="interpreting-assembly-quality" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="interpreting-assembly-quality" class="callout-inner">
<h3 class="callout-title">Interpreting assembly quality</h3>
<div class="callout-content">
<p>Very roughly:</p>
<ul>
<li>If total length is close to the expected genome size → good
coverage.</li>
<li>If there are <em>very many</em> short contigs → assembly may be
fragmented.</li>
<li>A few long contigs with reasonable coverage values → more contiguous
assembly.</li>
</ul>
<p>We will not perfect the assembly in this workshop, but we will
generate a reasonable draft for annotation.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="preparing-for-annotation">Preparing for annotation<a class="anchor" aria-label="anchor" href="#preparing-for-annotation"></a>
</h2>
<hr class="half-width">
<p>Prokka (Episode 11) will take the assembled contigs as input.<br>
For example:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="at">--outdir</span> annotation/sample01_prokka        <span class="at">--prefix</span> sample01        assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p>So the <strong>main output</strong> of SPAdes to remember is:</p>
<pre class="text"><code><span><span class="va">assembly</span><span class="op">/</span><span class="va">sample01_spades</span><span class="op">/</span><span class="va">contigs.fasta</span></span></code></pre>
<p>Make sure you know:</p>
<ul>
<li>where it lives,</li>
<li>how large it is,</li>
<li>how many contigs it contains.</li>
</ul>
<p>These details will be useful when interpreting annotation
results.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-identify-inputs-and-outputs-for-spades" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-identify-inputs-and-outputs-for-spades" class="callout-inner">
<h3 class="callout-title">Exercise: Identify inputs and outputs for
SPAdes</h3>
<div class="callout-content">
<p>From your project directory, run:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">ls</span> trimmed</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="fu">ls</span> assembly</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which files in <code>trimmed/</code> should be used as SPAdes
inputs?</li>
<li>What will the SPAdes output directory be called for
<code>sample01</code>?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-construct-a-spades-command" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-construct-a-spades-command" class="callout-inner">
<h3 class="callout-title">Exercise: Construct a SPAdes command</h3>
<div class="callout-content">
<p>Write down a complete SPAdes command using:</p>
<ul>
<li><code>trimmed/sample01_R1.trimmed.fastq.gz</code></li>
<li><code>trimmed/sample01_R2.trimmed.fastq.gz</code></li>
<li>output directory <code>assembly/sample01_spades</code>
</li>
<li>4 threads and 16 GB of memory</li>
</ul>
<p>Compare your command with a neighbor or with the instructor’s
example.</p>
</div>
</div>
</div>
<div id="exercise-explore-the-assembly-directory" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-explore-the-assembly-directory" class="callout-inner">
<h3 class="callout-title">Exercise: Explore the assembly directory</h3>
<div class="callout-content">
<p>After SPAdes has finished (or using a pre-computed result):</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">ls</span> assembly/sample01_spades</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which files look most important for downstream analysis?</li>
<li>Where is the <code>contigs.fasta</code> file located?</li>
<li>What other information is captured in <code>spades.log</code> or
<code>params.txt</code>?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-basic-assembly-stats" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-basic-assembly-stats" class="callout-inner">
<h3 class="callout-title">Exercise: Basic assembly stats</h3>
<div class="callout-content">
<p>Using the assembled contigs:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"^&gt;"</span> assembly/sample01_spades/contigs.fasta</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="ex">seqkit</span> stats assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How many contigs are there?</li>
<li>What is the total assembled length?</li>
<li>How does this compare to the expected genome size?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>
<em>De novo</em> assembly reconstructs a genome from overlapping
reads, producing contigs and scaffolds.</li>
<li>SPAdes uses paired-end trimmed reads as input and writes outputs to
an assembly directory.</li>
<li>
<code>contigs.fasta</code> is the key file for downstream annotation
with Prokka.</li>
<li>Simple statistics (number of contigs, total length, max length) help
assess assembly quality.</li>
<li>Keeping assembly outputs in a dedicated <code>assembly/</code>
directory supports a tidy, reproducible workflow.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-11-annotation-prokka"><p>Content from <a href="11-annotation-prokka.html">Genome Annotation with Prokka</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/11-annotation-prokka.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is genome annotation, and why do we perform it after
assembly?</li>
<li>How do we run Prokka on an assembled microbial genome?</li>
<li>What do Prokka’s output files contain?</li>
<li>How does annotation fit into the full workflow from raw reads →
assembly → biological interpretation?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the purpose of microbial genome annotation.</li>
<li>Run Prokka on an assembled genome using reproducible project
organization.</li>
<li>Identify and interpret key annotation outputs (<code>.gff</code>,
<code>.gbk</code>, <code>.faa</code>, <code>.ffn</code>,
<code>.txt</code>).</li>
<li>Understand how Prokka uses contigs to predict genes, proteins, and
RNA features.</li>
<li>Prepare annotated files for downstream exploration or comparative
genomics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>In this final episode, learners annotate the genome assembled in
Episode 10. Prokka identifies coding sequences (CDS), rRNAs, tRNAs, and
other features, producing files suitable for exploration, visualization,
or comparative genomics.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-genome-annotation">What is genome annotation?<a class="anchor" aria-label="anchor" href="#what-is-genome-annotation"></a>
</h2>
<hr class="half-width">
<p>Genome annotation assigns biological meaning to assembled
sequences.</p>
<p>Prokka identifies:</p>
<ul>
<li><strong>protein‑coding genes (CDS)</strong></li>
<li><strong>tRNAs and rRNAs</strong></li>
<li><strong>signal peptides</strong></li>
<li>
<strong>functional annotations</strong> (based on databases)</li>
</ul>
<p>Annotation answers biological questions such as:</p>
<ul>
<li>How many genes are present?</li>
<li>What pathways or functions might the organism have?</li>
<li>Are there mobile elements or phage regions?</li>
</ul></section><section><h2 class="section-heading" id="inputs-and-outputs">Inputs and outputs<a class="anchor" aria-label="anchor" href="#inputs-and-outputs"></a>
</h2>
<hr class="half-width">
<p>Prokka requires one main input:</p>
<pre><code><span><span class="va">assembly</span><span class="op">/</span><span class="va">sample01_spades</span><span class="op">/</span><span class="va">contigs.fasta</span></span></code></pre>
<p>This is the cleaned, assembled genome from SPAdes.</p>
<p>Prokka outputs many files in a dedicated directory:</p>
<ul>
<li>
<code>sample01.gff</code> — combined annotation + sequence</li>
<li>
<code>sample01.gbk</code> — GenBank format (useful for
visualization)</li>
<li>
<code>sample01.faa</code> — predicted protein sequences</li>
<li>
<code>sample01.ffn</code> — nucleotide sequences of CDS</li>
<li>
<code>sample01.fna</code> — contig sequences</li>
<li>
<code>sample01.tbl</code> — feature table</li>
<li>
<code>sample01.txt</code> — summary statistics</li>
</ul>
<p>Keeping these in <code>annotation/</code> supports reproducibility
and tidy organization.</p>
</section><section><h2 class="section-heading" id="running-prokka">Running Prokka<a class="anchor" aria-label="anchor" href="#running-prokka"></a>
</h2>
<hr class="half-width">
<p>Navigate to your project directory:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">cd</span> ~/myproject</span></code></pre>
</div>
<p>Run Prokka on your assembled contigs:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">prokka</span>   <span class="at">--outdir</span> annotation/sample01_prokka   <span class="at">--prefix</span> sample01   <span class="at">--cpus</span> 4   <span class="at">--locustag</span> S01   <span class="at">--addgenes</span>   <span class="at">--mincontiglen</span> 200   assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p>Options used:</p>
<ul>
<li>
<code>--outdir</code> — directory for results</li>
<li>
<code>--prefix</code> — prefix for output filenames</li>
<li>
<code>--cpus</code> — number of CPU threads</li>
<li>
<code>--locustag</code> — prefix for gene IDs</li>
<li>
<code>--addgenes</code> — adds gene features for each CDS</li>
<li>
<code>--mincontiglen</code> — ignore very short contigs</li>
</ul>
<div id="prokka-and-reproducibility" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="prokka-and-reproducibility" class="callout-inner">
<h3 class="callout-title">Prokka and reproducibility</h3>
<div class="callout-content">
<p>Prokka reports:</p>
<ul>
<li>exact parameters used,</li>
<li>software versions,</li>
<li>database versions,</li>
</ul>
<p>which is critical for reproducible microbial genomics workflows.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="inspecting-annotation-outputs">Inspecting annotation outputs<a class="anchor" aria-label="anchor" href="#inspecting-annotation-outputs"></a>
</h2>
<hr class="half-width">
<p>List results:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ls</span> annotation/sample01_prokka</span></code></pre>
</div>
<p>Start with the summary:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.txt</span></code></pre>
</div>
<p>Look for:</p>
<ul>
<li>total CDS</li>
<li>number of tRNAs and rRNAs</li>
<li>contig counts</li>
<li>annotation completeness</li>
</ul>
<p>Explore protein sequences:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.faa</span></code></pre>
</div>
<p>Explore the GFF file:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.gff</span></code></pre>
</div>
<p>This file contains all annotated features along with coordinates and
metadata.</p>
</section><section><h2 class="section-heading" id="connecting-annotation-to-the-workflow">Connecting annotation to the workflow<a class="anchor" aria-label="anchor" href="#connecting-annotation-to-the-workflow"></a>
</h2>
<hr class="half-width">
<p>You have now completed the pipeline:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Raw reads</strong><br>
</li>
<li>
<strong>QC (FastQC)</strong><br>
</li>
<li>
<strong>Trimming</strong><br>
</li>
<li>
<strong>Assembly (SPAdes)</strong><br>
</li>
<li><strong>Annotation (Prokka)</strong></li>
</ol>
<p>Prokka’s outputs can be used for:</p>
<ul>
<li>exploring gene functions,</li>
<li>identifying potential pathways,</li>
<li>comparing genomes,</li>
<li>visualizing features in genome browsers.</li>
</ul>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-run-prokka-or-inspect-a-pre-computed-run" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-prokka-or-inspect-a-pre-computed-run" class="callout-inner">
<h3 class="callout-title">Exercise: Run Prokka (or inspect a
pre-computed run)</h3>
<div class="callout-content">
<p>Run:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="at">--outdir</span> annotation/sample01_prokka        <span class="at">--prefix</span> sample01        assembly/sample01_spades/contigs.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How long did it take?</li>
<li>How many CDS, tRNAs, and rRNAs were identified?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-explore-annotation-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-explore-annotation-files" class="callout-inner">
<h3 class="callout-title">Exercise: Explore annotation files</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.txt</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.gff</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">less</span> annotation/sample01_prokka/sample01.faa</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What does the summary (<code>.txt</code>) reveal?</li>
<li>What kinds of features appear in the <code>.gff</code> file?</li>
<li>How are protein IDs represented in <code>.faa</code>?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-advanced-thinking-optional" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-advanced-thinking-optional" class="callout-inner">
<h3 class="callout-title">Exercise: Advanced thinking (optional)</h3>
<div class="callout-content">
<p>Discuss:</p>
<ul>
<li>Why might different annotation tools give different gene
counts?</li>
<li>How could you incorporate annotation results into comparative
genomics?</li>
<li>What additional analyses might follow annotation?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Prokka provides rapid, standardized annotation for bacterial
genomes.</li>
<li>Its outputs include GFF, GBK, FAA, FFN, and summary files.</li>
<li>The annotation step completes the pipeline from raw reads to curated
genome features.</li>
<li>Good project organization ensures annotation results remain
traceable and reproducible.</li>
</ul>
</div>
</div>
</div>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:ekwealor@sfsu.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.3" class="external-link">sandpaper (0.17.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.9" class="external-link">varnish (1.0.9)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "genomics, microbiome, assembly, bioinformatics, Unix shell, sequencing, annotation, contigs",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "identifier": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "dateCreated": "2025-12-25",
  "dateModified": "2025-12-04",
  "datePublished": "2025-12-04"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

