<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Microbial Genomics Workshop: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="../favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Microbial Genomics Workshop
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics Workshop
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="Extra_lesson.html">File manipulation</a></li>
<li><a class="dropdown-item" href="etherpad_template.html">Etherpad template</a></li>
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="resources.html">Resources</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics Workshop
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-why-cloud-computing.html">Why Cloud Computing for Microbial Genomics</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-shell.html">Introducing the Shell</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-the-filesystem.html">The Filesystem</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-working-with-files.html">Working With Files</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-tidiness.html">Tidiness &amp; Project Organization</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-project-planning.html">Project Planning</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-background.html">Sequence File Formats</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-quality-control.html">Quality Control of Reads</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-trimming.html">Trimming and Cleaning Reads</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-assembly-spades.html">Genome Assembly with SPAdes</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-annotation-prokka.html">Genome Annotation with Prokka</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <li><a href="Extra_lesson.html">File manipulation</a></li>
<li><a href="etherpad_template.html">Etherpad template</a></li>
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="resources.html">Resources</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-why-cloud-computing"><p>Content from <a href="01-why-cloud-computing.html">Why Cloud Computing for Microbial Genomics</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/01-why-cloud-computing.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why is cloud/HPC computing necessary for microbial genomics?</li>
<li>What limitations do laptops have for assembly and annotation?</li>
<li>How does using a remote Unix server improve reproducibility and
workflow stability?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why genomics workflows require more RAM/CPU than personal
computers provide.</li>
<li>Describe the benefits of using a remote server for QC, trimming,
assembly, and annotation.</li>
<li>Recognize how compute resources (CPUs, RAM, storage) impact SPAdes
and Prokka.</li>
<li>Interpret server resource information and understand why it matters
for microbial genomics.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode integrates core concepts from the Unix tutorial, genome
assembly workflow, and annotation workflow to explain <em>why</em> we
rely on powerful remote systems.</p>
</blockquote>
<section><h2 class="section-heading" id="why-local-machines-arent-enough">Why local machines aren’t enough<a class="anchor" aria-label="anchor" href="#why-local-machines-arent-enough"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics data are large:</p>
<ul>
<li>Illumina paired-end datasets can be large, exceeding even <strong>20
GB</strong>.</li>
<li>Each FASTQ file may contain <strong>millions of reads</strong>.</li>
<li>Assembly tools like <strong>SPAdes</strong> can require more than
<strong>16 GB of RAM</strong>.</li>
<li>Annotation tools like <strong>Prokka</strong> load large databases
into memory (RAM).</li>
</ul>
<p>Most laptops cannot handle these resource requirements. Typical
problems include:</p>
<ul>
<li>Programs freezing due to insufficient memory.</li>
<li>Disk space running out during assembly.</li>
<li>Analyses taking many hours or failing entirely.</li>
</ul>
<p>A remote server solves these issues!</p>
<div id="vocabulary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="vocabulary" class="callout-inner">
<h3 class="callout-title">Vocabulary</h3>
<div class="callout-content">
<ul>
<li>
<strong>CPU / core</strong> — A processing unit that executes
instructions.</li>
<li>
<strong>RAM</strong> — Memory used by running programs; a key
limiter for assembly.</li>
<li>
<strong>Disk storage</strong> — Where FASTQ, FASTA, output files,
and logs live.</li>
<li>
<strong>Network filesystem</strong> — Shared storage accessible to
all users.</li>
<li>
<strong>Remote server</strong> — A powerful, multi-user machine
accessed via SSH.</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="why-cloud-or-hpc-systems-excel-at-microbial-genomics">Why cloud or HPC systems excel at microbial genomics<a class="anchor" aria-label="anchor" href="#why-cloud-or-hpc-systems-excel-at-microbial-genomics"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="more-cpus-faster-assembly">1. More CPUs → Faster assembly<a class="anchor" aria-label="anchor" href="#more-cpus-faster-assembly"></a>
</h3>
<p>Later, we’ll use the assembly softwsre SPAdes. SPAdes supports
threading, which means it can do multiple tasks (or pieces of the task)
simultaneously:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">spades.py</span> <span class="at">--threads</span> 10</span></code></pre>
</div>
<p>More threads = faster graph construction and contig generation (i.e.,
faster assembly).</p>
</div>
<div class="section level3">
<h3 id="more-ram-successful-assembly">2. More RAM → Successful assembly<a class="anchor" aria-label="anchor" href="#more-ram-successful-assembly"></a>
</h3>
<p>If SPAdes runs out of memory, it crashes.<br>
Servers commonly provide <strong>32–256 GB RAM</strong>, or more.</p>
</div>
<div class="section level3">
<h3 id="high-throughput-storage">3. High-throughput storage<a class="anchor" aria-label="anchor" href="#high-throughput-storage"></a>
</h3>
<p>Assemblies generate many large temporary files.<br>
A server’s storage avoids the “No space left on device” errors common on
laptops.</p>
</div>
<div class="section level3">
<h3 id="reproducibility">4. Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h3>
<p>Shared environments load consistent versions:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">fastqc</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">trimmomatic</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">spades.py</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">prokka</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="stability">5. Stability<a class="anchor" aria-label="anchor" href="#stability"></a>
</h3>
<p>Jobs can run for <strong>hours</strong> without interruption using
<code>screen</code> or <code>tmux</code>.</p>
</div>
</section><section><h2 class="section-heading" id="logging-into-the-remote-system">Logging into the remote system<a class="anchor" aria-label="anchor" href="#logging-into-the-remote-system"></a>
</h2>
<hr class="half-width">
<p>Your instructor will provide:</p>
<ul>
<li>server hostname<br>
</li>
<li>username<br>
</li>
<li>temporary password</li>
</ul>
<p>Connect using SSH (macOS/Linux):</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>server-address<span class="op">&gt;</span></span></code></pre>
</div>
<p>On Windows, use:</p>
<ul>
<li>Windows Terminal<br>
</li>
<li>PowerShell<br>
</li>
<li>PuTTY<br>
</li>
<li>MobaXterm</li>
</ul>
<p>Once connected, you will see a shell prompt:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
<p>Try:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">whoami</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">hostname</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>These confirm you are on the <strong>remote</strong> machine.</p>
<div id="exercise-why-cant-we-run-spades-on-a-laptop" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-why-cant-we-run-spades-on-a-laptop" class="callout-inner">
<h3 class="callout-title">Exercise: Why can’t we run SPAdes on a
laptop?</h3>
<div class="callout-content">
<p>Discuss:</p>
<ol style="list-style-type: decimal">
<li>Why does SPAdes need so much RAM?</li>
<li>How large are typical FASTQ input files?</li>
<li>What happens if your laptop suspends or sleeps during a long
run?</li>
<li>How would cloud/HPC resources solve these problems?</li>
</ol>
</div>
</div>
</div>
<div id="exercise-estimate-your-dataset-size" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-estimate-your-dataset-size" class="callout-inner">
<h3 class="callout-title">Exercise: Estimate your dataset size</h3>
<div class="callout-content">
<p>Later in the workshop you will download real FASTQ files.<br>
Before assembly, check their sizes:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">du</span> <span class="at">-h</span> reads/</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How large are they?</li>
<li>Would your laptop have enough free disk space?</li>
<li>Would you feel confident running SPAdes locally?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Sequencing data are large and computationally expensive to
assemble.</li>
<li>Cloud/HPC computing provides the RAM, CPUs, and disk space needed
for assembly and annotation.</li>
<li>You access the remote system through SSH and run all commands <em>on
the server</em>, not locally.</li>
<li>Tools like FastQC, Trimmomatic, SPAdes, and Prokka all benefit from
shared, stable compute environments.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-introduction-to-shell"><p>Content from <a href="02-introduction-to-shell.html">Introducing the Shell</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/02-introduction-to-shell.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the Unix shell?</li>
<li>How do I run commands on the workshop server?</li>
<li>How do I navigate files and folders using basic Unix commands?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what the shell is and why it is essential for microbial
genomics workflows.</li>
<li>Run simple shell commands (<code>whoami</code>, <code>pwd</code>,
<code>ls</code>) on a remote server.</li>
<li>Navigate the directory structure using <code>cd</code>, absolute
paths, and relative paths.</li>
<li>Understand how commands, arguments, and options fit together.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is adapted from the Data Carpentry <em>Introduction to
the Command Line for Genomics</em> lesson and enhanced with concepts
needed for microbial genome QC, trimming, assembly, and annotation
workflows.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-the-shell">What is the shell?<a class="anchor" aria-label="anchor" href="#what-is-the-shell"></a>
</h2>
<hr class="half-width">
<p>The <strong>shell</strong> is a program that interprets your commands
and runs other programs.<br>
Rather than clicking through folders, you <strong>type commands</strong>
that give you precise control over:</p>
<ul>
<li>navigating files,</li>
<li>running tools (FastQC, SPAdes, Prokka),</li>
<li>organizing sequencing data,</li>
<li>automating workflows.</li>
</ul>
<p>Because genome-scale tools run on <strong>remote Linux
servers</strong>, learning the shell is essential.</p>
<p>When you log in via SSH, you land at a prompt that looks something
like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">learner01@genomics-server:~$</span></span></code></pre>
</div>
<p>Everything typed after this prompt is a command sent to the remote
machine.</p>
</section><section><h2 class="section-heading" id="your-first-commands">Your first commands<a class="anchor" aria-label="anchor" href="#your-first-commands"></a>
</h2>
<hr class="half-width">
<p>Try these:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">whoami</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">hostname</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<ul>
<li>
<code>whoami</code> shows your username (on the <strong>remote
server</strong>, not your laptop).</li>
<li>
<code>hostname</code> tells you which machine you’re logged
into.</li>
<li>
<code>pwd</code> prints your <strong>present working
directory</strong>, letting you confirm where you are.</li>
</ul>
<p>These help ensure you’re operating on the server—not your local
machine.</p>
</section><section><h2 class="section-heading" id="listing-files-with-ls">Listing files with <code>ls</code>
<a class="anchor" aria-label="anchor" href="#listing-files-with-ls"></a>
</h2>
<hr class="half-width">
<p>Use <code>ls</code> to see what’s in your current directory:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>Variants:</p>
<ul>
<li>
<code>ls -l</code> – long listing</li>
<li>
<code>ls -a</code> – show hidden files (those beginning with
<code>.</code>)</li>
<li>
<code>ls -lh</code> – human-readable file sizes</li>
</ul>
<p>This is how you inspect data directories like:</p>
<pre><code>reads/
trimmed/
assembly/
annotation/</code></pre>
<p>as the workshop progresses.</p>
</section><section><h2 class="section-heading" id="moving-around-with-cd">Moving around with <code>cd</code>
<a class="anchor" aria-label="anchor" href="#moving-around-with-cd"></a>
</h2>
<hr class="half-width">
<p>You will move between directories constantly during QC, trimming,
assembly, and annotation.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> directoryname</span></code></pre>
</div>
<p>To go <strong>home</strong>:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> ~</span></code></pre>
</div>
<p>To go <strong>up one level</strong>:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre>
</div>
<p>To move multiple levels:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> ../../</span></code></pre>
</div>
<p>To go to a specific path:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="bu">cd</span> /home/learner01/mydata/</span></code></pre>
</div>
<p>Understanding paths is critical when feeding files to tools like
FastQC or SPAdes.</p>
<div id="absolute-vs-relative-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="absolute-vs-relative-paths" class="callout-inner">
<h3 class="callout-title">Absolute vs relative paths</h3>
<div class="callout-content">
<ul>
<li><p><strong>Absolute path:</strong> starts from the root
(<code>/</code>), e.g.<br><code>/home/learner01/mydata/reads/ERR435025_R1.fastq.gz</code></p></li>
<li><p><strong>Relative path:</strong> starts from your current
directory, e.g.<br><code>reads/ERR435025_R1.fastq.gz</code></p></li>
</ul>
<p>SPAdes, FastQC, and Prokka all accept either — but incorrect paths
are a <em>very</em> common beginner error.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="command-structure">Command structure<a class="anchor" aria-label="anchor" href="#command-structure"></a>
</h2>
<hr class="half-width">
<p>Commands often look like this:</p>
<pre><code>command -option argument</code></pre>
<p>Example:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> /home/learner01/</span></code></pre>
</div>
<p>Later, when you run SPAdes, you’ll see the same structure:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">spades.py</span> <span class="at">-1</span> reads/R1.fastq.gz <span class="at">-2</span> reads/R2.fastq.gz <span class="at">--threads</span> 10 <span class="at">--memory</span> 120</span></code></pre>
</div>
<p>The shell simply interprets the command, options, and arguments you
provide.</p>
<div id="exercise-where-are-you" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-where-are-you" class="callout-inner">
<h3 class="callout-title">Exercise: Where are you?</h3>
<div class="callout-content">
<p>Run:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<ul>
<li>What directory are you in?</li>
<li>Is it your home directory?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-navigate-with-cd" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-navigate-with-cd" class="callout-inner">
<h3 class="callout-title">Exercise: Navigate with <code>cd</code>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Create a directory:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">mkdir</span> practice</span></code></pre>
</div>
</li>
<li>
<p>Move into it:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">cd</span> practice</span></code></pre>
</div>
</li>
<li>
<p>Verify location using:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
</li>
</ol>
</div>
</div>
</div>
<div id="exercise-inspect-files-with-ls" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-files-with-ls" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect files with
<code>ls</code>
</h3>
<div class="callout-content">
<p>In your home directory, run:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-a</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which files are hidden?</li>
<li>Which items are directories?</li>
<li>What do the permission strings (like <code>drwxr-xr-x</code>)
represent?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The <strong>shell</strong> is the primary interface for microbial
genomics workflows.</li>
<li>Commands like <code>whoami</code>, <code>hostname</code>, and
<code>pwd</code> confirm where you are working.</li>
<li>
<code>ls</code> lists files; <code>cd</code> moves between
directories.</li>
<li>Understanding paths is essential for running tools like FastQC,
SPAdes, and Prokka.</li>
<li>All commands run on the <strong>remote server</strong>, not your
local machine.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-the-filesystem"><p>Content from <a href="03-the-filesystem.html">The Filesystem</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/03-the-filesystem.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How is the Unix filesystem organized?</li>
<li>How do I check where I am in the directory structure?</li>
<li>How do I move between directories using paths?</li>
<li>What is the difference between absolute and relative paths?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe how the Unix filesystem is structured.</li>
<li>Use <code>pwd</code> to confirm your current location.</li>
<li>Navigate into and out of directories using <code>cd</code>.</li>
<li>Distinguish between absolute and relative paths.</li>
<li>Understand how paths matter for genomic workflows later on.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode draws directly from the Unix tutorial PDF (Exercises
2–5) to teach the filesystem concepts required before handling FASTQ
files, assemblies, or annotations.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-the-filesystem">What is the filesystem?<a class="anchor" aria-label="anchor" href="#what-is-the-filesystem"></a>
</h2>
<hr class="half-width">
<p>The Unix filesystem is a <strong>tree-like structure</strong> of
directories (folders). At the very top is the root:</p>
<pre><code>/</code></pre>
<p>Below this, you will find directories such as <code>/home</code>,
<code>/usr</code>, <code>/etc</code>, etc.</p>
<p>When you log into the workshop server, you start in <strong>your home
directory</strong>, represented by:</p>
<pre><code>~</code></pre>
<p>You can confirm this using:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>This prints your <strong>present working directory</strong> — your
current location.</p>
</section><section><h2 class="section-heading" id="listing-the-contents-of-a-directory">Listing the contents of a directory<a class="anchor" aria-label="anchor" href="#listing-the-contents-of-a-directory"></a>
</h2>
<hr class="half-width">
<p>Use <code>ls</code> to see what is inside your current directory:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>To view more details:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-la</span></span></code></pre>
</div>
<ul>
<li>
<code>-l</code> shows long format</li>
<li>
<code>-h</code> shows sizes in human-readable units</li>
<li>
<code>-a</code> includes hidden files (names beginning with
<code>.</code>)</li>
</ul></section><section><h2 class="section-heading" id="moving-between-directories-with-cd">Moving between directories with <code>cd</code>
<a class="anchor" aria-label="anchor" href="#moving-between-directories-with-cd"></a>
</h2>
<hr class="half-width">
<p>The command <code>cd</code> (<strong>change directory</strong>) moves
you into a different folder.</p>
<p>Example:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="bu">cd</span> mydata</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre>
</div>
<div class="section level3">
<h3 id="special-directories">Special directories<a class="anchor" aria-label="anchor" href="#special-directories"></a>
</h3>
<ul>
<li>
<code>~</code> — your home directory<br>
</li>
<li>
<code>.</code> — the current directory<br>
</li>
<li>
<code>..</code> — the parent directory (one level up)</li>
</ul>
<p>From the Unix tutorial PDF (Exercise 5), you can chain these
together:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> ~/temp/stuff/things</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="bu">cd</span> ../../..</span></code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="absolute-vs-relative-paths">Absolute vs relative paths<a class="anchor" aria-label="anchor" href="#absolute-vs-relative-paths"></a>
</h2>
<hr class="half-width">
<div id="absolute-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="absolute-paths" class="callout-inner">
<h3 class="callout-title">Absolute paths</h3>
<div class="callout-content">
<p>Start with <code>/</code> and specify the full path from the
root.</p>
<p>Example:</p>
<pre><code>/home/learner01/mydata/reads/</code></pre>
<div class="section level3">
<h3 id="relative-paths">Relative paths<a class="anchor" aria-label="anchor" href="#relative-paths"></a>
</h3>
<p>Start from your current directory.</p>
<p>Example (from within <code>mydata</code>):</p>
<pre><code>reads/</code></pre>
<p>Understanding this distinction prevents mistakes later when running
tools like FastQC, SPAdes, and Prokka, all of which require accurate
file paths.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="combining-pwd-cd-and-ls">Combining <code>pwd</code>, <code>cd</code>, and
<code>ls</code>
<a class="anchor" aria-label="anchor" href="#combining-pwd-cd-and-ls"></a>
</h2>
<hr class="half-width">
<p>These commands form the navigation toolkit you will use throughout
the entire workshop:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">pwd</span>      <span class="co"># where am I?</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">ls</span>       <span class="co"># what is here?</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="bu">cd</span> dir   <span class="co"># go somewhere else</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="bu">cd</span> ..    <span class="co"># go up one level</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="bu">cd</span> ~     <span class="co"># return home</span></span></code></pre>
</div>
<p>You will use them constantly once you begin working with:</p>
<ul>
<li>
<code>reads/</code> directories,</li>
<li>
<code>trimmed/</code> sequence folders,</li>
<li>SPAdes output directories,</li>
<li>Prokka annotation folders.</li>
</ul>
<div id="exercise-explore-your-home-directory" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-explore-your-home-directory" class="callout-inner">
<h3 class="callout-title">Exercise: Explore your home directory</h3>
<div class="callout-content">
<p>Run:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ol style="list-style-type: decimal">
<li>Which files or directories do you see?</li>
<li>Which items are directories?</li>
<li>Are any hidden files present?</li>
</ol>
</div>
</div>
</div>
<div id="exercise-practice-moving-around" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-practice-moving-around" class="callout-inner">
<h3 class="callout-title">Exercise: Practice moving around</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Create a small directory structure:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> practice/level1/level2</span></code></pre>
</div>
</li>
<li>
<p>Navigate into the deepest level:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="bu">cd</span> practice/level1/level2</span></code></pre>
</div>
</li>
<li><p>Use <code>pwd</code> — what path do you see?</p></li>
<li><p>Navigate back to your home directory in <strong>one
command</strong>.</p></li>
</ol>
</div>
</div>
</div>
<div id="exercise-absolute-or-relative" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-absolute-or-relative" class="callout-inner">
<h3 class="callout-title">Exercise: Absolute or relative?</h3>
<div class="callout-content">
<p>For each path below, decide whether it is absolute or relative:</p>
<ol style="list-style-type: decimal">
<li><code>/home/learner01</code></li>
<li><code>reads/ERR435025_R1.fastq.gz</code></li>
<li><code>../trimmed/</code></li>
<li><code>/usr/bin/</code></li>
<li><code>./scripts/run_spades.sh</code></li>
</ol>
<p>Discuss why absolute paths are often safer when running long
workflows.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The Unix filesystem is a tree with <code>/</code> at the root and
<code>~</code> as your home.</li>
<li>
<code>pwd</code> tells you where you are; <code>ls</code> shows what
is in a directory.</li>
<li>
<code>cd</code> moves you between directories using absolute or
relative paths.</li>
<li>Understanding paths is essential for accessing sequencing files and
running genomics tools.</li>
<li>Good navigation skills reduce errors when working with large
datasets.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-working-with-files"><p>Content from <a href="04-working-with-files.html">Working With Files</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/04-working-with-files.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 55 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I create, view, copy, move, and delete files in Unix?</li>
<li>How do I redirect output to files?</li>
<li>How do these skills prepare us for handling FASTQ, FASTA, QC,
trimming, and assembly outputs?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create empty files using <code>touch</code>.</li>
<li>View file contents using <code>cat</code>, <code>less</code>, and
redirection.</li>
<li>Copy and move files using <code>cp</code> and <code>mv</code>.</li>
<li>Delete files and directories safely using <code>rm</code> and
<code>rmdir</code>.</li>
<li>Use redirection (<code>&gt;</code>, <code>&gt;&gt;</code>) to create
and modify files.</li>
<li>Build comfort with file manipulation before working with large
genomics datasets.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode incorporates content from the Unix tutorial PDF
(Exercises 7–11), focusing on file creation, editing, copying, moving,
and deletion. These skills directly support genomics workflows where
FASTQ, trimmed reads, assemblies, and annotation outputs must be
manipulated.</p>
</blockquote>
<section><h2 class="section-heading" id="creating-new-files-with-touch">Creating new files with <code>touch</code>
<a class="anchor" aria-label="anchor" href="#creating-new-files-with-touch"></a>
</h2>
<hr class="half-width">
<p>From the Unix tutorial (Exercise 7), <code>touch</code> creates an
empty file:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">touch</span> notes.txt</span></code></pre>
</div>
<p>Check that it exists:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span></span></code></pre>
</div>
<p>You will later use this skill when organizing metadata, scripts,
logs, and assembly outputs.</p>
</section><section><h2 class="section-heading" id="redirecting-output-with-and">Redirecting output with <code>&gt;</code> and
<code>&gt;&gt;</code>
<a class="anchor" aria-label="anchor" href="#redirecting-output-with-and"></a>
</h2>
<hr class="half-width">
<p>The <code>echo</code> command prints text.<br>
Pair it with redirection to send output into files (Unix tutorial
Exercise 8).</p>
<div class="section level3">
<h3 id="overwrite-a-file">Overwrite a file:<a class="anchor" aria-label="anchor" href="#overwrite-a-file"></a>
</h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Quality control complete."</span> <span class="op">&gt;</span> log.txt</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="append-to-a-file">Append to a file:<a class="anchor" aria-label="anchor" href="#append-to-a-file"></a>
</h3>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Assembly started at: </span><span class="va">$(</span><span class="fu">date</span><span class="va">)</span><span class="st">"</span> <span class="op">&gt;&gt;</span> log.txt</span></code></pre>
</div>
<div id="overwrite-vs-append" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="overwrite-vs-append" class="callout-inner">
<h3 class="callout-title">Overwrite vs append</h3>
<div class="callout-content">
<ul>
<li>
<code>&gt;</code> replaces the entire file (use cautiously).</li>
<li>
<code>&gt;&gt;</code> adds new text at the end (safe for
logging).</li>
</ul>
<p>These are especially important when capturing quality control and
assembly logs later.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="viewing-files-with-cat-and-less">Viewing files with <code>cat</code> and <code>less</code>
<a class="anchor" aria-label="anchor" href="#viewing-files-with-cat-and-less"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<code>cat filename</code> prints the whole file at once.</li>
<li>
<code>less filename</code> allows scrolling (<code>q</code> to
quit).</li>
</ul>
<p>These tools are essential for inspecting:</p>
<ul>
<li>FASTQ headers</li>
<li>log files</li>
<li>assembly reports</li>
<li>annotation summaries</li>
</ul>
<p>Example:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">less</span> log.txt</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="copying-files-with-cp">Copying files with <code>cp</code>
<a class="anchor" aria-label="anchor" href="#copying-files-with-cp"></a>
</h2>
<hr class="half-width">
<p>From the Unix tutorial (Exercise 9):</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">cp</span> original.txt backup.txt</span></code></pre>
</div>
<p>Copy to a directory:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">cp</span> notes.txt archive/</span></code></pre>
</div>
<p>Copy entire directories:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> results/ results_backup/</span></code></pre>
</div>
<p>This becomes essential when backing up read files before quality
trimming or assembly.</p>
</section><section><h2 class="section-heading" id="moving-or-renaming-files-with-mv">Moving or renaming files with <code>mv</code>
<a class="anchor" aria-label="anchor" href="#moving-or-renaming-files-with-mv"></a>
</h2>
<hr class="half-width">
<p><code>mv</code> can move <strong>or</strong> rename files:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">mv</span> draft.txt final.txt</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">mv</span> final.txt documents/</span></code></pre>
</div>
<p>You will later use this to organize outputs such as:</p>
<ul>
<li><code>ERR435025_assembly/</code></li>
<li>trimmed reads directories</li>
<li>annotation output folders</li>
</ul></section><section><h2 class="section-heading" id="removing-files-and-directories">Removing files and directories<a class="anchor" aria-label="anchor" href="#removing-files-and-directories"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="remove-a-file">Remove a file:<a class="anchor" aria-label="anchor" href="#remove-a-file"></a>
</h3>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">rm</span> oldfile.txt</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="remove-an-empty-directory">Remove an empty directory:<a class="anchor" aria-label="anchor" href="#remove-an-empty-directory"></a>
</h3>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">rmdir</span> emptydir</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="remove-a-directory-with-contents">Remove a directory <em>with contents</em>:<a class="anchor" aria-label="anchor" href="#remove-a-directory-with-contents"></a>
</h3>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-rf</span> tempdir</span></code></pre>
</div>
<div id="warning-rm--rf" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="warning-rm--rf" class="callout-inner">
<h3 class="callout-title">Warning: <code>rm -rf</code>
</h3>
<div class="callout-content">
<p>This command deletes <strong>everything</strong> in a directory
<strong>permanently</strong> and <strong>cannot be undone</strong>.</p>
<p>Use it carefully, especially when working with important read files
or assemblies.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="editing-files-with-nano">Editing files with <code>nano</code>
<a class="anchor" aria-label="anchor" href="#editing-files-with-nano"></a>
</h2>
<hr class="half-width">
<p>Open a file for editing:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">nano</span> notes.txt</span></code></pre>
</div>
<p>Use:</p>
<ul>
<li>
<code>Ctrl + O</code> to save</li>
<li>
<code>Ctrl + X</code> to exit</li>
</ul>
<p>Log files, small scripts, and metadata files can all be edited this
way.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-create-and-inspect-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-create-and-inspect-files" class="callout-inner">
<h3 class="callout-title">Exercise: Create and inspect files</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Create a file:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">touch</span> example.txt</span></code></pre>
</div>
</li>
<li>
<p>Add text using <code>&gt;</code>:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"First line"</span> <span class="op">&gt;</span> example.txt</span></code></pre>
</div>
</li>
<li>
<p>Append text:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Second line"</span> <span class="op">&gt;&gt;</span> example.txt</span></code></pre>
</div>
</li>
<li>
<p>View using:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">cat</span> example.txt</span></code></pre>
</div>
</li>
</ol>
<p><strong>Questions:</strong><br>
- What happened when you used <code>&gt;</code>?<br>
- What changed when you used <code>&gt;&gt;</code>?</p>
</div>
</div>
</div>
<div id="exercise-copy-and-move-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-copy-and-move-files" class="callout-inner">
<h3 class="callout-title">Exercise: Copy and move files</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Copy your file:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">cp</span> example.txt example_copy.txt</span></code></pre>
</div>
</li>
<li>
<p>Make a directory:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">mkdir</span> workspace</span></code></pre>
</div>
</li>
<li>
<p>Move both files into the directory:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">mv</span> example<span class="pp">*</span>.txt workspace/</span></code></pre>
</div>
</li>
</ol>
<p><strong>Question:</strong><br>
- How can you confirm that the files moved successfully?</p>
</div>
</div>
</div>
<div id="exercise-remove-files-safely" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-remove-files-safely" class="callout-inner">
<h3 class="callout-title">Exercise: Remove files safely</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>
<p>Make a directory with files:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">mkdir</span> testdir</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"data"</span> <span class="op">&gt;</span> testdir/a.txt</span></code></pre>
</div>
</li>
<li>
<p>Try:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">rmdir</span> testdir</span></code></pre>
</div>
<p>Why does it fail?</p>
</li>
<li>
<p>Remove the directory properly:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> testdir</span></code></pre>
</div>
</li>
</ol>
<p>Discuss safety concerns around using <code>rm -rf</code>.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>
<code>touch</code> creates files; <code>cat</code> and
<code>less</code> view them.</li>
<li>
<code>cp</code> copies files; <code>mv</code> moves or renames
them.</li>
<li>
<code>rm</code> and <code>rmdir</code> delete files or
directories.</li>
<li>
<code>&gt;</code> overwrites files; <code>&gt;&gt;</code> appends to
them.</li>
<li>These skills are essential for managing reads, QC logs, assemblies,
and annotation outputs.</li>
<li>Be extremely careful when removing directories, especially with
<code>rm -rf</code>.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-05-tidiness"><p>Content from <a href="05-tidiness.html">Tidiness &amp; Project Organization</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/05-tidiness.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How should we organize files for a microbial genomics project?</li>
<li>Why is tidy organization essential for QC, trimming, assembly, and
annotation workflows?</li>
<li>What naming conventions and directory structures prevent errors
later?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the importance of tidy project organization in
genomics.</li>
<li>Create a clear directory structure for raw reads, trimmed reads, QC
outputs, assemblies, and annotations.</li>
<li>Apply consistent and meaningful file naming conventions.</li>
<li>Understand how well‑organized projects reduce mistakes in downstream
commands.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode uses best practices from the Unix tutorial and prepares
for handling large datasets and multi-step genomics workflows in later
episodes.</p>
</blockquote>
<section><h2 class="section-heading" id="why-organization-matters-in-genomics">Why organization matters in genomics<a class="anchor" aria-label="anchor" href="#why-organization-matters-in-genomics"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics projects quickly accumulate many files:</p>
<ul>
<li>raw FASTQ files<br>
</li>
<li>QC reports<br>
</li>
<li>trimmed reads<br>
</li>
<li>assembly outputs<br>
</li>
<li>annotation files<br>
</li>
<li>logs and scripts</li>
</ul>
<p>Without structure, it becomes easy to:</p>
<ul>
<li>run tools on the wrong files,<br>
</li>
<li>overwrite results,<br>
</li>
<li>mix up samples, or<br>
</li>
<li>accidentally delete important data.</li>
</ul>
<p>A tidy filesystem prevents these mistakes.</p>
</section><section><h2 class="section-heading" id="recommended-project-structure">Recommended project structure<a class="anchor" aria-label="anchor" href="#recommended-project-structure"></a>
</h2>
<hr class="half-width">
<p>A common layout is:</p>
<pre><code>project/
├── raw_reads/
├── qc_reports/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
<p>As taught in earlier Unix episodes, this structure uses directories
effectively to keep related files together.</p>
</section><section><h2 class="section-heading" id="naming-conventions">Naming conventions<a class="anchor" aria-label="anchor" href="#naming-conventions"></a>
</h2>
<hr class="half-width">
<p>Use filenames that are:</p>
<ul>
<li>
<strong>descriptive</strong><br>
</li>
<li>
<strong>consistent</strong><br>
</li>
<li>
<strong>machine‑friendly</strong> (no spaces!)</li>
</ul>
<p>Examples:</p>
<ul>
<li><code>sample01_R1.fastq.gz</code></li>
<li><code>sample01_R2.fastq.gz</code></li>
<li><code>ERR435025_assembly/</code></li>
<li><code>prokka_sample01/</code></li>
</ul>
<p>Avoid:</p>
<ul>
<li>
<code>my file.fastq</code><br>
</li>
<li>
<code>data stuff/</code><br>
</li>
<li><code>final_FINAL_reallyfinal.txt</code></li>
</ul>
<div id="tip-use-hyphens-or-underscores" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-use-hyphens-or-underscores" class="callout-inner">
<h3 class="callout-title">Tip: Use hyphens or underscores</h3>
<div class="callout-content">
<p>Good:</p>
<pre><code>sample01_R1.fastq.gz
run-2025-reads/</code></pre>
<p>Bad:</p>
<pre><code>Sample 1 Reads/</code></pre>
<p>Spaces break commands unless quoted—dangerous in genomics
workflows.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="planning-for-downstream-tools">Planning for downstream tools<a class="anchor" aria-label="anchor" href="#planning-for-downstream-tools"></a>
</h2>
<hr class="half-width">
<p>Later episodes will use:</p>
<ul>
<li>
<strong>FastQC</strong> → expects clear input filenames<br>
</li>
<li>
<strong>Trimming tools</strong> → produce new FASTQ files<br>
</li>
<li>
<strong>SPAdes</strong> → creates many output directories<br>
</li>
<li>
<strong>Prokka</strong> → generates several annotation files</li>
</ul>
<p>A well‑organized structure ensures:</p>
<ul>
<li>safe separation of raw vs processed data<br>
</li>
<li>clarity during command execution<br>
</li>
<li>easier troubleshooting<br>
</li>
<li>reproducible workflows</li>
</ul>
<div id="exercise-build-your-project-structure" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-build-your-project-structure" class="callout-inner">
<h3 class="callout-title">Exercise: Build your project structure</h3>
<div class="callout-content">
<p>Create the recommended structure:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> project/<span class="dt">{raw_reads</span><span class="op">,</span><span class="dt">qc_reports</span><span class="op">,</span><span class="dt">trimmed</span><span class="op">,</span><span class="dt">assembly</span><span class="op">,</span><span class="dt">annotation</span><span class="op">,</span><span class="dt">scripts}</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which directories will hold input FASTQ files?</li>
<li>Where will SPAdes results go?</li>
<li>Where should you store logs or scripts?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-rename-files-for-tidiness" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-rename-files-for-tidiness" class="callout-inner">
<h3 class="callout-title">Exercise: Rename files for tidiness</h3>
<div class="callout-content">
<p>Suppose you have files:</p>
<pre><code><span><span class="va">read1.fq.gz</span></span>
<span><span class="va">read2.fq.gz</span></span></code></pre>
<p>Rename them using a good naming convention:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">mv</span> read1.fq.gz sample01_R1.fastq.gz</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">mv</span> read2.fq.gz sample01_R2.fastq.gz</span></code></pre>
</div>
<p><strong>Question:</strong><br>
Why is this better for downstream analysis?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Good organization prevents errors in multi-step genomics
workflows.</li>
<li>Use meaningful directory structures: raw reads, QC reports, trimmed
data, assemblies, annotations.</li>
<li>Consistent, space‑free filenames make commands easier and
safer.</li>
<li>Tidy projects support reproducibility and smooth downstream
analysis.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-06-project-planning"><p>Content from <a href="06-project-planning.html">Project Planning</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/06-project-planning.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How should I plan a microbial genomics analysis before touching the
data?</li>
<li>What directories, files, and metadata should I set up ahead of
time?</li>
<li>How do I structure a workflow so QC, trimming, assembly, and
annotation run smoothly?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Anticipate the workflow steps in a microbial genomics project.</li>
<li>Plan a directory layout suitable for QC, trimming, assembly, and
annotation workflows.</li>
<li>Identify metadata you must collect before running tools.</li>
<li>Understand how planning improves reproducibility and avoids
downstream errors.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode prepares learners for the workflow used in Episodes 7–11
by building the project plan and directory structure needed for quality
control, trimming, assembly (SPAdes), and annotation (Prokka).<br>
It draws on concepts from the Unix tutorial and best practices for
genome analysis, but does <strong>not</strong> yet introduce FASTQ or
tool-specific PDF content.</p>
</blockquote>
<section><h2 class="section-heading" id="why-planning-matters">Why planning matters<a class="anchor" aria-label="anchor" href="#why-planning-matters"></a>
</h2>
<hr class="half-width">
<p>Microbial genomics workflows involve multiple steps:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Quality control</strong> (FastQC)</li>
<li><strong>Trimming</strong></li>
<li>
<strong>Genome assembly</strong> (SPAdes)</li>
<li><strong>Assembly evaluation &amp; filtering</strong></li>
<li>
<strong>Annotation</strong> (Prokka)</li>
<li><strong>Archiving results &amp; documenting methods</strong></li>
</ol>
<p>Each step produces files, logs, and subdirectories. Without
planning:</p>
<ul>
<li>files become scattered across your home directory,</li>
<li>output overwrites previous results,</li>
<li>commands reference the wrong paths,</li>
<li>reproducibility becomes impossible.</li>
</ul>
<p>A simple, thoughtful plan prevents all of this.</p>
</section><section><h2 class="section-heading" id="anticipating-the-workflow">Anticipating the workflow<a class="anchor" aria-label="anchor" href="#anticipating-the-workflow"></a>
</h2>
<hr class="half-width">
<p>Before analyzing data, ask:</p>
<ul>
<li>Where will I store raw FASTQ files?</li>
<li>Where will QC reports go?</li>
<li>Where will trimmed reads go?</li>
<li>Where should the SPAdes assembly output live?</li>
<li>Where should I store annotation results?</li>
<li>Where will scripts or notes be kept?</li>
</ul>
<p>This planning mirrors the Unix skills covered in Episodes
1–5—especially filesystem navigation and working with files.</p>
</section><section><h2 class="section-heading" id="recommended-directory-skeleton">Recommended directory skeleton<a class="anchor" aria-label="anchor" href="#recommended-directory-skeleton"></a>
</h2>
<hr class="half-width">
<p>You can create a clean starting layout:</p>
<pre><code>myproject/
├── metadata/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
├── annotation/
└── scripts/</code></pre>
<p>Purpose of each:</p>
<ul>
<li>
<strong>metadata/</strong>: sample sheets, run notes<br>
</li>
<li>
<strong>raw_reads/</strong>: original FASTQ files<br>
</li>
<li>
<strong>qc/</strong>: FastQC or other QC outputs<br>
</li>
<li>
<strong>trimmed/</strong>: files after trimming<br>
</li>
<li>
<strong>assembly/</strong>: SPAdes output folders<br>
</li>
<li>
<strong>annotation/</strong>: Prokka output<br>
</li>
<li>
<strong>scripts/</strong>: shell scripts, helper commands</li>
</ul>
<p>This structure mirrors the layout used throughout the workshop and is
consistent with good practice in microbial genomics research.</p>
<div id="planning-tip" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="planning-tip" class="callout-inner">
<h3 class="callout-title">Planning tip</h3>
<div class="callout-content">
<p>Never mix <strong>raw data</strong> with <strong>processed
data</strong>.<br>
Raw FASTQ files should remain untouched and stored in their own
directory.<br>
All processing steps should write outputs into new directories.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="metadata-planning">Metadata planning<a class="anchor" aria-label="anchor" href="#metadata-planning"></a>
</h2>
<hr class="half-width">
<p>Before running any tool, you should prepare:</p>
<ul>
<li>sample names<br>
</li>
<li>read file names (R1, R2)<br>
</li>
<li>sequencing platform<br>
</li>
<li>expected genome size (helps anticipate memory needs)<br>
</li>
<li>notes on experimental conditions</li>
</ul>
<p>A simple spreadsheet often works best.<br>
This metadata informs file naming and makes SPAdes/Prokka outputs easier
to interpret.</p>
</section><section><h2 class="section-heading" id="planning-for-tool-inputs-outputs">Planning for tool inputs &amp; outputs<a class="anchor" aria-label="anchor" href="#planning-for-tool-inputs-outputs"></a>
</h2>
<hr class="half-width">
<p>This episode does <strong>not</strong> invoke SPAdes or Prokka
directly, but prepares for their behavior:</p>
<ul>
<li>SPAdes produces an output directory with many files (logs, contigs,
scaffolds, temporary files).</li>
<li>Prokka outputs multiple files (<code>.gff</code>, <code>.faa</code>,
<code>.ffn</code>, <code>.tbl</code>, etc.).</li>
</ul>
<p>Ensuring you already have dedicated directories avoids clutter and
accidental overwriting.</p>
<div id="exercise-create-your-project-layout" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-create-your-project-layout" class="callout-inner">
<h3 class="callout-title">Exercise: Create your project layout</h3>
<div class="callout-content">
<p>Make the directory skeleton:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> myproject/<span class="dt">{metadata</span><span class="op">,</span><span class="dt">raw_reads</span><span class="op">,</span><span class="dt">qc</span><span class="op">,</span><span class="dt">trimmed</span><span class="op">,</span><span class="dt">assembly</span><span class="op">,</span><span class="dt">annotation</span><span class="op">,</span><span class="dt">scripts}</span></span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Which folder will contain raw FASTQ files?</li>
<li>Where will SPAdes output go?</li>
<li>Where should your Prokka annotation end up?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-plan-metadata-for-one-sample" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-plan-metadata-for-one-sample" class="callout-inner">
<h3 class="callout-title">Exercise: Plan metadata for one sample</h3>
<div class="callout-content">
<p>Create a text file in the <code>metadata/</code> directory:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="st">"sample_id\tread1\tread2\nsample01\tsample01_R1.fastq.gz\tsample01_R2.fastq.gz"</span> <span class="op">&gt;</span> metadata/samples.tsv</span></code></pre>
</div>
<p>Then inspect it:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">cat</span> metadata/samples.tsv</span></code></pre>
</div>
<p><strong>Question:</strong><br>
Why is it helpful to prepare metadata before running analysis tools?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Planning your project structure prevents mistakes during multi-step
analysis.</li>
<li>Separate raw data, QC results, trimmed reads, assemblies, and
annotations.</li>
<li>Metadata prepared early improves reproducibility and keeps analyses
organized.</li>
<li>This planning sets the stage for handling FASTQ files starting in
Episode 7.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-07-background"><p>Content from <a href="07-background.html">Sequence File Formats</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/07-background.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are FASTA, FASTQ, and SAM/BAM file formats?</li>
<li>How are sequence, identifiers, and quality scores stored?</li>
<li>Why are many sequence files compressed, and what does
<code>.gz</code> mean?</li>
<li>How do these formats fit into the microbial genome assembly and
annotation workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Recognize and describe the structure of FASTA and FASTQ files.</li>
<li>Understand how quality scores are represented in FASTQ.</li>
<li>Explain the role of SAM/BAM files in mapping reads to
references.</li>
<li>Identify common filename extensions used in microbial genomics.</li>
<li>Connect file formats to later workflow steps: QC, trimming,
assembly, and annotation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode incorporates concepts from the <em>Sequence File
Formats</em> notes (FASTA, FASTQ, SAM/BAM) and prepares learners to
interpret the files used in later episodes for QC, assembly, and
annotation.</p>
</blockquote>
<section><h2 class="section-heading" id="overview-of-common-file-types">Overview of common file types<a class="anchor" aria-label="anchor" href="#overview-of-common-file-types"></a>
</h2>
<hr class="half-width">
<p>In microbial genomics, you will frequently encounter:</p>
<ul>
<li>
<strong>FASTA</strong> – sequence and identifier(s) only<br>
</li>
<li>
<strong>FASTQ</strong> – sequence <em>and</em> per-base quality
scores<br>
</li>
<li>
<strong>SAM/BAM</strong> – alignments of reads to a reference</li>
</ul>
<p>File extensions help you keep track:</p>
<ul>
<li>
<code>.fasta</code>, <code>.fa</code>, <code>.fna</code>,
<code>.faa</code>, <code>.ffn</code>, <code>.frn</code>
</li>
<li>
<code>.fastq</code>, <code>.fq</code>, <code>.fastq.gz</code>
</li>
<li>
<code>.sam</code>, <code>.bam</code>
</li>
</ul>
<div id="plain-text-vs-binary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="plain-text-vs-binary" class="callout-inner">
<h3 class="callout-title">Plain text vs binary</h3>
<div class="callout-content">
<ul>
<li>FASTA, FASTQ, and SAM are <strong>plain text</strong> (you can read
them with <code>less</code>).</li>
<li>BAM is a <strong>binary</strong> version of SAM (smaller and faster
to process, not human-readable).</li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="fasta-format">FASTA format<a class="anchor" aria-label="anchor" href="#fasta-format"></a>
</h2>
<hr class="half-width">
<p>FASTA files store sequence(s) with headers. From your sequence
formats notes:</p>
<ul>
<li>Header lines start with <code>&gt;</code><br>
</li>
<li>The header is followed by one or more lines of sequence</li>
</ul>
<p>Example:</p>
<pre class="text"><code>&gt;gi|212499741|ref|YP_002308549.1| DNA polymerase [Bacteriophage APSE-2]
MQNLLFCDLETYSDIPINCGTHRYAENAEILLFAYAYNHAPVKVWDVTQDKTMPADLKAYFDDSEILTVW
HNGGMFDTVILKRVLNIDLPLSRVHDTLVQALAHGLPGALGLLCDIFNVNSDKAKDKEGKALISLLCKPR</code></pre>
<p>Key points:</p>
<ul>
<li>A FASTA file often contains <strong>many</strong> sequences, one
after another.</li>
<li>FASTA is used for genomes, contigs, proteins, etc.</li>
<li>Later, SPAdes assemblies and Prokka annotations will use
FASTA-formatted files.</li>
</ul></section><section><h2 class="section-heading" id="fastq-format">FASTQ format<a class="anchor" aria-label="anchor" href="#fastq-format"></a>
</h2>
<hr class="half-width">
<p>FASTQ is a <strong>high-density format for short reads</strong> (≤300
nt) that stores both sequence and quality.</p>
<p>Each record has <strong>four lines</strong>:</p>
<ol style="list-style-type: decimal">
<li>
<code>@</code> plus an identifier<br>
</li>
<li>the raw sequence<br>
</li>
<li>
<code>+</code> (optionally followed by the identifier)<br>
</li>
<li>ASCII-encoded quality string</li>
</ol>
<p>Example (adapted from your notes):</p>
<pre class="text"><code>@DBRHHJN1:387:H89AEADXX:2:1101:1811:1988 1:N:0:TCCTGAGCCTCTCTAT
GTATTGGATATGTCCTCTATTAAACCTTGTGATGAGGAAGCTGTCTGTCTCTTATACACATCTCCGAGCCCACGAGACTCCTGAGCA
+
BB@FFFFFHHHHHJJJJJJJJJJJJJJJJIJJJJJJJJJIJJJIJJJIIIJJJJIJIJJJJJJJJIJHIJJJHHFFDDEEEDDCDDD</code></pre>
<ul>
<li>Line 1: read name, including run, lane, tile, x/y position,
etc.</li>
<li>Line 2: sequence (A/C/G/T/N).</li>
<li>Line 4: quality scores encoded as ASCII characters.</li>
</ul>
<p>These FASTQ files are the <strong>starting point</strong> for QC,
trimming, and assembly.</p>
<div id="why-.fastq.gz" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="why-.fastq.gz" class="callout-inner">
<h3 class="callout-title">Why <code>.fastq.gz</code>?</h3>
<div class="callout-content">
<p>FASTQ files are large.<br>
They are often compressed with <strong>gzip</strong>, resulting in
<code>.fastq.gz</code> files.</p>
<p>Most tools (FastQC, trimmers, SPAdes) can read gzipped FASTQ
directly, so you rarely need to decompress them manually.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="sam-and-bam-formats">SAM and BAM formats<a class="anchor" aria-label="anchor" href="#sam-and-bam-formats"></a>
</h2>
<hr class="half-width">
<p>After assembling or mapping reads to a reference, alignments are
often stored in <strong>SAM</strong> (text) or <strong>BAM</strong>
(binary) format.</p>
<p>A (simplified) SAM snippet from your notes:</p>
<pre class="text"><code>@HD VN:1.3 SO:coordinate
@SQ SN:ref LN:45
r001 163 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *
r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *</code></pre>
<p>Each line after the header contains:</p>
<ul>
<li>read name</li>
<li>flags</li>
<li>reference name</li>
<li>position</li>
<li>CIGAR string (match/mismatch/insertions/deletions)</li>
<li>sequence and quality</li>
</ul>
<p>Although this workshop focuses on <strong>de novo assembly</strong>
and annotation, you may still encounter SAM/BAM when evaluating
assemblies or mapping reads back to contigs.</p>
</section><section><h2 class="section-heading" id="where-formats-fit-into-the-workflow">Where formats fit into the workflow<a class="anchor" aria-label="anchor" href="#where-formats-fit-into-the-workflow"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<strong>FASTQ</strong> – raw reads → input to QC and trimming<br>
</li>
<li>
<strong>FASTQ (trimmed)</strong> – input to SPAdes<br>
</li>
<li>
<strong>FASTA</strong> – contigs/scaffolds from assembly → input to
annotation (Prokka)<br>
</li>
<li>
<strong>FASTA (proteins, CDS)</strong> – output from Prokka<br>
</li>
<li>
<strong>SAM/BAM</strong> – optional, for read mapping to references
or assemblies</li>
</ul>
<p>Understanding these formats now will make later steps much less
mysterious.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-identify-format-by-eye" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-identify-format-by-eye" class="callout-inner">
<h3 class="callout-title">Exercise: Identify format by eye</h3>
<div class="callout-content">
<p>Your instructor will show you several short text snippets (or
files).<br>
For each one, decide whether it is:</p>
<ul>
<li>FASTA</li>
<li>FASTQ</li>
<li>SAM</li>
</ul>
<p><strong>Questions:</strong></p>
<ul>
<li>What clues tell you the format?</li>
<li>How many lines per record can you see?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-inspect-a-fastq-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-a-fastq-file" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect a FASTQ file</h3>
<div class="callout-content">
<p>On the server, navigate to where a small sample FASTQ file is stored
and run:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">zcat</span> sample.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 8</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How many complete records are shown in 8 lines?</li>
<li>Which lines are the sequences?</li>
<li>Which lines are the quality scores?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-inspect-an-assembly-fasta-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-inspect-an-assembly-fasta-file" class="callout-inner">
<h3 class="callout-title">Exercise: Inspect an assembly FASTA file</h3>
<div class="callout-content">
<p>Later, after running SPAdes, you will have a scaffolds FASTA
file.<br>
Practice on a small example (or instructor-provided file):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span> scaffolds.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What does each header look like?</li>
<li>Can you tell how many contigs are in the file
(<code>grep -c '^&gt;' scaffolds.fasta</code>)?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>FASTA stores sequence(s) with header lines starting with
<code>&gt;</code>.</li>
<li>FASTQ stores reads with <strong>four lines per record</strong>,
including quality scores.</li>
<li>Many FASTQ files are compressed as <code>.fastq.gz</code> to save
space.</li>
<li>SAM/BAM store read alignments to a reference genome or
assembly.</li>
<li>Understanding these formats is essential for QC, trimming, assembly,
and annotation.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-08-quality-control"><p>Content from <a href="08-quality-control.html">Quality Control of Reads</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/08-quality-control.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we perform quality control (QC) on raw sequencing reads?</li>
<li>What kinds of issues can QC reveal in microbial genomics
datasets?</li>
<li>How do we run FastQC on the command line and interpret its basic
outputs?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why QC is a critical first step before trimming and
assembly.</li>
<li>Run FastQC on compressed FASTQ files using the command line.</li>
<li>Locate and view FastQC HTML reports.</li>
<li>Identify major QC metrics: per-base quality, sequence length,
overrepresented sequences, adapters.</li>
<li>Connect QC findings to downstream decisions (trimming, filtering,
assembly).</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode incorporates the FastQC step described in the
<em>Assembling a Microbial Genome</em> activity and connects it to the
sequence file formats from Episode 7. We focus on command‑line usage and
interpretation of key metrics, even if full QC is abbreviated in
class.</p>
</blockquote>
<section><h2 class="section-heading" id="why-quality-control">Why quality control?<a class="anchor" aria-label="anchor" href="#why-quality-control"></a>
</h2>
<hr class="half-width">
<p>Raw Illumina reads may contain:</p>
<ul>
<li>low‑quality bases at the ends of reads,</li>
<li>leftover adapter sequences,</li>
<li>duplicated reads,</li>
<li>unexpected sequence content.</li>
</ul>
<p>These issues can:</p>
<ul>
<li>decrease assembly quality,</li>
<li>increase misassemblies,</li>
<li>waste computational resources.</li>
</ul>
<p>Quality control with tools like <strong>FastQC</strong> helps you
decide:</p>
<ul>
<li>whether trimming is necessary,</li>
<li>which trimming parameters to use,</li>
<li>whether your data looks as expected.</li>
</ul></section><section><h2 class="section-heading" id="fastqc-overview">FastQC overview<a class="anchor" aria-label="anchor" href="#fastqc-overview"></a>
</h2>
<hr class="half-width">
<p>FastQC can be run:</p>
<ul>
<li>through a <strong>graphical user interface (GUI)</strong>, or</li>
<li>on the <strong>command line</strong>.</li>
</ul>
<p>For a remote server / cloud workflow, the <strong>command‑line
approach</strong> is most practical, as described in your assembly
activity.</p>
<p>FastQC accepts <strong>FASTQ</strong> (and <code>.fastq.gz</code>)
files and produces:</p>
<ul>
<li>an <code>.html</code> report (interactive)</li>
<li>a <code>.zip</code> file containing detailed data</li>
</ul></section><section><h2 class="section-heading" id="running-fastqc-on-the-command-line">Running FastQC on the command line<a class="anchor" aria-label="anchor" href="#running-fastqc-on-the-command-line"></a>
</h2>
<hr class="half-width">
<p>From your assembly exercise, a typical command is:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">fastqc</span> reads/ERR435025_R1.fastq.gz reads/ERR435025_R2.fastq.gz</span></code></pre>
</div>
<p>This:</p>
<ul>
<li>reads compressed FASTQ files from the <code>reads/</code>
directory,</li>
<li>writes HTML and ZIP output files in the current directory.</li>
</ul>
<p>You can choose an output directory with <code>-o</code>:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> qc</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc reads/ERR435025_R1.fastq.gz reads/ERR435025_R2.fastq.gz</span></code></pre>
</div>
<p>After running, list the QC directory:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span> qc</span></code></pre>
</div>
<p>You should see files like:</p>
<ul>
<li><code>ERR435025_R1_fastqc.html</code></li>
<li><code>ERR435025_R2_fastqc.html</code></li>
<li>matching <code>.zip</code> files</li>
</ul>
<div id="viewing-fastqc-reports" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="viewing-fastqc-reports" class="callout-inner">
<h3 class="callout-title">Viewing FastQC reports</h3>
<div class="callout-content">
<p>On a remote server, you have two main options:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Download the HTML files</strong> to your local machine (via
<code>scp</code>, SFTP, or a client like Cyberduck) and open them in a
browser.</li>
<li>If available, open them directly in a remote browser session
(e.g. <code>firefox</code> or similar), as suggested in your assembly
notes.</li>
</ol>
<p>In most workshops, downloading and viewing locally is simplest.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="key-fastqc-modules">Key FastQC modules<a class="anchor" aria-label="anchor" href="#key-fastqc-modules"></a>
</h2>
<hr class="half-width">
<p>Some important panels in the FastQC HTML report include:</p>
<ul>
<li>
<strong>Per base sequence quality</strong>
<ul>
<li>Boxplots of quality at each position in the read.<br>
</li>
<li>Low quality at the 3′ end often suggests the need for trimming.</li>
</ul>
</li>
<li>
<strong>Per sequence quality scores</strong>
<ul>
<li>Distribution of average quality per read.</li>
</ul>
</li>
<li>
<strong>Per base sequence content</strong>
<ul>
<li>Proportions of A/C/G/T at each base position.</li>
</ul>
</li>
<li>
<strong>Per sequence GC content</strong>
<ul>
<li>Deviations may indicate contamination.</li>
</ul>
</li>
<li>
<strong>Overrepresented sequences</strong>
<ul>
<li>May reveal adapters or other technical artifacts.</li>
</ul>
</li>
</ul>
<p>Your goal is not to memorize every module but to recognize:</p>
<ul>
<li>whether data quality looks generally <strong>good</strong> or
<strong>poor</strong>, and</li>
<li>whether adapters or other contaminants are present.</li>
</ul></section><section><h2 class="section-heading" id="qc-in-the-genome-assembly-workflow">QC in the genome assembly workflow<a class="anchor" aria-label="anchor" href="#qc-in-the-genome-assembly-workflow"></a>
</h2>
<hr class="half-width">
<p>In the assembly activity, QC is described as an early step, even if
it is sometimes skipped in class to save time.</p>
<p>The typical real‑world workflow is:</p>
<ol style="list-style-type: decimal">
<li>Run <strong>FastQC</strong> on raw FASTQ files.</li>
<li>Use the reports to:
<ul>
<li>decide on trimming parameters,</li>
<li>confirm read length and quality.</li>
</ul>
</li>
<li>Perform <strong>trimming</strong> (next episode).</li>
<li>Optionally, run FastQC again on trimmed reads.</li>
</ol>
<p>Even when abbreviated, understanding the QC step helps interpret
downstream assembly quality and potential problems.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-run-fastqc-on-one-read-file" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-fastqc-on-one-read-file" class="callout-inner">
<h3 class="callout-title">Exercise: Run FastQC on one read file</h3>
<div class="callout-content">
<p>Assuming your raw reads are in <code>raw_reads/</code> or
<code>reads/</code>, run:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> qc</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc raw_reads/sample01_R1.fastq.gz</span></code></pre>
</div>
<p>Then:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">ls</span> qc</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What output files were created?</li>
<li>How can you transfer the HTML file to your local machine for
viewing?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-interpret-a-fastqc-report" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-interpret-a-fastqc-report" class="callout-inner">
<h3 class="callout-title">Exercise: Interpret a FastQC report</h3>
<div class="callout-content">
<p>Working from an instructor‑provided FastQC HTML report:</p>
<ul>
<li>Examine the <strong>Per base sequence quality</strong> module.</li>
<li>Check whether most bases are in the green (high quality) zone.</li>
<li>Look at the <strong>Overrepresented sequences</strong> module.</li>
</ul>
<p><strong>Questions:</strong></p>
<ul>
<li>Would you recommend trimming this dataset? Why or why not?</li>
<li>Do you see evidence of adapter contamination?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-connect-qc-to-downstream-steps" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-connect-qc-to-downstream-steps" class="callout-inner">
<h3 class="callout-title">Exercise: Connect QC to downstream steps</h3>
<div class="callout-content">
<p>Discuss with a partner:</p>
<ol style="list-style-type: decimal">
<li>How could poor QC (low quality, many adapters) affect:
<ul>
<li>genome assembly with SPAdes?</li>
<li>annotation with Prokka?</li>
</ul>
</li>
<li>How might good QC save time and avoid misleading results?</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>QC is a critical first step before trimming and assembly in
microbial genomics.</li>
<li>FastQC summarizes read quality, sequence content, GC content, and
overrepresented sequences.</li>
<li>FastQC can read compressed <code>.fastq.gz</code> files and produces
HTML reports for review.</li>
<li>QC results guide trimming choices and help interpret assembly
quality.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-09-trimming"><p>Content from <a href="09-trimming.html">Trimming and Cleaning Reads</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/09-trimming.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why do we trim and filter raw reads before assembly?</li>
<li>What kinds of problems can trimming fix?</li>
<li>How do we run a typical trimming tool on paired-end FASTQ
files?</li>
<li>How do trimmed reads fit into the downstream SPAdes and Prokka
workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain why adapter and quality trimming are important for microbial
genome assembly.</li>
<li>Describe common trimming operations (adapter removal, quality
trimming, length filtering).</li>
<li>Run a representative trimming command on paired-end FASTQ
files.</li>
<li>Verify that trimmed reads are created and understand where they
should be stored.</li>
<li>Connect trimming decisions to FastQC results and assembly
quality.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode builds on the QC step from Episode 8 and the sequencing
file formats from Episode 7. While your assembly PDF focuses on QC and
assembly, trimming is a common intermediate step that improves assembly
performance and quality, so we discuss it explicitly here.</p>
</blockquote>
<section><h2 class="section-heading" id="why-trim-reads">Why trim reads?<a class="anchor" aria-label="anchor" href="#why-trim-reads"></a>
</h2>
<hr class="half-width">
<p>After QC with FastQC, you may discover:</p>
<ul>
<li>low-quality bases at the <strong>3′ ends</strong> of reads,</li>
<li>
<strong>adapter sequences</strong> still present,</li>
<li>very short or poor-quality reads.</li>
</ul>
<p>Leaving these issues uncorrected can lead to:</p>
<ul>
<li>fragmented assemblies,</li>
<li>misassemblies,</li>
<li>wasted compute and storage.</li>
</ul>
<p>Trimming tools remove or correct problematic parts of reads before
assembly.</p>
<p>Common trimming goals:</p>
<ul>
<li>remove adapters,</li>
<li>trim low-quality bases from ends,</li>
<li>discard reads that become too short.</li>
</ul></section><section><h2 class="section-heading" id="typical-trimming-workflow">Typical trimming workflow<a class="anchor" aria-label="anchor" href="#typical-trimming-workflow"></a>
</h2>
<hr class="half-width">
<p>A trimming tool will:</p>
<ul>
<li>take <strong>paired-end FASTQ</strong> files as input,</li>
<li>write new <strong>trimmed</strong> FASTQ files to a different
directory,</li>
<li>optionally produce a log summarizing how many reads were kept or
discarded.</li>
</ul>
<p>A generic command-line pattern looks like:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">trimmer</span>   <span class="at">-i</span> raw_reads/sample01_R1.fastq.gz   <span class="at">-I</span> raw_reads/sample01_R2.fastq.gz   <span class="at">-o</span> trimmed/sample01_R1.trimmed.fastq.gz   <span class="at">-O</span> trimmed/sample01_R2.trimmed.fastq.gz   <span class="pp">[</span><span class="ss">options...</span><span class="pp">]</span></span></code></pre>
</div>
<p>The exact options and flags vary by tool (e.g., Trimmomatic, fastp,
Cutadapt), but this pattern is common:</p>
<ul>
<li>
<strong>input reads</strong> → raw FASTQ files,</li>
<li>
<strong>output reads</strong> → trimmed FASTQ files in a separate
directory.</li>
</ul>
<div id="directory-layout-reminder" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="directory-layout-reminder" class="callout-inner">
<h3 class="callout-title">Directory layout reminder</h3>
<div class="callout-content">
<p>From Episode 6 (Project Planning), trimmed reads should go in a
dedicated directory, e.g.:</p>
<pre><code>myproject/
├── raw_reads/
├── qc/
├── trimmed/
├── assembly/
└── annotation/</code></pre>
<p>Never overwrite your raw FASTQ files when trimming.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="connecting-qc-to-trimming-choices">Connecting QC to trimming choices<a class="anchor" aria-label="anchor" href="#connecting-qc-to-trimming-choices"></a>
</h2>
<hr class="half-width">
<p>FastQC reports from Episode 8 help answer:</p>
<ul>
<li>Are per-base quality scores dropping sharply at the end of
reads?</li>
<li>Do we see clear evidence of adapter contamination?</li>
<li>Are there overrepresented sequences that look like adapters?</li>
</ul>
<p>Based on these findings, you might choose to:</p>
<ul>
<li>trim the last <strong>10–20 bases</strong> of each read,</li>
<li>remove known adapter sequences,</li>
<li>filter out reads below a certain minimum length.</li>
</ul>
<p>You don’t need to memorize specific thresholds, but you should
understand:</p>
<ul>
<li>QC → informs trimming<br>
</li>
<li>Trimming → improves assembly input</li>
</ul></section><section><h2 class="section-heading" id="after-trimming-checking-results">After trimming: checking results<a class="anchor" aria-label="anchor" href="#after-trimming-checking-results"></a>
</h2>
<hr class="half-width">
<p>After trimming, you should:</p>
<ol style="list-style-type: decimal">
<li>Confirm that the new files exist:</li>
</ol>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span> trimmed/</span></code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li>Optionally, run FastQC again on the trimmed reads:</li>
</ol>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc trimmed/sample01_R1.trimmed.fastq.gz trimmed/sample01_R2.trimmed.fastq.gz</span></code></pre>
</div>
<ol start="3" style="list-style-type: decimal">
<li>Compare QC reports <strong>before</strong> and
<strong>after</strong> trimming:</li>
</ol>
<ul>
<li>Are low-quality tails reduced?</li>
<li>Are adapter sequences gone or reduced?</li>
<li>Are length distributions as expected?</li>
</ul>
<p>The trimmed FASTQ files will be the inputs to <strong>SPAdes</strong>
in Episode 10.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-plan-your-trimming-command" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-plan-your-trimming-command" class="callout-inner">
<h3 class="callout-title">Exercise: Plan your trimming command</h3>
<div class="callout-content">
<p>Working with a partner and based on your FastQC results:</p>
<ol style="list-style-type: decimal">
<li>Decide:
<ul>
<li>Do you need adapter trimming?</li>
<li>Do you need to cut low-quality bases from the ends?</li>
<li>What should your minimum read length be?</li>
</ul>
</li>
<li>Draft a pseudocode command (not necessarily using a specific tool
yet), e.g.:</li>
</ol>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">trimmer</span> <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  <span class="at">-i</span> raw_reads/sample01_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">-I</span> raw_reads/sample01_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">-o</span> trimmed/sample01_R1.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">-O</span> trimmed/sample01_R2.trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  [quality and adapter options...]</span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-check-your-trimmed-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-check-your-trimmed-files" class="callout-inner">
<h3 class="callout-title">Exercise: Check your trimmed files</h3>
<div class="callout-content">
<p>After running a trimming tool (demonstrated by the instructor):</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">ls</span> trimmed/</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">zcat</span> trimmed/sample01_R1.trimmed.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 4</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Does the file name clearly indicate that the reads are trimmed?</li>
<li>Can you see that the sequences have been shortened relative to the
raw reads?</li>
<li>How many records remain compared to the original file
(<code>zgrep -c '^@'</code>)?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-qc-after-trimming" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-qc-after-trimming" class="callout-inner">
<h3 class="callout-title">Exercise: QC after trimming</h3>
<div class="callout-content">
<p>If time permits, run FastQC on the trimmed reads:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-o</span> qc trimmed/sample01_R1.trimmed.fastq.gz trimmed/sample01_R2.trimmed.fastq.gz</span></code></pre>
</div>
<p>Compare:</p>
<ul>
<li>Per-base quality profiles before vs after trimming.</li>
<li>Overrepresented sequences in both reports.</li>
</ul>
<p><strong>Question:</strong><br>
Do you think the data are now ready to be used as input to SPAdes?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Trimming removes low-quality bases and adapters that can interfere
with assembly.</li>
<li>Trimmed reads should be saved in a separate <code>trimmed/</code>
directory; raw data must not be overwritten.</li>
<li>QC (FastQC) guides trimming decisions and can be rerun afterward to
confirm improvements.</li>
<li>Trimmed FASTQ files will be used as inputs to SPAdes in the genome
assembly step.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-10-assembly-spades"><p>Content from <a href="10-assembly-spades.html">Genome Assembly with SPAdes</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/10-assembly-spades.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we assemble microbial genomes from paired-end reads?</li>
<li>How do we run SPAdes on the command line?</li>
<li>What are the key SPAdes options we need to set (input files,
threads, memory, output directory)?</li>
<li>What files does SPAdes produce and which ones are most
important?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the role of de novo assembly in microbial genomics.</li>
<li>Run SPAdes on paired-end reads using appropriate options.</li>
<li>Locate SPAdes output files (especially <code>contigs.fasta</code>
and <code>scaffolds.fasta</code>).</li>
<li>Perform a basic first inspection of the assembly output.</li>
<li>Connect assembly results to later annotation with Prokka.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode uses the microbial genome assembly activity you provided
and adapts the SPAdes workflow for a Carpentries-style lesson. It
assumes reads have already been organized, QC’d, and optionally trimmed
in earlier episodes.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-genome-assembly">What is genome assembly?<a class="anchor" aria-label="anchor" href="#what-is-genome-assembly"></a>
</h2>
<hr class="half-width">
<p>In <strong>de novo genome assembly</strong>, we reconstruct a genome
from many short reads:</p>
<ul>
<li>Raw reads (FASTQ) → overlap / k-mer graph → contigs →
scaffolds.</li>
<li>For bacterial genomes, final sizes are often <strong>2–6
Mb</strong>.</li>
<li>Assemblers like <strong>SPAdes</strong> are optimized for microbial
genomes.</li>
</ul>
<p>Our workflow:</p>
<ol style="list-style-type: decimal">
<li>Ensure reads are available in a project directory.</li>
<li>Run SPAdes with paired-end inputs.</li>
<li>Inspect key output files (especially the assembly FASTA).</li>
<li>Filter or post-process the assembly (covered here at a basic
level).</li>
</ol></section><section><h2 class="section-heading" id="setting-up-the-assembly-directory">Setting up the assembly directory<a class="anchor" aria-label="anchor" href="#setting-up-the-assembly-directory"></a>
</h2>
<hr class="half-width">
<p>Following the earlier project structure, we will work in the
<code>assembly/</code> directory:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">cd</span> myproject</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span> assembly</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>You should see that your <strong>current working directory</strong>
is <code>.../myproject/assembly</code>.</p>
<p>Make sure your reads are accessible, for example:</p>
<ul>
<li><code>../trimmed/sample01_R1.trimmed.fastq.gz</code></li>
<li><code>../trimmed/sample01_R2.trimmed.fastq.gz</code></li>
</ul>
<p>or, if trimming was skipped in this run:</p>
<ul>
<li><code>../raw_reads/sample01_R1.fastq.gz</code></li>
<li><code>../raw_reads/sample01_R2.fastq.gz</code></li>
</ul></section><section><h2 class="section-heading" id="running-spades">Running SPAdes<a class="anchor" aria-label="anchor" href="#running-spades"></a>
</h2>
<hr class="half-width">
<p>SPAdes is run with a command that specifies:</p>
<ul>
<li>
<code>-1</code> – path to R1 FASTQ,</li>
<li>
<code>-2</code> – path to R2 FASTQ,</li>
<li>
<code>--threads</code> – number of CPU cores,</li>
<li>
<code>--memory</code> – maximum RAM (in GB),</li>
<li>
<code>-k</code> – list of k-mer sizes,</li>
<li>
<code>-o</code> – output directory.</li>
</ul>
<p>A typical command, adapted from your assembly activity, looks
like:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">spades.py</span> <span class="dt">\</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="at">-1</span> ../raw_reads/ERR435025_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="at">-2</span> ../raw_reads/ERR435025_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">--threads</span> 10 <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="at">--memory</span> 120 <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">-k</span> 21,33,55,77,99,127 <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">-o</span> ERR435025_assembly</span></code></pre>
</div>
<p>Adjust:</p>
<ul>
<li>input paths (<code>-1</code>, <code>-2</code>) to match your
dataset,</li>
<li>
<code>--threads</code> based on the number of cores available,</li>
<li>
<code>--memory</code> based on server limits,</li>
<li>
<code>-k</code> if needed (these k-mers are typical for Illumina
reads).</li>
</ul>
<div id="running-spades-in-a-long-session" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="running-spades-in-a-long-session" class="callout-inner">
<h3 class="callout-title">Running SPAdes in a long session</h3>
<div class="callout-content">
<p>Genome assembly can take <strong>many minutes to hours</strong>,
depending on:</p>
<ul>
<li>read depth,</li>
<li>genome size,</li>
<li>server load,</li>
<li>number of threads and memory.</li>
</ul>
<p>On shared systems, it is common to run SPAdes inside a session
manager like <code>screen</code> or <code>tmux</code> so that the
process keeps running even if your SSH connection drops.</p>
<p>Your instructor may demonstrate this, or the run may be
pre-computed.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="inspecting-spades-output">Inspecting SPAdes output<a class="anchor" aria-label="anchor" href="#inspecting-spades-output"></a>
</h2>
<hr class="half-width">
<p>Once SPAdes finishes, you should have an output directory (for
example <code>ERR435025_assembly/</code>):</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">ls</span> ERR435025_assembly</span></code></pre>
</div>
<p>Common files and directories include:</p>
<ul>
<li>
<code>contigs.fasta</code> – assembled contigs</li>
<li>
<code>scaffolds.fasta</code> – assembled scaffolds (contigs linked
by paired-end information)</li>
<li>
<code>spades.log</code> – log file describing the run</li>
<li>
<code>params.txt</code> – parameters used</li>
<li>other intermediate files and reports</li>
</ul>
<p>We will usually use <strong><code>scaffolds.fasta</code></strong> (or
sometimes <code>contigs.fasta</code>) as the input for downstream
steps.</p>
<p>Copy the scaffold file into a convenient location:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">cp</span> ERR435025_assembly/scaffolds.fasta .</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<p>You should now see <code>scaffolds.fasta</code> in your
<code>assembly/</code> directory.</p>
</section><section><h2 class="section-heading" id="first-look-at-the-assembly">First look at the assembly<a class="anchor" aria-label="anchor" href="#first-look-at-the-assembly"></a>
</h2>
<hr class="half-width">
<p>Use commands from earlier Unix episodes to inspect the assembly:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span> scaffolds.fasta</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">'^&gt;'</span> scaffolds.fasta</span></code></pre>
</div>
<ul>
<li>
<code>head</code> shows the first few contigs.</li>
<li>
<code>grep -c '^&gt;'</code> counts how many contigs are present
(each header starts with <code>&gt;</code>).</li>
</ul>
<p>If helper scripts are available (as in your assembly notes), you can
compute stats such as:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">seqstat</span> scaffolds.fasta</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">N50_count.pl</span> scaffolds.fasta</span></code></pre>
</div>
<p>These provide:</p>
<ul>
<li>total assembly length,</li>
<li>number of contigs,</li>
<li>N50 (a measure of contig length distribution).</li>
</ul>
<p>Later, you may want to filter contigs by length (for example keeping
only those ≥ 1,000 bp) before annotation, but the basic SPAdes assembly
output is already enough to proceed to Prokka.</p>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-run-spades-or-examine-a-prepared-run" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-spades-or-examine-a-prepared-run" class="callout-inner">
<h3 class="callout-title">Exercise: Run SPAdes (or examine a prepared
run)</h3>
<div class="callout-content">
<p>If computing resources allow, run SPAdes on your sample:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> myproject/assembly</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="ex">spades.py</span> <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">-1</span> ../raw_reads/sample01_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="at">-2</span> ../raw_reads/sample01_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">--threads</span> 8 <span class="dt">\</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="at">--memory</span> 32 <span class="dt">\</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>  <span class="at">-k</span> 21,33,55,77,99,127 <span class="dt">\</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="at">-o</span> sample01_assembly</span></code></pre>
</div>
<p>If SPAdes has been pre-run for you, skip directly to examining the
output directory.</p>
<p><strong>Questions:</strong></p>
<ul>
<li>How long did the run take?</li>
<li>Did SPAdes report any warnings or errors in
<code>spades.log</code>?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-count-contigs-and-inspect-the-assembly" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-count-contigs-and-inspect-the-assembly" class="callout-inner">
<h3 class="callout-title">Exercise: Count contigs and inspect the
assembly</h3>
<div class="callout-content">
<p>After SPAdes finishes (or using a provided assembly):</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> myproject/assembly</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">cp</span> sample01_assembly/scaffolds.fasta .</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">'^&gt;'</span> scaffolds.fasta</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="fu">head</span> scaffolds.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>How many contigs are present?</li>
<li>What does the first header line look like?</li>
<li>How might you identify the length of a specific contig?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-think-ahead-to-annotation" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-think-ahead-to-annotation" class="callout-inner">
<h3 class="callout-title">Exercise: Think ahead to annotation</h3>
<div class="callout-content">
<p>Discuss with a partner:</p>
<ol style="list-style-type: decimal">
<li>Why might we want to filter out very short contigs before
annotation?</li>
<li>Which file from the SPAdes output will we give to Prokka in the next
episode?</li>
<li>What information about the assembly might be important to record in
your project metadata?</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>SPAdes performs de novo genome assembly from paired-end reads.</li>
<li>You must provide paths to R1 and R2 reads, thread count, memory
limit, and an output directory.</li>
<li>SPAdes produces many files; <code>contigs.fasta</code> and
<code>scaffolds.fasta</code> are the most important for downstream
analysis.</li>
<li>Basic inspection of the assembly (number of contigs, total size,
N50) helps assess quality.</li>
<li>The assembly FASTA file will be the input to Prokka for genome
annotation.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-11-annotation-prokka"><p>Content from <a href="11-annotation-prokka.html">Genome Annotation with Prokka</a></p>
<hr>
<p>Last updated on 2025-12-04 |

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/episodes/11-annotation-prokka.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is genome annotation, and why do we perform it after
assembly?</li>
<li>How do we run Prokka on an assembled microbial genome?</li>
<li>What Prokka options are most important for typical bacterial
genomes?</li>
<li>What kinds of output files does Prokka produce, and how can we use
them?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the purpose of genome annotation in microbial genomics.</li>
<li>Run Prokka on an assembled genome using sensible options.</li>
<li>Interpret the most important annotation output files (GFF, FAA, FFN,
GBK, etc.).</li>
<li>Understand the difference between annotating a single genome, a bin,
or a metagenome.</li>
<li>Connect annotation results to downstream analyses (comparative
genomics, visualization, etc.).</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<blockquote>
<p>This episode is based on your Prokka annotation exercise and follows
directly from the SPAdes assembly in Episode 10. We assume that a
scaffold FASTA file is available and that contigs have already been
filtered to a reasonable minimum length.</p>
</blockquote>
<section><h2 class="section-heading" id="what-is-genome-annotation">What is genome annotation?<a class="anchor" aria-label="anchor" href="#what-is-genome-annotation"></a>
</h2>
<hr class="half-width">
<p>Genome annotation is the process of:</p>
<ul>
<li>finding genes (CDS) and other features (tRNAs, rRNAs),</li>
<li>predicting their functions,</li>
<li>assigning standardized names and identifiers.</li>
</ul>
<p>For a microbial genome, annotation helps you answer:</p>
<ul>
<li>How many protein-coding genes are present?</li>
<li>Are there plasmids, phage, or mobile elements?</li>
<li>Which pathways or functions are encoded?</li>
</ul>
<p><strong>Prokka</strong> is a widely used tool for <strong>rapid,
automated annotation</strong> of bacterial and archaeal genomes.</p>
</section><section><h2 class="section-heading" id="inputs-and-outputs-of-prokka">Inputs and outputs of Prokka<a class="anchor" aria-label="anchor" href="#inputs-and-outputs-of-prokka"></a>
</h2>
<hr class="half-width">
<p>Prokka expects:</p>
<ul>
<li>a FASTA file of contigs (e.g. scaffolds from SPAdes),</li>
<li>basic options describing the run and the organism.</li>
</ul>
<p>It produces:</p>
<ul>
<li>
<code>.gff</code> – combined sequence and annotation</li>
<li>
<code>.gbk</code> – GenBank-format file</li>
<li>
<code>.faa</code> – protein sequences</li>
<li>
<code>.ffn</code> – nucleotide sequences of genes</li>
<li>
<code>.fna</code> – nucleotide sequences of contigs</li>
<li>
<code>.tbl</code> – feature table</li>
<li>
<code>.txt</code> – plain-text summary</li>
<li>other helper files</li>
</ul>
<p>These outputs can be loaded into genome browsers, comparative
genomics tools, or downstream pipelines.</p>
</section><section><h2 class="section-heading" id="basic-prokka-command">Basic Prokka command<a class="anchor" aria-label="anchor" href="#basic-prokka-command"></a>
</h2>
<hr class="half-width">
<p>From your Prokka exercise, the generic form is:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="pp">[</span><span class="ss">options</span><span class="pp">]</span> mysequences.fna</span></code></pre>
</div>
<p>For a typical genome or metagenome bin, your notes suggest:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="dt">\</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="at">--outdir</span> OUTPUTDIRECTORY <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="at">--prefix</span> OUTPUTFILE_PREFIX <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">--addgenes</span> <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  <span class="at">--locustag</span> MYGEN <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">--increment</span> 5 <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">--cpus</span> 20 <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  <span class="at">--mincontiglen</span> 1000 <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  MYSEQUENCEFILE</span></code></pre>
</div>
<p>We will adapt this to our workshop data.</p>
</section><section><h2 class="section-heading" id="running-prokka-on-a-spades-assembly">Running Prokka on a SPAdes assembly<a class="anchor" aria-label="anchor" href="#running-prokka-on-a-spades-assembly"></a>
</h2>
<hr class="half-width">
<p>Assume you have a scaffold FASTA file from Episode 10, such as:</p>
<ul>
<li><code>scaffolds.fasta</code></li>
<li>or a filtered version with contigs ≥ 1000 bp.</li>
</ul>
<p>From your <code>annotation/</code> directory:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> myproject/annotation</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">cp</span> ../assembly/scaffolds.fasta sample01.scaffolds.fasta</span></code></pre>
</div>
<p>Run Prokka:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  <span class="at">--outdir</span> sample01_prokka <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">--prefix</span> sample01 <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">--addgenes</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">--locustag</span> S01 <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">--increment</span> 5 <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">--cpus</span> 8 <span class="dt">\</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">--mincontiglen</span> 1000 <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  sample01.scaffolds.fasta</span></code></pre>
</div>
<p>Adjust:</p>
<ul>
<li>
<code>--cpus</code> for your system,</li>
<li>
<code>--locustag</code> and <code>--prefix</code> for your organism
or project,</li>
<li>
<code>--mincontiglen</code> if you want a different minimum contig
size.</li>
</ul>
<div id="prokka-for-metagenomes" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="prokka-for-metagenomes" class="callout-inner">
<h3 class="callout-title">Prokka for metagenomes</h3>
<div class="callout-content">
<p>Your notes distinguish:</p>
<ul>
<li>
<p><strong>Genome or bin</strong> annotation:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="at">--outdir</span> OUT <span class="at">--prefix</span> PREFIX <span class="at">--addgenes</span> <span class="at">--locustag</span> TAG <span class="at">--increment</span> 5 <span class="at">--cpus</span> 20 <span class="at">--mincontiglen</span> 1000 MYSEQUENCEFILE</span></code></pre>
</div>
</li>
<li>
<p><strong>Entire metagenome</strong> annotation:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">prokka</span> <span class="at">--outdir</span> OUT <span class="at">--prefix</span> PREFIX <span class="at">--addgenes</span> <span class="at">--metagenome</span> <span class="at">--locustag</span> TAG <span class="at">--increment</span> 5 <span class="at">--cpus</span> 20 <span class="at">--mincontiglen</span> 1000 MYSEQUENCEFILE</span></code></pre>
</div>
</li>
</ul>
<p>The <code>--metagenome</code> flag adjusts gene prediction for
fragmented genomes.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="inspecting-prokka-output">Inspecting Prokka output<a class="anchor" aria-label="anchor" href="#inspecting-prokka-output"></a>
</h2>
<hr class="half-width">
<p>After Prokka finishes, list the output directory:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">ls</span> sample01_prokka</span></code></pre>
</div>
<p>Look for:</p>
<ul>
<li><code>sample01.gff</code></li>
<li><code>sample01.gbk</code></li>
<li><code>sample01.faa</code></li>
<li><code>sample01.ffn</code></li>
<li><code>sample01.fna</code></li>
<li><code>sample01.tbl</code></li>
<li><code>sample01.txt</code></li>
</ul>
<p>You can inspect the summary file:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">less</span> sample01.txt</span></code></pre>
</div>
<p>It typically includes:</p>
<ul>
<li>number of CDS,</li>
<li>number of tRNAs and rRNAs,</li>
<li>contig statistics,</li>
<li>basic annotation summary.</li>
</ul>
<p>View the GFF header:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">head</span> sample01.gff</span></code></pre>
</div>
<p>This file is especially useful because it:</p>
<ul>
<li>combines sequence,</li>
<li>feature coordinates,</li>
<li>annotations,</li>
</ul>
<p>in a single standardized format.</p>
</section><section><h2 class="section-heading" id="using-annotation-results">Using annotation results<a class="anchor" aria-label="anchor" href="#using-annotation-results"></a>
</h2>
<hr class="half-width">
<p>Annotation outputs can be used for:</p>
<ul>
<li>
<strong>exploratory analysis</strong> – counting genes, inspecting
interesting CDS,</li>
<li>
<strong>comparative genomics</strong> – clustering proteins from
<code>.faa</code> files,</li>
<li>
<strong>visualization</strong> – loading <code>.gbk</code> or
<code>.gff</code> into genome browsers,</li>
<li>
<strong>downstream pipelines</strong> – feeding <code>.gff</code>
and <code>.fna</code> into other tools.</li>
</ul>
<p>In this workshop, we focus on:</p>
<ul>
<li>understanding what Prokka did,</li>
<li>identifying key outcome metrics,</li>
<li>connecting the results back to the assembly used as input.</li>
</ul>
<hr></section><div class="section level1">
<h1 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h1>
<div id="exercise-run-prokka-on-your-assembly" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-run-prokka-on-your-assembly" class="callout-inner">
<h3 class="callout-title">Exercise: Run Prokka on your assembly</h3>
<div class="callout-content">
<p>Assuming your assembly FASTA is in <code>assembly/</code>:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> myproject/annotation</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">cp</span> ../assembly/scaffolds.fasta sample01.scaffolds.fasta</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="ex">prokka</span> <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  <span class="at">--outdir</span> sample01_prokka <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  <span class="at">--prefix</span> sample01 <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  <span class="at">--addgenes</span> <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  <span class="at">--locustag</span> S01 <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  <span class="at">--increment</span> 5 <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  <span class="at">--cpus</span> 4 <span class="dt">\</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  <span class="at">--mincontiglen</span> 1000 <span class="dt">\</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  sample01.scaffolds.fasta</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>Did Prokka complete without errors?</li>
<li>How long did the run take?</li>
<li>How many CDS, tRNAs, and rRNAs were predicted (check
<code>sample01.txt</code>)?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-explore-annotation-files" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-explore-annotation-files" class="callout-inner">
<h3 class="callout-title">Exercise: Explore annotation files</h3>
<div class="callout-content">
<p>From within <code>sample01_prokka</code>:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">head</span> sample01.txt</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">head</span> sample01.gff</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">head</span> sample01.faa</span></code></pre>
</div>
<p><strong>Questions:</strong></p>
<ul>
<li>What information is summarized in <code>sample01.txt</code>?</li>
<li>How are features represented in the GFF file?</li>
<li>What does a typical protein header in <code>sample01.faa</code> look
like?</li>
</ul>
</div>
</div>
</div>
<div id="exercise-genome-vs-metagenome-mode" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise-genome-vs-metagenome-mode" class="callout-inner">
<h3 class="callout-title">Exercise: Genome vs metagenome mode</h3>
<div class="callout-content">
<p>Discuss with a partner:</p>
<ol style="list-style-type: decimal">
<li>When would you use <code>--metagenome</code> with Prokka?</li>
<li>What kinds of datasets (e.g., bins, MAGs, entire metagenomes) might
benefit from <code>--metagenome</code>?</li>
<li>For the assembly produced in this workshop, is
<code>--metagenome</code> appropriate or not?</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Prokka performs rapid, automated annotation of bacterial and
archaeal genomes.</li>
<li>You must provide an assembly FASTA file and specify sensible options
(output directory, prefix, locus tag, CPUs, minimum contig length).</li>
<li>Prokka generates many outputs; the most important include
<code>.gff</code>, <code>.gbk</code>, <code>.faa</code>,
<code>.ffn</code>, <code>.fna</code>, and <code>.txt</code>.</li>
<li>Annotation results summarize genes, rRNAs, tRNAs, and other
features, and can be used in downstream comparative and visualization
workflows.</li>
<li>The annotated genome is the culmination of the QC → trimming →
assembly → annotation pipeline used in this workshop.</li>
</ul>
</div>
</div>
</div>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/jenna-tb-ekwealor/microbial-genomics/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:ekwealor@sfsu.edu">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.3" class="external-link">sandpaper (0.17.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.9" class="external-link">varnish (1.0.9)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "genomics, microbiome, assembly, bioinformatics, Unix shell, sequencing, annotation, contigs",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "identifier": "https://jenna-tb-ekwealor.github.io/microbial-genomics/instructor/aio.html",
  "dateCreated": "2025-12-25",
  "dateModified": "2025-12-04",
  "datePublished": "2025-12-04"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

